--- a/com/mojang/math/Matrix4f.java
+++ b/com/mojang/math/Matrix4f.java
@@ -2,25 +2,26 @@
 
 import java.nio.FloatBuffer;
 import java.util.Random;
+import net.minecraft.world.phys.Vec3;
 
 public final class Matrix4f
 {
-    protected float m00;
-    protected float m01;
-    protected float m02;
-    protected float m03;
-    protected float m10;
-    protected float m11;
-    protected float m12;
-    protected float m13;
-    protected float m20;
-    protected float m21;
-    protected float m22;
-    protected float m23;
-    protected float m30;
-    protected float m31;
-    protected float m32;
-    protected float m33;
+    public float m00;
+    public float m01;
+    public float m02;
+    public float m03;
+    public float m10;
+    public float m11;
+    public float m12;
+    public float m13;
+    public float m20;
+    public float m21;
+    public float m22;
+    public float m23;
+    public float m30;
+    public float m31;
+    public float m32;
+    public float m33;
 
     public Matrix4f()
     {
@@ -467,165 +468,191 @@
         return matrix4f;
     }
 
-    public float getTransformX(float x, float y, float z, float w)
+    public float getTransformX(float p_getTransformX_1_, float p_getTransformX_2_, float p_getTransformX_3_, float p_getTransformX_4_)
     {
-        return this.m00 * x + this.m01 * y + this.m02 * z + this.m03 * w;
+        return this.m00 * p_getTransformX_1_ + this.m01 * p_getTransformX_2_ + this.m02 * p_getTransformX_3_ + this.m03 * p_getTransformX_4_;
     }
 
-    public float getTransformY(float x, float y, float z, float w)
+    public float getTransformY(float p_getTransformY_1_, float p_getTransformY_2_, float p_getTransformY_3_, float p_getTransformY_4_)
     {
-        return this.m10 * x + this.m11 * y + this.m12 * z + this.m13 * w;
+        return this.m10 * p_getTransformY_1_ + this.m11 * p_getTransformY_2_ + this.m12 * p_getTransformY_3_ + this.m13 * p_getTransformY_4_;
     }
 
-    public float getTransformZ(float x, float y, float z, float w)
-    {
-        return this.m20 * x + this.m21 * y + this.m22 * z + this.m23 * w;
-    }
-
-    public float getTransformW(float x, float y, float z, float w)
-    {
-        return this.m30 * x + this.m31 * y + this.m32 * z + this.m33 * w;
-    }
-
-    public void mulTranslate(float xIn, float yIn, float zIn)
-    {
-        this.m03 += this.m00 * xIn + this.m01 * yIn + this.m02 * zIn;
-        this.m13 += this.m10 * xIn + this.m11 * yIn + this.m12 * zIn;
-        this.m23 += this.m20 * xIn + this.m21 * yIn + this.m22 * zIn;
-        this.m33 += this.m30 * xIn + this.m31 * yIn + this.m32 * zIn;
-    }
-
-    public void mulScale(float xIn, float yIn, float zIn)
-    {
-        this.m00 *= xIn;
-        this.m01 *= yIn;
-        this.m02 *= zIn;
-        this.m10 *= xIn;
-        this.m11 *= yIn;
-        this.m12 *= zIn;
-        this.m20 *= xIn;
-        this.m21 *= yIn;
-        this.m22 *= zIn;
-        this.m30 *= xIn;
-        this.m31 *= yIn;
-        this.m32 *= zIn;
-    }
-
-    public void setRandom(Random r)
-    {
-        this.m00 = r.nextFloat();
-        this.m01 = r.nextFloat();
-        this.m02 = r.nextFloat();
-        this.m03 = r.nextFloat();
-        this.m10 = r.nextFloat();
-        this.m11 = r.nextFloat();
-        this.m12 = r.nextFloat();
-        this.m13 = r.nextFloat();
-        this.m20 = r.nextFloat();
-        this.m21 = r.nextFloat();
-        this.m22 = r.nextFloat();
-        this.m23 = r.nextFloat();
-        this.m30 = r.nextFloat();
-        this.m31 = r.nextFloat();
-        this.m32 = r.nextFloat();
-        this.m33 = r.nextFloat();
-    }
-
-    public void write(float[] floatArrayIn)
-    {
-        floatArrayIn[0] = this.m00;
-        floatArrayIn[1] = this.m01;
-        floatArrayIn[2] = this.m02;
-        floatArrayIn[3] = this.m03;
-        floatArrayIn[4] = this.m10;
-        floatArrayIn[5] = this.m11;
-        floatArrayIn[6] = this.m12;
-        floatArrayIn[7] = this.m13;
-        floatArrayIn[8] = this.m20;
-        floatArrayIn[9] = this.m21;
-        floatArrayIn[10] = this.m22;
-        floatArrayIn[11] = this.m23;
-        floatArrayIn[12] = this.m30;
-        floatArrayIn[13] = this.m31;
-        floatArrayIn[14] = this.m32;
-        floatArrayIn[15] = this.m33;
-    }
-
-    public Matrix4f(float[] values)
-    {
-        this.m00 = values[0];
-        this.m01 = values[1];
-        this.m02 = values[2];
-        this.m03 = values[3];
-        this.m10 = values[4];
-        this.m11 = values[5];
-        this.m12 = values[6];
-        this.m13 = values[7];
-        this.m20 = values[8];
-        this.m21 = values[9];
-        this.m22 = values[10];
-        this.m23 = values[11];
-        this.m30 = values[12];
-        this.m31 = values[13];
-        this.m32 = values[14];
-        this.m33 = values[15];
-    }
-
-    public void set(Matrix4f mat)
-    {
-        this.m00 = mat.m00;
-        this.m01 = mat.m01;
-        this.m02 = mat.m02;
-        this.m03 = mat.m03;
-        this.m10 = mat.m10;
-        this.m11 = mat.m11;
-        this.m12 = mat.m12;
-        this.m13 = mat.m13;
-        this.m20 = mat.m20;
-        this.m21 = mat.m21;
-        this.m22 = mat.m22;
-        this.m23 = mat.m23;
-        this.m30 = mat.m30;
-        this.m31 = mat.m31;
-        this.m32 = mat.m32;
-        this.m33 = mat.m33;
-    }
-
-    public void add(Matrix4f other)
-    {
-        this.m00 += other.m00;
-        this.m01 += other.m01;
-        this.m02 += other.m02;
-        this.m03 += other.m03;
-        this.m10 += other.m10;
-        this.m11 += other.m11;
-        this.m12 += other.m12;
-        this.m13 += other.m13;
-        this.m20 += other.m20;
-        this.m21 += other.m21;
-        this.m22 += other.m22;
-        this.m23 += other.m23;
-        this.m30 += other.m30;
-        this.m31 += other.m31;
-        this.m32 += other.m32;
-        this.m33 += other.m33;
+    public float getTransformZ(float p_getTransformZ_1_, float p_getTransformZ_2_, float p_getTransformZ_3_, float p_getTransformZ_4_)
+    {
+        return this.m20 * p_getTransformZ_1_ + this.m21 * p_getTransformZ_2_ + this.m22 * p_getTransformZ_3_ + this.m23 * p_getTransformZ_4_;
+    }
+
+    public float getTransformW(float p_getTransformW_1_, float p_getTransformW_2_, float p_getTransformW_3_, float p_getTransformW_4_)
+    {
+        return this.m30 * p_getTransformW_1_ + this.m31 * p_getTransformW_2_ + this.m32 * p_getTransformW_3_ + this.m33 * p_getTransformW_4_;
+    }
+
+    public void mulTranslate(float p_mulTranslate_1_, float p_mulTranslate_2_, float p_mulTranslate_3_)
+    {
+        this.m03 += this.m00 * p_mulTranslate_1_ + this.m01 * p_mulTranslate_2_ + this.m02 * p_mulTranslate_3_;
+        this.m13 += this.m10 * p_mulTranslate_1_ + this.m11 * p_mulTranslate_2_ + this.m12 * p_mulTranslate_3_;
+        this.m23 += this.m20 * p_mulTranslate_1_ + this.m21 * p_mulTranslate_2_ + this.m22 * p_mulTranslate_3_;
+        this.m33 += this.m30 * p_mulTranslate_1_ + this.m31 * p_mulTranslate_2_ + this.m32 * p_mulTranslate_3_;
+    }
+
+    public void mulScale(float p_mulScale_1_, float p_mulScale_2_, float p_mulScale_3_)
+    {
+        this.m00 *= p_mulScale_1_;
+        this.m01 *= p_mulScale_2_;
+        this.m02 *= p_mulScale_3_;
+        this.m10 *= p_mulScale_1_;
+        this.m11 *= p_mulScale_2_;
+        this.m12 *= p_mulScale_3_;
+        this.m20 *= p_mulScale_1_;
+        this.m21 *= p_mulScale_2_;
+        this.m22 *= p_mulScale_3_;
+        this.m30 *= p_mulScale_1_;
+        this.m31 *= p_mulScale_2_;
+        this.m32 *= p_mulScale_3_;
+    }
+
+    public void setRandom(Random p_setRandom_1_)
+    {
+        this.m00 = p_setRandom_1_.nextFloat();
+        this.m01 = p_setRandom_1_.nextFloat();
+        this.m02 = p_setRandom_1_.nextFloat();
+        this.m03 = p_setRandom_1_.nextFloat();
+        this.m10 = p_setRandom_1_.nextFloat();
+        this.m11 = p_setRandom_1_.nextFloat();
+        this.m12 = p_setRandom_1_.nextFloat();
+        this.m13 = p_setRandom_1_.nextFloat();
+        this.m20 = p_setRandom_1_.nextFloat();
+        this.m21 = p_setRandom_1_.nextFloat();
+        this.m22 = p_setRandom_1_.nextFloat();
+        this.m23 = p_setRandom_1_.nextFloat();
+        this.m30 = p_setRandom_1_.nextFloat();
+        this.m31 = p_setRandom_1_.nextFloat();
+        this.m32 = p_setRandom_1_.nextFloat();
+        this.m33 = p_setRandom_1_.nextFloat();
+    }
+
+    public void write(float[] p_write_1_)
+    {
+        p_write_1_[0] = this.m00;
+        p_write_1_[1] = this.m01;
+        p_write_1_[2] = this.m02;
+        p_write_1_[3] = this.m03;
+        p_write_1_[4] = this.m10;
+        p_write_1_[5] = this.m11;
+        p_write_1_[6] = this.m12;
+        p_write_1_[7] = this.m13;
+        p_write_1_[8] = this.m20;
+        p_write_1_[9] = this.m21;
+        p_write_1_[10] = this.m22;
+        p_write_1_[11] = this.m23;
+        p_write_1_[12] = this.m30;
+        p_write_1_[13] = this.m31;
+        p_write_1_[14] = this.m32;
+        p_write_1_[15] = this.m33;
+    }
+
+    public Matrix4f(float[] p_i242105_1_)
+    {
+        this.m00 = p_i242105_1_[0];
+        this.m01 = p_i242105_1_[1];
+        this.m02 = p_i242105_1_[2];
+        this.m03 = p_i242105_1_[3];
+        this.m10 = p_i242105_1_[4];
+        this.m11 = p_i242105_1_[5];
+        this.m12 = p_i242105_1_[6];
+        this.m13 = p_i242105_1_[7];
+        this.m20 = p_i242105_1_[8];
+        this.m21 = p_i242105_1_[9];
+        this.m22 = p_i242105_1_[10];
+        this.m23 = p_i242105_1_[11];
+        this.m30 = p_i242105_1_[12];
+        this.m31 = p_i242105_1_[13];
+        this.m32 = p_i242105_1_[14];
+        this.m33 = p_i242105_1_[15];
+    }
+
+    public void set(Matrix4f p_set_1_)
+    {
+        this.m00 = p_set_1_.m00;
+        this.m01 = p_set_1_.m01;
+        this.m02 = p_set_1_.m02;
+        this.m03 = p_set_1_.m03;
+        this.m10 = p_set_1_.m10;
+        this.m11 = p_set_1_.m11;
+        this.m12 = p_set_1_.m12;
+        this.m13 = p_set_1_.m13;
+        this.m20 = p_set_1_.m20;
+        this.m21 = p_set_1_.m21;
+        this.m22 = p_set_1_.m22;
+        this.m23 = p_set_1_.m23;
+        this.m30 = p_set_1_.m30;
+        this.m31 = p_set_1_.m31;
+        this.m32 = p_set_1_.m32;
+        this.m33 = p_set_1_.m33;
+    }
+
+    public void add(Matrix4f p_add_1_)
+    {
+        this.m00 += p_add_1_.m00;
+        this.m01 += p_add_1_.m01;
+        this.m02 += p_add_1_.m02;
+        this.m03 += p_add_1_.m03;
+        this.m10 += p_add_1_.m10;
+        this.m11 += p_add_1_.m11;
+        this.m12 += p_add_1_.m12;
+        this.m13 += p_add_1_.m13;
+        this.m20 += p_add_1_.m20;
+        this.m21 += p_add_1_.m21;
+        this.m22 += p_add_1_.m22;
+        this.m23 += p_add_1_.m23;
+        this.m30 += p_add_1_.m30;
+        this.m31 += p_add_1_.m31;
+        this.m32 += p_add_1_.m32;
+        this.m33 += p_add_1_.m33;
     }
 
-    public void multiplyBackward(Matrix4f other)
+    public void multiplyBackward(Matrix4f p_multiplyBackward_1_)
     {
-        Matrix4f matrix4f = other.copy();
+        Matrix4f matrix4f = p_multiplyBackward_1_.copy();
         matrix4f.multiply(this);
         this.set(matrix4f);
     }
 
-    public void setTranslation(float x, float y, float z)
+    public void setTranslation(float p_setTranslation_1_, float p_setTranslation_2_, float p_setTranslation_3_)
     {
         this.m00 = 1.0F;
         this.m11 = 1.0F;
         this.m22 = 1.0F;
         this.m33 = 1.0F;
-        this.m03 = x;
-        this.m13 = y;
-        this.m23 = z;
+        this.m03 = p_setTranslation_1_;
+        this.m13 = p_setTranslation_2_;
+        this.m23 = p_setTranslation_3_;
+    }
+
+    @Deprecated
+    public void read(FloatBuffer floatBufferIn)
+    {
+        this.m00 = floatBufferIn.get(bufferIndex(0, 0));
+        this.m01 = floatBufferIn.get(bufferIndex(0, 1));
+        this.m02 = floatBufferIn.get(bufferIndex(0, 2));
+        this.m03 = floatBufferIn.get(bufferIndex(0, 3));
+        this.m10 = floatBufferIn.get(bufferIndex(1, 0));
+        this.m11 = floatBufferIn.get(bufferIndex(1, 1));
+        this.m12 = floatBufferIn.get(bufferIndex(1, 2));
+        this.m13 = floatBufferIn.get(bufferIndex(1, 3));
+        this.m20 = floatBufferIn.get(bufferIndex(2, 0));
+        this.m21 = floatBufferIn.get(bufferIndex(2, 1));
+        this.m22 = floatBufferIn.get(bufferIndex(2, 2));
+        this.m23 = floatBufferIn.get(bufferIndex(2, 3));
+        this.m30 = floatBufferIn.get(bufferIndex(3, 0));
+        this.m31 = floatBufferIn.get(bufferIndex(3, 1));
+        this.m32 = floatBufferIn.get(bufferIndex(3, 2));
+        this.m33 = floatBufferIn.get(bufferIndex(3, 3));
+    }
+
+    public Vec3 translationComponent()
+    {
+        return new Vec3((double)this.m30, (double)this.m31, (double)this.m32);
     }
 }
