diff -r -U 3 minecraft\net\minecraft\server\level\TickingTracker.java minecraft_patched\net\minecraft\server\level\TickingTracker.java
--- minecraft\net\minecraft\server\level\TickingTracker.java
+++ minecraft_patched\net\minecraft\server\level\TickingTracker.java
@@ -65,7 +65,7 @@
       for(Entry<SortedArraySet<Ticket<?>>> entry : this.f_184143_.long2ObjectEntrySet()) {
          for(Ticket<?> ticket : entry.getValue()) {
             if (ticket.m_9428_() == TicketType.f_9444_) {
-               list.add(Pair.of(ticket, entry.getLongKey()));
+               list.add(Pair.of((Ticket<ChunkPos>)ticket, entry.getLongKey()));
             }
          }
       }

