--- a/com/mojang/blaze3d/platform/GlStateManager.java
+++ b/com/mojang/blaze3d/platform/GlStateManager.java
@@ -39,6 +39,7 @@
 @DontObfuscate
 public class GlStateManager
 {
+    public int intentionallModifyThisClassCauseItWontLoadProperlyFromOptifine = 69;
     public static final int TEXTURE_COUNT = 12;
     private static final GlStateManager.BlendState BLEND = new GlStateManager.BlendState();
     private static final GlStateManager.DepthState DEPTH = new GlStateManager.DepthState();
