--- a/com/mojang/blaze3d/platform/Window.java
+++ b/com/mojang/blaze3d/platform/Window.java
@@ -165,7 +165,6 @@
         GL.createCapabilities();
         this.setMode();
         this.refreshFramebufferSize();
-        GLFW.glfwSetFramebufferSizeCallback(this.window, this::onFramebufferResize);
         GLFW.glfwSetWindowPosCallback(this.window, this::onMove);
         GLFW.glfwSetWindowSizeCallback(this.window, this::onResize);
         GLFW.glfwSetWindowFocusCallback(this.window, this::onFocus);
@@ -296,9 +295,9 @@
                 memorystack.close();
             }
         }
-        catch (IOException ioexception1)
+        catch (IOException ioexception)
         {
-            LOGGER.error("Couldn't set icon", (Throwable)ioexception1);
+            LOGGER.error("Couldn't set icon", (Throwable)ioexception);
         }
     }
 
@@ -375,7 +374,7 @@
     {
         RenderSystem.assertThread(RenderSystem::isOnRenderThreadOrInit);
         this.vsync = pVsyncEnabled;
-        GLFW.glfwSwapInterval(pVsyncEnabled ? 1 : 0);
+        GLFW.glfwSwapInterval(0);
     }
 
     public void close()
@@ -398,15 +397,15 @@
     {
         if (pWindowPointer == this.window)
         {
-            int i = this.getWidth();
-            int j = this.getHeight();
+            int i = this.getScreenWidth();
+            int j = this.getScreenHeight();
 
             if (p_85417_ != 0 && pFramebufferWidth != 0)
             {
                 this.framebufferWidth = p_85417_;
                 this.framebufferHeight = pFramebufferWidth;
 
-                if (this.getWidth() != i || this.getHeight() != j)
+                if (this.getScreenWidth() != i || this.getScreenHeight() != j)
                 {
                     this.eventHandler.resizeDisplay();
                 }
@@ -425,20 +424,24 @@
 
         if (this.framebufferHeight == 0 || this.framebufferWidth == 0)
         {
-            EarlyProgressVisualization.INSTANCE.updateFBSize((w) ->
+            EarlyProgressVisualization.INSTANCE.updateFBSize((p_lambda$updateFramebufferSize$4_1_) ->
             {
-                this.framebufferWidth = w;
-            }, (h) ->
+                this.framebufferWidth = p_lambda$updateFramebufferSize$4_1_;
+            }, (p_lambda$updateFramebufferSize$5_1_) ->
             {
-                this.framebufferHeight = h;
+                this.framebufferHeight = p_lambda$updateFramebufferSize$5_1_;
             });
         }
     }
 
     private void onResize(long pWindowPointer, int p_85429_, int pWindowWidth)
     {
-        this.width = p_85429_;
-        this.height = pWindowWidth;
+        if (pWindowWidth * p_85429_ != 0)
+        {
+            this.width = p_85429_;
+            this.height = pWindowWidth;
+            this.eventHandler.resizeDisplay();
+        }
     }
 
     private void onFocus(long pWindowPointer, boolean p_85394_)
@@ -586,13 +589,12 @@
     {
         int i;
 
-        for (i = 1; i != pGuiScale && i < this.framebufferWidth && i < this.framebufferHeight && this.framebufferWidth / (i + 1) >= 320 && this.framebufferHeight / (i + 1) >= 240; ++i)
+        for (i = 1; i != pGuiScale && i < this.width && i < this.height && this.width / (i + 1) >= 320 && this.height / (i + 1) >= 240; ++i)
         {
         }
 
         if (pForceUnicode && i % 2 != 0)
         {
-            ++i;
         }
 
         return i;
@@ -601,10 +603,10 @@
     public void setGuiScale(double pScaleFactor)
     {
         this.guiScale = pScaleFactor;
-        int i = (int)((double)this.framebufferWidth / pScaleFactor);
-        this.guiScaledWidth = (double)this.framebufferWidth / pScaleFactor > (double)i ? i + 1 : i;
-        int j = (int)((double)this.framebufferHeight / pScaleFactor);
-        this.guiScaledHeight = (double)this.framebufferHeight / pScaleFactor > (double)j ? j + 1 : j;
+        int i = (int)((double)this.width / pScaleFactor);
+        this.guiScaledWidth = (double)this.width / pScaleFactor > (double)i ? i + 1 : i;
+        int j = (int)((double)this.height / pScaleFactor);
+        this.guiScaledHeight = (double)this.height / pScaleFactor > (double)j ? j + 1 : j;
     }
 
     public void setTitle(String pTitle)
@@ -624,12 +626,12 @@
 
     public int getWidth()
     {
-        return this.framebufferWidth;
+        return Minecraft.getInstance().getMainRenderTarget().viewWidth;
     }
 
     public int getHeight()
     {
-        return this.framebufferHeight;
+        return Minecraft.getInstance().getMainRenderTarget().viewHeight;
     }
 
     public void setWidth(int p_166451_)
@@ -688,9 +690,9 @@
         InputConstants.updateRawMouseInput(this.window, pValue);
     }
 
-    public void resizeFramebuffer(int width, int height)
+    public void resizeFramebuffer(int p_resizeFramebuffer_1_, int p_resizeFramebuffer_2_)
     {
-        this.onFramebufferResize(this.window, width, height);
+        this.onFramebufferResize(this.window, p_resizeFramebuffer_1_, p_resizeFramebuffer_2_);
     }
 
     public boolean isClosed()
