--- a/net/minecraft/client/main/Main.java
+++ b/net/minecraft/client/main/Main.java
@@ -54,6 +54,11 @@
         optionparser.accepts("demo");
         optionparser.accepts("disableMultiplayer");
         optionparser.accepts("disableChat");
+        optionparser.accepts("demo");
+        optionparser.accepts("kiosk");
+        optionparser.accepts("viewonly");
+        optionparser.accepts("katvr");
+        optionparser.accepts("infinadeck");
         optionparser.accepts("fullscreen");
         optionparser.accepts("checkGlErrors");
         OptionSpec<Void> optionspec = optionparser.accepts("jfrProfile");
@@ -196,6 +201,7 @@
         }
         catch (Throwable throwable1)
         {
+            LOGGER.warn("Failed to something: ", throwable1);
             CrashReport crashreport = CrashReport.forThrowable(throwable1, "Initializing game");
             CrashReportCategory crashreportcategory = crashreport.addCategory("Initialization");
             NativeModuleLister.addCrashSection(crashreportcategory);
@@ -255,9 +261,9 @@
                 thread1.join();
             }
         }
-        catch (InterruptedException interruptedexception)
+        catch (InterruptedException interruptedexception1)
         {
-            LOGGER.error("Exception during client thread shutdown", (Throwable)interruptedexception);
+            LOGGER.error("Exception during client thread shutdown", (Throwable)interruptedexception1);
         }
         finally
         {
