--- a/net/minecraft/client/particle/ParticleEngine.java
+++ b/net/minecraft/client/particle/ParticleEngine.java
@@ -346,7 +346,7 @@
     @Nullable
     private <T extends ParticleOptions> Particle makeParticle(T pParticleData, double pX, double p_107398_, double pY, double p_107400_, double pZ, double p_107402_)
     {
-        ParticleProvider<T> particleprovider = this.providers.get(Registry.PARTICLE_TYPE.getKey(pParticleData.getType()));
+        ParticleProvider<T> particleprovider = (ParticleProvider<T>) this.providers.get(Registry.PARTICLE_TYPE.getKey(pParticleData.getType()));
         return particleprovider == null ? null : particleprovider.createParticle(pParticleData, this.level, pX, p_107398_, pY, p_107400_, pZ, p_107402_);
     }
 
@@ -674,8 +674,14 @@
         return this.trackedParticleCounts.getInt(p_172280_) < p_172280_.getLimit();
     }
 
-    private boolean reuseBarrierParticle(Particle var1, Queue<Particle> var2)
+    private boolean reuseBarrierParticle(Particle var1, Queue<Particle> deque)
     {
+    	Particle var4;
+        for(Iterator it = deque.iterator(); it.hasNext(); var4 = (Particle)it.next()) {
+           ;
+        }
+
+        return false;
     }
 
     public void addBlockHitEffects(BlockPos pos, BlockHitResult target)
