--- a/net/optifine/Config.java
+++ b/net/optifine/Config.java
@@ -142,14 +142,15 @@
 
     public static void initGameSettings(Options settings)
     {
-        if (gameSettings == null)
-        {
-            gameSettings = settings;
-            updateAvailableProcessors();
-            ReflectorForge.putLaunchBlackboard("optifine.ForgeSplashCompatible", Boolean.TRUE);
-            antialiasingLevel = gameSettings.ofAaLevel;
-        }
-    }
+    	//Vivecraft why do we do this?
+    	//if (gameSettings == null)
+    	//{
+	    	gameSettings = settings;
+	    	updateAvailableProcessors();
+	    	ReflectorForge.putLaunchBlackboard("optifine.ForgeSplashCompatible", Boolean.TRUE);
+	    	antialiasingLevel = gameSettings.ofAaLevel;
+    	//}
+}
 
     public static void initDisplay()
     {
@@ -161,14 +162,11 @@
 
     public static void checkInitialized()
     {
-        if (!initialized)
+        if (!initialized && Minecraft.getInstance().getWindow() != null)
         {
-            if (Minecraft.getInstance().getWindow() != null)
-            {
-                initialized = true;
-                checkOpenGlCaps();
-                startVersionCheckThread();
-            }
+            initialized = true;
+            checkOpenGlCaps();
+            startVersionCheckThread();
         }
     }
 
@@ -401,9 +399,9 @@
                 }
             }
         }
-        catch (Exception exception)
+        catch (Exception exception1)
         {
-            error("", exception);
+            error("", exception1);
             return def;
         }
     }
@@ -451,9 +449,9 @@
             String s = GL11.glGetString(GL11.GL_EXTENSIONS);
             return s.split(" ");
         }
-        catch (Exception exception)
+        catch (Exception exception1)
         {
-            error("", exception);
+            error("", exception1);
             return new String[0];
         }
     }
@@ -508,9 +506,9 @@
                 }
             }
         }
-        catch (Throwable throwable)
+        catch (Throwable throwable1)
         {
-            warn(throwable.getClass().getName() + ": " + throwable.getMessage());
+            warn(throwable1.getClass().getName() + ": " + throwable1.getMessage());
         }
     }
 
@@ -1111,7 +1109,7 @@
             Resource resource = resourceManager.getResource(location);
             return resource != null;
         }
-        catch (IOException ioexception)
+        catch (IOException ioexception1)
         {
             return false;
         }
@@ -1165,7 +1163,7 @@
                     astring[i] = apackresources[i].getName();
                 }
 
-                return arrayToString((Object[])astring);
+                return arrayToString(astring);
             }
         }
     }
@@ -2293,14 +2291,7 @@
 
     public static boolean isIntegratedServerRunning()
     {
-        if (minecraft.getSingleplayerServer() == null)
-        {
-            return false;
-        }
-        else
-        {
-            return minecraft.isLocalServer();
-        }
+        return minecraft.getSingleplayerServer() == null ? false : minecraft.isLocalServer();
     }
 
     public static IntBuffer createDirectIntBuffer(int capacity)
