--- a/net/minecraft/client/KeyboardHandler.java
+++ b/net/minecraft/client/KeyboardHandler.java
@@ -43,6 +43,9 @@
 import net.optifine.reflect.Reflector;
 import net.optifine.shaders.Shaders;
 import net.optifine.shaders.gui.GuiShaderOptions;
+import org.vivecraft.gameplay.screenhandlers.RadialHandler;
+import org.vivecraft.provider.ControllerType;
+import org.vivecraft.settings.VRHotkeys;
 
 public class KeyboardHandler
 {
@@ -411,6 +414,26 @@
     {
         if (p_90894_ == this.minecraft.getWindow().getWindow())
         {
+            if (p_90895_ == 256 && p_90897_ == 1)
+            {
+                if (org.vivecraft.gameplay.screenhandlers.KeyboardHandler.Showing)
+                {
+                    org.vivecraft.gameplay.screenhandlers.KeyboardHandler.setOverlayShowing(false);
+                    return;
+                }
+
+                if (RadialHandler.isShowing())
+                {
+                    RadialHandler.setOverlayShowing(false, (ControllerType)null);
+                    return;
+                }
+            }
+
+            if (VRHotkeys.handleKeyboardInputs(p_90895_, p_90896_, p_90897_, p_90898_))
+            {
+                return;
+            }
+
             if (this.debugCrashKeyTime > 0L)
             {
                 if (!InputConstants.isKeyDown(Minecraft.getInstance().getWindow().getWindow(), 67) || !InputConstants.isKeyDown(Minecraft.getInstance().getWindow().getWindow(), 292))
@@ -442,16 +465,17 @@
 
                     if (this.minecraft.options.keyScreenshot.matches(p_90895_, p_90896_))
                     {
-                        if (Screen.hasControlDown())
-                        {
-                        }
-
-                        Screenshot.grab(this.minecraft.gameDirectory, this.minecraft.getWindow().getWidth(), this.minecraft.getWindow().getHeight(), this.minecraft.getMainRenderTarget(), (p_90916_1_) ->
-                        {
-                            this.minecraft.execute(() -> {
-                                this.minecraft.gui.getChat().addMessage(p_90916_1_);
-                            });
-                        });
+//                        if (Screen.hasControlDown())
+//                        {
+//                        }
+//
+//                        Screenshot.grab(this.minecraft.gameDirectory, this.minecraft.getWindow().getWidth(), this.minecraft.getWindow().getHeight(), this.minecraft.getMainRenderTarget(), (p_90916_1_) ->
+//                        {
+//                            this.minecraft.execute(() -> {
+//                                this.minecraft.gui.getChat().addMessage(p_90916_1_);
+//                            });
+//                        });
+                    	this.minecraft.grabScreenShot = true;
                         return;
                     }
                 }
@@ -537,7 +561,7 @@
                 }
             }
 
-            if (this.minecraft.screen == null || this.minecraft.screen.passEvents)
+            if ((this.minecraft.screen == null || this.minecraft.screen.passEvents) && !InputConstants.isKeyDown(345))
             {
                 InputConstants.Key inputconstants$key = InputConstants.getKey(p_90895_, p_90896_);
 
@@ -595,6 +619,7 @@
                         if (p_90895_ == 290)
                         {
                             this.minecraft.options.hideGui = !this.minecraft.options.hideGui;
+                            this.minecraft.vrSettings.saveOptions();
                         }
                     }
 
@@ -619,7 +644,7 @@
         }
     }
 
-    private void charTyped(long p_90890_, int p_90891_, int p_90892_)
+    public void charTyped(long p_90890_, int p_90891_, int p_90892_)
     {
         if (p_90890_ == this.minecraft.getWindow().getWindow())
         {
