--- a/net/minecraft/client/Option.java
+++ b/net/minecraft/client/Option.java
@@ -25,388 +25,393 @@
 
 public abstract class Option
 {
-    public static final ProgressOption BIOME_BLEND_RADIUS = new ProgressOption("options.biomeBlendRadius", 0.0D, 7.0D, 1.0F, (p_216607_0_) ->
+    public static final ProgressOption BIOME_BLEND_RADIUS = new ProgressOption("options.biomeBlendRadius", 0.0D, 7.0D, 1.0F, (p_lambda$static$0_0_) ->
     {
-        return (double)p_216607_0_.biomeBlendRadius;
-    }, (p_216660_0_, p_216660_1_) ->
+        return (double)p_lambda$static$0_0_.biomeBlendRadius;
+    }, (p_lambda$static$1_0_, p_lambda$static$1_1_) ->
     {
-        p_216660_0_.biomeBlendRadius = Mth.clamp((int)p_216660_1_.doubleValue(), 0, 7);
+        p_lambda$static$1_0_.biomeBlendRadius = Mth.clamp((int)p_lambda$static$1_1_.doubleValue(), 0, 7);
         Minecraft.getInstance().levelRenderer.allChanged();
-    }, (p_216595_0_, p_216595_1_) ->
+    }, (p_lambda$static$2_0_, p_lambda$static$2_1_) ->
     {
-        double d0 = p_216595_1_.get(p_216595_0_);
+        double d0 = p_lambda$static$2_1_.get(p_lambda$static$2_0_);
         int i = (int)d0 * 2 + 1;
-        return p_216595_1_.genericValueLabel(new TranslatableComponent("options.biomeBlendRadius." + i));
+        return p_lambda$static$2_1_.genericValueLabel(new TranslatableComponent("options.biomeBlendRadius." + i));
     });
-    public static final ProgressOption CHAT_HEIGHT_FOCUSED = new ProgressOption("options.chat.height.focused", 0.0D, 1.0D, 0.0F, (p_216587_0_) ->
+    public static final ProgressOption CHAT_HEIGHT_FOCUSED = new ProgressOption("options.chat.height.focused", 0.0D, 1.0D, 0.0F, (p_lambda$static$3_0_) ->
     {
-        return p_216587_0_.chatHeightFocused;
-    }, (p_216600_0_, p_216600_1_) ->
+        return p_lambda$static$3_0_.chatHeightFocused;
+    }, (p_lambda$static$4_0_, p_lambda$static$4_1_) ->
     {
-        p_216600_0_.chatHeightFocused = p_216600_1_;
+        p_lambda$static$4_0_.chatHeightFocused = p_lambda$static$4_1_;
         Minecraft.getInstance().gui.getChat().rescaleChat();
-    }, (p_216642_0_, p_216642_1_) ->
+    }, (p_lambda$static$5_0_, p_lambda$static$5_1_) ->
     {
-        double d0 = p_216642_1_.toPct(p_216642_1_.get(p_216642_0_));
-        return p_216642_1_.pixelValueLabel(ChatComponent.getHeight(d0));
+        double d0 = p_lambda$static$5_1_.toPct(p_lambda$static$5_1_.get(p_lambda$static$5_0_));
+        return p_lambda$static$5_1_.pixelValueLabel(ChatComponent.getHeight(d0));
     });
-    public static final ProgressOption CHAT_HEIGHT_UNFOCUSED = new ProgressOption("options.chat.height.unfocused", 0.0D, 1.0D, 0.0F, (p_216611_0_) ->
+    public static final ProgressOption CHAT_HEIGHT_UNFOCUSED = new ProgressOption("options.chat.height.unfocused", 0.0D, 1.0D, 0.0F, (p_lambda$static$6_0_) ->
     {
-        return p_216611_0_.chatHeightUnfocused;
-    }, (p_216650_0_, p_216650_1_) ->
+        return p_lambda$static$6_0_.chatHeightUnfocused;
+    }, (p_lambda$static$7_0_, p_lambda$static$7_1_) ->
     {
-        p_216650_0_.chatHeightUnfocused = p_216650_1_;
+        p_lambda$static$7_0_.chatHeightUnfocused = p_lambda$static$7_1_;
         Minecraft.getInstance().gui.getChat().rescaleChat();
-    }, (p_216604_0_, p_216604_1_) ->
+    }, (p_lambda$static$8_0_, p_lambda$static$8_1_) ->
     {
-        double d0 = p_216604_1_.toPct(p_216604_1_.get(p_216604_0_));
-        return p_216604_1_.pixelValueLabel(ChatComponent.getHeight(d0));
+        double d0 = p_lambda$static$8_1_.toPct(p_lambda$static$8_1_.get(p_lambda$static$8_0_));
+        return p_lambda$static$8_1_.pixelValueLabel(ChatComponent.getHeight(d0));
     });
-    public static final ProgressOption CHAT_OPACITY = new ProgressOption("options.chat.opacity", 0.0D, 1.0D, 0.0F, (p_216649_0_) ->
+    public static final ProgressOption CHAT_OPACITY = new ProgressOption("options.chat.opacity", 0.0D, 1.0D, 0.0F, (p_lambda$static$9_0_) ->
     {
-        return p_216649_0_.chatOpacity;
-    }, (p_216578_0_, p_216578_1_) ->
+        return p_lambda$static$9_0_.chatOpacity;
+    }, (p_lambda$static$10_0_, p_lambda$static$10_1_) ->
     {
-        p_216578_0_.chatOpacity = p_216578_1_;
+        p_lambda$static$10_0_.chatOpacity = p_lambda$static$10_1_;
         Minecraft.getInstance().gui.getChat().rescaleChat();
-    }, (p_216592_0_, p_216592_1_) ->
+    }, (p_lambda$static$11_0_, p_lambda$static$11_1_) ->
     {
-        double d0 = p_216592_1_.toPct(p_216592_1_.get(p_216592_0_));
-        return p_216592_1_.percentValueLabel(d0 * 0.9D + 0.1D);
+        double d0 = p_lambda$static$11_1_.toPct(p_lambda$static$11_1_.get(p_lambda$static$11_0_));
+        return p_lambda$static$11_1_.percentValueLabel(d0 * 0.9D + 0.1D);
     });
-    public static final ProgressOption CHAT_SCALE = new ProgressOption("options.chat.scale", 0.0D, 1.0D, 0.0F, (p_216591_0_) ->
+    public static final ProgressOption CHAT_SCALE = new ProgressOption("options.chat.scale", 0.0D, 1.0D, 0.0F, (p_lambda$static$12_0_) ->
     {
-        return p_216591_0_.chatScale;
-    }, (p_216624_0_, p_216624_1_) ->
+        return p_lambda$static$12_0_.chatScale;
+    }, (p_lambda$static$13_0_, p_lambda$static$13_1_) ->
     {
-        p_216624_0_.chatScale = p_216624_1_;
+        p_lambda$static$13_0_.chatScale = p_lambda$static$13_1_;
         Minecraft.getInstance().gui.getChat().rescaleChat();
-    }, (p_216637_0_, p_216637_1_) ->
+    }, (p_lambda$static$14_0_, p_lambda$static$14_1_) ->
     {
-        double d0 = p_216637_1_.toPct(p_216637_1_.get(p_216637_0_));
-        return (Component)(d0 == 0.0D ? CommonComponents.optionStatus(p_216637_1_.getCaption(), false) : p_216637_1_.percentValueLabel(d0));
+        double d0 = p_lambda$static$14_1_.toPct(p_lambda$static$14_1_.get(p_lambda$static$14_0_));
+        return (Component)(d0 == 0.0D ? CommonComponents.optionStatus(p_lambda$static$14_1_.getCaption(), false) : p_lambda$static$14_1_.percentValueLabel(d0));
     });
-    public static final ProgressOption CHAT_WIDTH = new ProgressOption("options.chat.width", 0.0D, 1.0D, 0.0F, (p_216601_0_) ->
+    public static final ProgressOption CHAT_WIDTH = new ProgressOption("options.chat.width", 0.0D, 1.0D, 0.0F, (p_lambda$static$15_0_) ->
     {
-        return p_216601_0_.chatWidth / 4.0571431D;
-    }, (p_216620_0_, p_216620_1_) ->
+        return p_lambda$static$15_0_.chatWidth / 4.0571431D;
+    }, (p_lambda$static$16_0_, p_lambda$static$16_1_) ->
     {
-        p_216620_1_ = p_216620_1_ * 4.0571431D;
-        p_216620_0_.chatWidth = p_216620_1_;
+        p_lambda$static$16_1_ = p_lambda$static$16_1_ * 4.0571431D;
+        p_lambda$static$16_0_.chatWidth = p_lambda$static$16_1_;
         Minecraft.getInstance().gui.getChat().rescaleChat();
-    }, (p_216673_0_, p_216673_1_) ->
+    }, (p_lambda$static$17_0_, p_lambda$static$17_1_) ->
     {
-        double d0 = p_216673_1_.toPct(p_216673_1_.get(p_216673_0_));
-        return p_216673_1_.pixelValueLabel(ChatComponent.getWidth(d0 * 4.0571431D));
+        double d0 = p_lambda$static$17_1_.toPct(p_lambda$static$17_1_.get(p_lambda$static$17_0_));
+        return p_lambda$static$17_1_.pixelValueLabel(ChatComponent.getWidth(d0 * 4.0571431D));
     });
-    public static final ProgressOption CHAT_LINE_SPACING = new ProgressOption("options.chat.line_spacing", 0.0D, 1.0D, 0.0F, (p_238287_0_) ->
+    public static final ProgressOption CHAT_LINE_SPACING = new ProgressOption("options.chat.line_spacing", 0.0D, 1.0D, 0.0F, (p_lambda$static$18_0_) ->
     {
-        return p_238287_0_.chatLineSpacing;
-    }, (p_238282_0_, p_238282_1_) ->
+        return p_lambda$static$18_0_.chatLineSpacing;
+    }, (p_lambda$static$19_0_, p_lambda$static$19_1_) ->
     {
-        p_238282_0_.chatLineSpacing = p_238282_1_;
-    }, (p_238297_0_, p_238297_1_) ->
+        p_lambda$static$19_0_.chatLineSpacing = p_lambda$static$19_1_;
+    }, (p_lambda$static$20_0_, p_lambda$static$20_1_) ->
     {
-        return p_238297_1_.percentValueLabel(p_238297_1_.toPct(p_238297_1_.get(p_238297_0_)));
+        return p_lambda$static$20_1_.percentValueLabel(p_lambda$static$20_1_.toPct(p_lambda$static$20_1_.get(p_lambda$static$20_0_)));
     });
-    public static final ProgressOption CHAT_DELAY = new ProgressOption("options.chat.delay_instant", 0.0D, 6.0D, 0.1F, (p_238271_0_) ->
+    public static final ProgressOption CHAT_DELAY = new ProgressOption("options.chat.delay_instant", 0.0D, 6.0D, 0.1F, (p_lambda$static$21_0_) ->
     {
-        return p_238271_0_.chatDelay;
-    }, (p_238273_0_, p_238273_1_) ->
+        return p_lambda$static$21_0_.chatDelay;
+    }, (p_lambda$static$22_0_, p_lambda$static$22_1_) ->
     {
-        p_238273_0_.chatDelay = p_238273_1_;
-    }, (p_238292_0_, p_238292_1_) ->
+        p_lambda$static$22_0_.chatDelay = p_lambda$static$22_1_;
+    }, (p_lambda$static$23_0_, p_lambda$static$23_1_) ->
     {
-        double d0 = p_238292_1_.get(p_238292_0_);
+        double d0 = p_lambda$static$23_1_.get(p_lambda$static$23_0_);
         return d0 <= 0.0D ? new TranslatableComponent("options.chat.delay_none") : new TranslatableComponent("options.chat.delay", String.format("%.1f", d0));
     });
-    public static final ProgressOption FOV = new ProgressOption("options.fov", 30.0D, 110.0D, 1.0F, (p_216655_0_) ->
+    public static final ProgressOption FOV = new ProgressOption("options.fov", 30.0D, 110.0D, 1.0F, (p_lambda$static$24_0_) ->
     {
-        return p_216655_0_.fov;
-    }, (p_216612_0_, p_216612_1_) ->
+        return p_lambda$static$24_0_.fov;
+    }, (p_lambda$static$25_0_, p_lambda$static$25_1_) ->
     {
-        p_216612_0_.fov = p_216612_1_;
-    }, (p_216590_0_, p_216590_1_) ->
+        p_lambda$static$25_0_.fov = p_lambda$static$25_1_;
+    }, (p_lambda$static$26_0_, p_lambda$static$26_1_) ->
     {
-        double d0 = p_216590_1_.get(p_216590_0_);
+        double d0 = p_lambda$static$26_1_.get(p_lambda$static$26_0_);
 
         if (d0 == 70.0D)
         {
-            return p_216590_1_.genericValueLabel(new TranslatableComponent("options.fov.min"));
+            return p_lambda$static$26_1_.genericValueLabel(new TranslatableComponent("options.fov.min"));
         }
         else {
-            return d0 == p_216590_1_.getMaxValue() ? p_216590_1_.genericValueLabel(new TranslatableComponent("options.fov.max")) : p_216590_1_.genericValueLabel((int)d0);
+            return d0 == p_lambda$static$26_1_.getMaxValue() ? p_lambda$static$26_1_.genericValueLabel(new TranslatableComponent("options.fov.max")) : p_lambda$static$26_1_.genericValueLabel((int)d0);
         }
     });
     private static final Component ACCESSIBILITY_TOOLTIP_FOV_EFFECT = new TranslatableComponent("options.fovEffectScale.tooltip");
-    public static final ProgressOption FOV_EFFECTS_SCALE = new ProgressOption("options.fovEffectScale", 0.0D, 1.0D, 0.0F, (p_216672_0_) ->
+    public static final ProgressOption FOV_EFFECTS_SCALE = new ProgressOption("options.fovEffectScale", 0.0D, 1.0D, 0.0F, (p_lambda$static$27_0_) ->
     {
-        return Math.pow((double)p_216672_0_.fovEffectScale, 2.0D);
-    }, (p_216608_0_, p_216608_1_) ->
+        return Math.pow((double)p_lambda$static$27_0_.fovEffectScale, 2.0D);
+    }, (p_lambda$static$28_0_, p_lambda$static$28_1_) ->
     {
-        p_216608_0_.fovEffectScale = Mth.sqrt(p_216608_1_);
-    }, (p_216645_0_, p_216645_1_) ->
+        p_lambda$static$28_0_.fovEffectScale = Mth.sqrt(p_lambda$static$28_1_);
+    }, (p_lambda$static$29_0_, p_lambda$static$29_1_) ->
     {
-        p_216645_1_.setTooltip(Minecraft.getInstance().font.split(ACCESSIBILITY_TOOLTIP_FOV_EFFECT, 200));
-        double d0 = p_216645_1_.toPct(p_216645_1_.get(p_216645_0_));
-        return d0 == 0.0D ? p_216645_1_.genericValueLabel(new TranslatableComponent("options.fovEffectScale.off")) : p_216645_1_.percentValueLabel(d0);
+        p_lambda$static$29_1_.setTooltip(Minecraft.getInstance().font.split(ACCESSIBILITY_TOOLTIP_FOV_EFFECT, 200));
+        double d0 = p_lambda$static$29_1_.toPct(p_lambda$static$29_1_.get(p_lambda$static$29_0_));
+        return d0 == 0.0D ? p_lambda$static$29_1_.genericValueLabel(new TranslatableComponent("options.fovEffectScale.off")) : p_lambda$static$29_1_.percentValueLabel(d0);
     });
     private static final Component ACCESSIBILITY_TOOLTIP_SCREEN_EFFECT = new TranslatableComponent("options.screenEffectScale.tooltip");
-    public static final ProgressOption SCREEN_EFFECTS_SCALE = new ProgressOption("options.screenEffectScale", 0.0D, 1.0D, 0.0F, (p_244419_0_) ->
+    public static final ProgressOption SCREEN_EFFECTS_SCALE = new ProgressOption("options.screenEffectScale", 0.0D, 1.0D, 0.0F, (p_lambda$static$30_0_) ->
     {
-        return (double)p_244419_0_.screenEffectScale;
-    }, (p_244415_0_, p_244415_1_) ->
+        return (double)p_lambda$static$30_0_.screenEffectScale;
+    }, (p_lambda$static$31_0_, p_lambda$static$31_1_) ->
     {
-        p_244415_0_.screenEffectScale = p_244415_1_.floatValue();
-    }, (p_244416_0_, p_244416_1_) ->
+        p_lambda$static$31_0_.screenEffectScale = p_lambda$static$31_1_.floatValue();
+    }, (p_lambda$static$32_0_, p_lambda$static$32_1_) ->
     {
-        p_244416_1_.setTooltip(Minecraft.getInstance().font.split(ACCESSIBILITY_TOOLTIP_SCREEN_EFFECT, 200));
-        double d0 = p_244416_1_.toPct(p_244416_1_.get(p_244416_0_));
-        return d0 == 0.0D ? p_244416_1_.genericValueLabel(new TranslatableComponent("options.screenEffectScale.off")) : p_244416_1_.percentValueLabel(d0);
+        p_lambda$static$32_1_.setTooltip(Minecraft.getInstance().font.split(ACCESSIBILITY_TOOLTIP_SCREEN_EFFECT, 200));
+        double d0 = p_lambda$static$32_1_.toPct(p_lambda$static$32_1_.get(p_lambda$static$32_0_));
+        return d0 == 0.0D ? p_lambda$static$32_1_.genericValueLabel(new TranslatableComponent("options.screenEffectScale.off")) : p_lambda$static$32_1_.percentValueLabel(d0);
     });
-    public static final ProgressOption FRAMERATE_LIMIT = new ProgressOption("options.framerateLimit", 0.0D, 260.0D, 5.0F, (p_244406_0_) ->
+    public static final ProgressOption FRAMERATE_LIMIT = new ProgressOption("options.framerateLimit", 0.0D, 260.0D, 5.0F, (p_lambda$static$33_0_) ->
     {
-        return p_244406_0_.enableVsync ? FRAMERATE_LIMIT.getMinValue() : (double)p_244406_0_.framerateLimit;
-    }, (p_244411_0_, p_244411_1_) ->
+        return p_lambda$static$33_0_.enableVsync ? 0.0D : (double)p_lambda$static$33_0_.framerateLimit;
+    }, (p_lambda$static$34_0_, p_lambda$static$34_1_) ->
     {
-        p_244411_0_.framerateLimit = (int)p_244411_1_.doubleValue();
-        p_244411_0_.enableVsync = false;
+        p_lambda$static$34_0_.framerateLimit = (int)p_lambda$static$34_1_.doubleValue();
+        p_lambda$static$34_0_.enableVsync = false;
 
-        if (p_244411_0_.framerateLimit <= 0)
+        if (p_lambda$static$34_0_.framerateLimit <= 0)
         {
-            p_244411_0_.framerateLimit = (int)FRAMERATE_LIMIT.getMaxValue();
-            p_244411_0_.enableVsync = true;
+            p_lambda$static$34_0_.framerateLimit = 260;
+            p_lambda$static$34_0_.enableVsync = true;
         }
 
-        p_244411_0_.updateVSync();
-        Minecraft.getInstance().getWindow().setFramerateLimit(p_244411_0_.framerateLimit);
-    }, (p_244413_0_, p_244413_1_) ->
+        p_lambda$static$34_0_.updateVSync();
+        Minecraft.getInstance().getWindow().setFramerateLimit(p_lambda$static$34_0_.framerateLimit);
+    }, (p_lambda$static$35_0_, p_lambda$static$35_1_) ->
     {
-        if (p_244413_0_.enableVsync)
+        if (p_lambda$static$35_0_.enableVsync)
         {
-            return p_244413_1_.genericValueLabel(new TranslatableComponent("of.options.framerateLimit.vsync"));
+            return p_lambda$static$35_1_.genericValueLabel(new TranslatableComponent("of.options.framerateLimit.vsync"));
         }
         else {
-            double d0 = p_244413_1_.get(p_244413_0_);
-            return d0 == p_244413_1_.getMaxValue() ? p_244413_1_.genericValueLabel(new TranslatableComponent("options.framerateLimit.max")) : p_244413_1_.genericValueLabel(new TranslatableComponent("options.framerate", (int)d0));
+            double d0 = p_lambda$static$35_1_.get(p_lambda$static$35_0_);
+            return d0 == p_lambda$static$35_1_.getMaxValue() ? p_lambda$static$35_1_.genericValueLabel(new TranslatableComponent("options.framerateLimit.max")) : p_lambda$static$35_1_.genericValueLabel(new TranslatableComponent("options.framerate", (int)d0));
         }
     });
-    public static final ProgressOption GAMMA = new ProgressOption("options.gamma", 0.0D, 1.0D, 0.0F, (p_216636_0_) ->
+    public static final ProgressOption GAMMA = new ProgressOption("options.gamma", -0.5D, 1.0D, 0.0F, (p_lambda$static$36_0_) ->
     {
-        return p_216636_0_.gamma;
-    }, (p_216651_0_, p_216651_1_) ->
+        return p_lambda$static$36_0_.gamma;
+    }, (p_lambda$static$37_0_, p_lambda$static$37_1_) ->
     {
-        p_216651_0_.gamma = p_216651_1_;
-    }, (p_216594_0_, p_216594_1_) ->
+        p_lambda$static$37_0_.gamma = p_lambda$static$37_1_;
+    }, (p_lambda$static$38_0_, p_lambda$static$38_1_) ->
     {
-        double d0 = p_216594_1_.toPct(p_216594_1_.get(p_216594_0_));
+        double d0 = p_lambda$static$38_1_.toPct(p_lambda$static$38_1_.get(p_lambda$static$38_0_));
 
-        if (d0 == 0.0D)
+        if (Math.abs(d0 - 0.3333333333333333D) < 0.005D)
+        {
+            return p_lambda$static$38_1_.genericValueLabel(new TranslatableComponent("options.gamma.min"));
+        }
+        else if (d0 == 1.0D)
         {
-            return p_216594_1_.genericValueLabel(new TranslatableComponent("options.gamma.min"));
+            return p_lambda$static$38_1_.genericValueLabel(new TranslatableComponent("options.gamma.max"));
         }
         else {
-            return d0 == 1.0D ? p_216594_1_.genericValueLabel(new TranslatableComponent("options.gamma.max")) : p_216594_1_.percentAddValueLabel((int)(d0 * 100.0D));
+            return d0 == 0.0D ? p_lambda$static$38_1_.genericValueLabel(new TranslatableComponent("vivecraft.options.gamma.cantseeshitcaptain")) : p_lambda$static$38_1_.genericValueLabel(new TextComponent((d0 >= 0.3333333333333333D ? "+" : "") + (int)((d0 * 1.5D - 0.5D) * 100.0D) + "%"));
         }
     });
-    public static final ProgressOption MIPMAP_LEVELS = new ProgressOption("options.mipmapLevels", 0.0D, 4.0D, 1.0F, (p_216667_0_) ->
+    public static final ProgressOption MIPMAP_LEVELS = new ProgressOption("options.mipmapLevels", 0.0D, 4.0D, 1.0F, (p_lambda$static$39_0_) ->
     {
-        return (double)p_216667_0_.mipmapLevels;
-    }, (p_216585_0_, p_216585_1_) ->
+        return (double)p_lambda$static$39_0_.mipmapLevels;
+    }, (p_lambda$static$40_0_, p_lambda$static$40_1_) ->
     {
-        p_216585_0_.mipmapLevels = (int)p_216585_1_.doubleValue();
-        p_216585_0_.updateMipmaps();
-    }, (p_216629_0_, p_216629_1_) ->
+        p_lambda$static$40_0_.mipmapLevels = (int)p_lambda$static$40_1_.doubleValue();
+        p_lambda$static$40_0_.updateMipmaps();
+    }, (p_lambda$static$41_0_, p_lambda$static$41_1_) ->
     {
-        double d0 = p_216629_1_.get(p_216629_0_);
+        double d0 = p_lambda$static$41_1_.get(p_lambda$static$41_0_);
 
         if (d0 >= 4.0D)
         {
-            return p_216629_1_.genericValueLabel(new TranslatableComponent("of.general.max"));
+            return p_lambda$static$41_1_.genericValueLabel(new TranslatableComponent("of.general.max"));
         }
         else {
-            return (Component)(d0 == 0.0D ? CommonComponents.optionStatus(p_216629_1_.getCaption(), false) : p_216629_1_.genericValueLabel((int)d0));
+            return (Component)(d0 == 0.0D ? CommonComponents.optionStatus(p_lambda$static$41_1_.getCaption(), false) : p_lambda$static$41_1_.genericValueLabel((int)d0));
         }
     });
-    public static final ProgressOption MOUSE_WHEEL_SENSITIVITY = new LogaritmicProgressOption("options.mouseWheelSensitivity", 0.01D, 10.0D, 0.01F, (p_216581_0_) ->
+    public static final ProgressOption MOUSE_WHEEL_SENSITIVITY = new LogaritmicProgressOption("options.mouseWheelSensitivity", 0.01D, 10.0D, 0.01F, (p_lambda$static$42_0_) ->
     {
-        return p_216581_0_.mouseWheelSensitivity;
-    }, (p_216628_0_, p_216628_1_) ->
+        return p_lambda$static$42_0_.mouseWheelSensitivity;
+    }, (p_lambda$static$43_0_, p_lambda$static$43_1_) ->
     {
-        p_216628_0_.mouseWheelSensitivity = p_216628_1_;
-    }, (p_216675_0_, p_216675_1_) ->
+        p_lambda$static$43_0_.mouseWheelSensitivity = p_lambda$static$43_1_;
+    }, (p_lambda$static$44_0_, p_lambda$static$44_1_) ->
     {
-        double d0 = p_216675_1_.toPct(p_216675_1_.get(p_216675_0_));
-        return p_216675_1_.genericValueLabel(new TextComponent(String.format("%.2f", p_216675_1_.toValue(d0))));
+        double d0 = p_lambda$static$44_1_.toPct(p_lambda$static$44_1_.get(p_lambda$static$44_0_));
+        return p_lambda$static$44_1_.genericValueLabel(new TextComponent(String.format("%.2f", p_lambda$static$44_1_.toValue(d0))));
     });
-    public static final BooleanOption RAW_MOUSE_INPUT = new BooleanOption("options.rawMouseInput", (p_225287_0_) ->
+    public static final BooleanOption RAW_MOUSE_INPUT = new BooleanOption("options.rawMouseInput", (p_lambda$static$45_0_) ->
     {
-        return p_225287_0_.rawMouseInput;
-    }, (p_225259_0_, p_225259_1_) ->
+        return p_lambda$static$45_0_.rawMouseInput;
+    }, (p_lambda$static$46_0_, p_lambda$static$46_1_) ->
     {
-        p_225259_0_.rawMouseInput = p_225259_1_;
+        p_lambda$static$46_0_.rawMouseInput = p_lambda$static$46_1_;
         Window window = Minecraft.getInstance().getWindow();
 
         if (window != null)
         {
-            window.updateRawMouseInput(p_225259_1_);
+            window.updateRawMouseInput(p_lambda$static$46_1_);
         }
     });
-    public static final ProgressOption RENDER_DISTANCE = new ProgressOption("options.renderDistance", 2.0D, 16.0D, 1.0F, (p_216658_0_) ->
+    public static final ProgressOption RENDER_DISTANCE = new ProgressOption("options.renderDistance", 2.0D, 16.0D, 1.0F, (p_lambda$static$47_0_) ->
     {
-        return (double)p_216658_0_.renderDistance;
-    }, (p_216579_0_, p_216579_1_) ->
+        return (double)p_lambda$static$47_0_.renderDistance;
+    }, (p_lambda$static$48_0_, p_lambda$static$48_1_) ->
     {
-        p_216579_0_.renderDistance = (int)p_216579_1_.doubleValue();
+        p_lambda$static$48_0_.renderDistance = (int)p_lambda$static$48_1_.doubleValue();
         Minecraft.getInstance().levelRenderer.needsUpdate();
-    }, (p_216664_0_, p_216664_1_) ->
+    }, (p_lambda$static$49_0_, p_lambda$static$49_1_) ->
     {
-        double d0 = p_216664_1_.get(p_216664_0_);
-        return p_216664_1_.genericValueLabel(new TranslatableComponent("options.chunks", (int)d0));
+        double d0 = p_lambda$static$49_1_.get(p_lambda$static$49_0_);
+        return p_lambda$static$49_1_.genericValueLabel(new TranslatableComponent("options.chunks", (int)d0));
     });
-    public static final ProgressOption ENTITY_DISTANCE_SCALING = new ProgressOption("options.entityDistanceScaling", 0.5D, 5.0D, 0.25F, (p_238324_0_) ->
+    public static final ProgressOption ENTITY_DISTANCE_SCALING = new ProgressOption("options.entityDistanceScaling", 0.5D, 5.0D, 0.25F, (p_lambda$static$50_0_) ->
     {
-        return (double)p_238324_0_.entityDistanceScaling;
-    }, (p_238256_0_, p_238256_1_) ->
+        return (double)p_lambda$static$50_0_.entityDistanceScaling;
+    }, (p_lambda$static$51_0_, p_lambda$static$51_1_) ->
     {
-        p_238256_0_.entityDistanceScaling = (float)p_238256_1_.doubleValue();
-    }, (p_238254_0_, p_238254_1_) ->
+        p_lambda$static$51_0_.entityDistanceScaling = (float)p_lambda$static$51_1_.doubleValue();
+    }, (p_lambda$static$52_0_, p_lambda$static$52_1_) ->
     {
-        double d0 = p_238254_1_.get(p_238254_0_);
-        return p_238254_1_.percentValueLabel(d0);
+        double d0 = p_lambda$static$52_1_.get(p_lambda$static$52_0_);
+        return p_lambda$static$52_1_.percentValueLabel(d0);
     });
-    public static final ProgressOption SENSITIVITY = new ProgressOption("options.sensitivity", 0.0D, 1.0D, 0.0F, (p_216654_0_) ->
+    public static final ProgressOption SENSITIVITY = new ProgressOption("options.sensitivity", 0.0D, 1.0D, 0.0F, (p_lambda$static$53_0_) ->
     {
-        return p_216654_0_.sensitivity;
-    }, (p_216644_0_, p_216644_1_) ->
+        return p_lambda$static$53_0_.sensitivity;
+    }, (p_lambda$static$54_0_, p_lambda$static$54_1_) ->
     {
-        p_216644_0_.sensitivity = p_216644_1_;
-    }, (p_216641_0_, p_216641_1_) ->
+        p_lambda$static$54_0_.sensitivity = p_lambda$static$54_1_;
+    }, (p_lambda$static$55_0_, p_lambda$static$55_1_) ->
     {
-        double d0 = p_216641_1_.toPct(p_216641_1_.get(p_216641_0_));
+        double d0 = p_lambda$static$55_1_.toPct(p_lambda$static$55_1_.get(p_lambda$static$55_0_));
 
         if (d0 == 0.0D)
         {
-            return p_216641_1_.genericValueLabel(new TranslatableComponent("options.sensitivity.min"));
+            return p_lambda$static$55_1_.genericValueLabel(new TranslatableComponent("options.sensitivity.min"));
         }
         else {
-            return d0 == 1.0D ? p_216641_1_.genericValueLabel(new TranslatableComponent("options.sensitivity.max")) : p_216641_1_.percentValueLabel(2.0D * d0);
+            return d0 == 1.0D ? p_lambda$static$55_1_.genericValueLabel(new TranslatableComponent("options.sensitivity.max")) : p_lambda$static$55_1_.percentValueLabel(2.0D * d0);
         }
     });
-    public static final ProgressOption TEXT_BACKGROUND_OPACITY = new ProgressOption("options.accessibility.text_background_opacity", 0.0D, 1.0D, 0.0F, (p_216597_0_) ->
+    public static final ProgressOption TEXT_BACKGROUND_OPACITY = new ProgressOption("options.accessibility.text_background_opacity", 0.0D, 1.0D, 0.0F, (p_lambda$static$56_0_) ->
     {
-        return p_216597_0_.textBackgroundOpacity;
-    }, (p_216593_0_, p_216593_1_) ->
+        return p_lambda$static$56_0_.textBackgroundOpacity;
+    }, (p_lambda$static$57_0_, p_lambda$static$57_1_) ->
     {
-        p_216593_0_.textBackgroundOpacity = p_216593_1_;
+        p_lambda$static$57_0_.textBackgroundOpacity = p_lambda$static$57_1_;
         Minecraft.getInstance().gui.getChat().rescaleChat();
-    }, (p_216626_0_, p_216626_1_) ->
+    }, (p_lambda$static$58_0_, p_lambda$static$58_1_) ->
     {
-        return p_216626_1_.percentValueLabel(p_216626_1_.toPct(p_216626_1_.get(p_216626_0_)));
+        return p_lambda$static$58_1_.percentValueLabel(p_lambda$static$58_1_.toPct(p_lambda$static$58_1_.get(p_lambda$static$58_0_)));
     });
-    public static final CycleOption AMBIENT_OCCLUSION = new CycleOption("options.ao", (p_216653_0_, p_216653_1_) ->
+    public static final CycleOption AMBIENT_OCCLUSION = new CycleOption("options.ao", (p_lambda$static$59_0_, p_lambda$static$59_1_) ->
     {
-        p_216653_0_.ambientOcclusion = AmbientOcclusionStatus.byId(p_216653_0_.ambientOcclusion.getId() + p_216653_1_);
+        p_lambda$static$59_0_.ambientOcclusion = AmbientOcclusionStatus.byId(p_lambda$static$59_0_.ambientOcclusion.getId() + p_lambda$static$59_1_);
         Minecraft.getInstance().levelRenderer.allChanged();
-    }, (p_216630_0_, p_216630_1_) ->
+    }, (p_lambda$static$60_0_, p_lambda$static$60_1_) ->
     {
-        return p_216630_1_.genericValueLabel(new TranslatableComponent(p_216630_0_.ambientOcclusion.getKey()));
+        return p_lambda$static$60_1_.genericValueLabel(new TranslatableComponent(p_lambda$static$60_0_.ambientOcclusion.getKey()));
     });
-    public static final CycleOption ATTACK_INDICATOR = new CycleOption("options.attackIndicator", (p_216615_0_, p_216615_1_) ->
+    public static final CycleOption ATTACK_INDICATOR = new CycleOption("options.attackIndicator", (p_lambda$static$61_0_, p_lambda$static$61_1_) ->
     {
-        p_216615_0_.attackIndicator = AttackIndicatorStatus.byId(p_216615_0_.attackIndicator.getId() + p_216615_1_);
-    }, (p_216609_0_, p_216609_1_) ->
+        p_lambda$static$61_0_.attackIndicator = AttackIndicatorStatus.byId(p_lambda$static$61_0_.attackIndicator.getId() + p_lambda$static$61_1_);
+    }, (p_lambda$static$62_0_, p_lambda$static$62_1_) ->
     {
-        return p_216609_1_.genericValueLabel(new TranslatableComponent(p_216609_0_.attackIndicator.getKey()));
+        return p_lambda$static$62_1_.genericValueLabel(new TranslatableComponent(p_lambda$static$62_0_.attackIndicator.getKey()));
     });
-    public static final CycleOption CHAT_VISIBILITY = new CycleOption("options.chat.visibility", (p_216640_0_, p_216640_1_) ->
+    public static final CycleOption CHAT_VISIBILITY = new CycleOption("options.chat.visibility", (p_lambda$static$63_0_, p_lambda$static$63_1_) ->
     {
-        p_216640_0_.chatVisibility = ChatVisiblity.byId((p_216640_0_.chatVisibility.getId() + p_216640_1_) % 3);
-    }, (p_216598_0_, p_216598_1_) ->
+        p_lambda$static$63_0_.chatVisibility = ChatVisiblity.byId((p_lambda$static$63_0_.chatVisibility.getId() + p_lambda$static$63_1_) % 3);
+    }, (p_lambda$static$64_0_, p_lambda$static$64_1_) ->
     {
-        return p_216598_1_.genericValueLabel(new TranslatableComponent(p_216598_0_.chatVisibility.getKey()));
+        return p_lambda$static$64_1_.genericValueLabel(new TranslatableComponent(p_lambda$static$64_0_.chatVisibility.getKey()));
     });
     private static final Component GRAPHICS_TOOLTIP_FAST = new TranslatableComponent("options.graphics.fast.tooltip");
     private static final Component GRAPHICS_TOOLTIP_FABULOUS = new TranslatableComponent("options.graphics.fabulous.tooltip", (new TranslatableComponent("options.graphics.fabulous")).withStyle(ChatFormatting.ITALIC));
     private static final Component GRAPHICS_TOOLTIP_FANCY = new TranslatableComponent("options.graphics.fancy.tooltip");
-    public static final CycleOption GRAPHICS = new CycleOption("options.graphics", (p_216577_0_, p_216577_1_) ->
+    public static final CycleOption GRAPHICS = new CycleOption("options.graphics", (p_lambda$static$65_0_, p_lambda$static$65_1_) ->
     {
         Minecraft minecraft = Minecraft.getInstance();
         GpuWarnlistManager gpuwarnlistmanager = minecraft.getGpuWarnlistManager();
 
-        if (p_216577_0_.graphicsMode == GraphicsStatus.FANCY && gpuwarnlistmanager.willShowWarning())
+        if (p_lambda$static$65_0_.graphicsMode == GraphicsStatus.FANCY && gpuwarnlistmanager.willShowWarning())
         {
             gpuwarnlistmanager.showWarning();
         }
         else {
-            p_216577_0_.graphicsMode = p_216577_0_.graphicsMode.cycleNext();
+            p_lambda$static$65_0_.graphicsMode = p_lambda$static$65_0_.graphicsMode.cycleNext();
 
-            if (p_216577_0_.graphicsMode == GraphicsStatus.FABULOUS && (Config.isShaders() || !GLX.isUsingFBOs() || !GlStateManager.supportsFramebufferBlit() || gpuwarnlistmanager.isSkippingFabulous()))
+            if (p_lambda$static$65_0_.graphicsMode == GraphicsStatus.FABULOUS && (Config.isShaders() || !GLX.isUsingFBOs() || !GlStateManager.supportsFramebufferBlit() || gpuwarnlistmanager.isSkippingFabulous()))
             {
-                p_216577_0_.graphicsMode = GraphicsStatus.FAST;
+                p_lambda$static$65_0_.graphicsMode = GraphicsStatus.FAST;
             }
 
-            p_216577_0_.updateRenderClouds();
+            p_lambda$static$65_0_.updateRenderClouds();
             minecraft.levelRenderer.allChanged();
+            minecraft.vrRenderer.reinitFrameBuffers("gfx setting change");
         }
-    }, (p_216633_0_, p_216633_1_) ->
+    }, (p_lambda$static$66_0_, p_lambda$static$66_1_) ->
     {
-        switch (p_216633_0_.graphicsMode)
+        switch (p_lambda$static$66_0_.graphicsMode)
         {
             case FAST:
-                p_216633_1_.setTooltip(Minecraft.getInstance().font.split(GRAPHICS_TOOLTIP_FAST, 200));
+                p_lambda$static$66_1_.setTooltip(Minecraft.getInstance().font.split(GRAPHICS_TOOLTIP_FAST, 200));
                 break;
 
             case FANCY:
-                p_216633_1_.setTooltip(Minecraft.getInstance().font.split(GRAPHICS_TOOLTIP_FANCY, 200));
+                p_lambda$static$66_1_.setTooltip(Minecraft.getInstance().font.split(GRAPHICS_TOOLTIP_FANCY, 200));
                 break;
 
             case FABULOUS:
-                p_216633_1_.setTooltip(Minecraft.getInstance().font.split(GRAPHICS_TOOLTIP_FABULOUS, 200));
+                p_lambda$static$66_1_.setTooltip(Minecraft.getInstance().font.split(GRAPHICS_TOOLTIP_FABULOUS, 200));
         }
 
-        MutableComponent mutablecomponent = new TranslatableComponent(p_216633_0_.graphicsMode.getKey());
-        return p_216633_0_.graphicsMode == GraphicsStatus.FABULOUS ? p_216633_1_.genericValueLabel(mutablecomponent.withStyle(ChatFormatting.ITALIC)) : p_216633_1_.genericValueLabel(mutablecomponent);
+        MutableComponent mutablecomponent = new TranslatableComponent(p_lambda$static$66_0_.graphicsMode.getKey());
+        return p_lambda$static$66_0_.graphicsMode == GraphicsStatus.FABULOUS ? p_lambda$static$66_1_.genericValueLabel(mutablecomponent.withStyle(ChatFormatting.ITALIC)) : p_lambda$static$66_1_.genericValueLabel(mutablecomponent);
     });
-    public static final CycleOption GUI_SCALE = new CycleOption("options.guiScale", (p_216674_0_, p_216674_1_) ->
+    public static final CycleOption GUI_SCALE = new CycleOption("options.guiScale", (p_lambda$static$67_0_, p_lambda$static$67_1_) ->
     {
-        p_216674_0_.guiScale = Mth.positiveModulo(p_216674_0_.guiScale + p_216674_1_, Minecraft.getInstance().getWindow().calculateScale(0, Minecraft.getInstance().isEnforceUnicode()) + 1);
-    }, (p_216668_0_, p_216668_1_) ->
+        p_lambda$static$67_0_.guiScaleUser = Mth.positiveModulo(p_lambda$static$67_0_.guiScaleUser + p_lambda$static$67_1_, Minecraft.getInstance().getWindow().calculateScale(0, Minecraft.getInstance().isEnforceUnicode()) + 1);
+    }, (p_lambda$static$68_0_, p_lambda$static$68_1_) ->
     {
-        return p_216668_0_.guiScale == 0 ? p_216668_1_.genericValueLabel(new TranslatableComponent("options.guiScale.auto")) : p_216668_1_.genericValueLabel(p_216668_0_.guiScale);
+        return p_lambda$static$68_0_.guiScaleUser == 0 ? p_lambda$static$68_1_.genericValueLabel(new TranslatableComponent("options.guiScale.auto")) : p_lambda$static$68_1_.genericValueLabel(p_lambda$static$68_0_.guiScaleUser);
     });
-    public static final CycleOption MAIN_HAND = new CycleOption("options.mainHand", (p_216584_0_, p_216584_1_) ->
+    public static final CycleOption MAIN_HAND = new CycleOption("options.mainHand", (p_lambda$static$69_0_, p_lambda$static$69_1_) ->
     {
-        p_216584_0_.mainHand = p_216584_0_.mainHand.getOpposite();
-    }, (p_216596_0_, p_216596_1_) ->
+        p_lambda$static$69_0_.mainHand = p_lambda$static$69_0_.mainHand.getOpposite();
+    }, (p_lambda$static$70_0_, p_lambda$static$70_1_) ->
     {
-        return p_216596_1_.genericValueLabel(p_216596_0_.mainHand.getName());
+        return p_lambda$static$70_1_.genericValueLabel(p_lambda$static$70_0_.mainHand.getName());
     });
-    public static final CycleOption NARRATOR = new CycleOption("options.narrator", (p_216648_0_, p_216648_1_) ->
+    public static final CycleOption NARRATOR = new CycleOption("options.narrator", (p_lambda$static$71_0_, p_lambda$static$71_1_) ->
     {
         if (NarratorChatListener.INSTANCE.isActive())
         {
-            p_216648_0_.narratorStatus = NarratorStatus.byId(p_216648_0_.narratorStatus.getId() + p_216648_1_);
+            p_lambda$static$71_0_.narratorStatus = NarratorStatus.byId(p_lambda$static$71_0_.narratorStatus.getId() + p_lambda$static$71_1_);
         }
         else {
-            p_216648_0_.narratorStatus = NarratorStatus.OFF;
+            p_lambda$static$71_0_.narratorStatus = NarratorStatus.OFF;
         }
 
-        NarratorChatListener.INSTANCE.updateNarratorStatus(p_216648_0_.narratorStatus);
-    }, (p_216632_0_, p_216632_1_) ->
+        NarratorChatListener.INSTANCE.updateNarratorStatus(p_lambda$static$71_0_.narratorStatus);
+    }, (p_lambda$static$72_0_, p_lambda$static$72_1_) ->
     {
-        return NarratorChatListener.INSTANCE.isActive() ? p_216632_1_.genericValueLabel(p_216632_0_.narratorStatus.getName()) : p_216632_1_.genericValueLabel(new TranslatableComponent("options.narrator.notavailable"));
+        return NarratorChatListener.INSTANCE.isActive() ? p_lambda$static$72_1_.genericValueLabel(p_lambda$static$72_0_.narratorStatus.getName()) : p_lambda$static$72_1_.genericValueLabel(new TranslatableComponent("options.narrator.notavailable"));
     });
-    public static final CycleOption PARTICLES = new CycleOption("options.particles", (p_216622_0_, p_216622_1_) ->
+    public static final CycleOption PARTICLES = new CycleOption("options.particles", (p_lambda$static$73_0_, p_lambda$static$73_1_) ->
     {
-        p_216622_0_.particles = ParticleStatus.byId(p_216622_0_.particles.getId() + p_216622_1_);
-    }, (p_216616_0_, p_216616_1_) ->
+        p_lambda$static$73_0_.particles = ParticleStatus.byId(p_lambda$static$73_0_.particles.getId() + p_lambda$static$73_1_);
+    }, (p_lambda$static$74_0_, p_lambda$static$74_1_) ->
     {
-        return p_216616_1_.genericValueLabel(new TranslatableComponent(p_216616_0_.particles.getKey()));
+        return p_lambda$static$74_1_.genericValueLabel(new TranslatableComponent(p_lambda$static$74_0_.particles.getKey()));
     });
-    public static final CycleOption RENDER_CLOUDS = new CycleOption("options.renderClouds", (p_216605_0_, p_216605_1_) ->
+    public static final CycleOption RENDER_CLOUDS = new CycleOption("options.renderClouds", (p_lambda$static$75_0_, p_lambda$static$75_1_) ->
     {
-        p_216605_0_.renderClouds = CloudStatus.byId(p_216605_0_.renderClouds.getId() + p_216605_1_);
+        p_lambda$static$75_0_.renderClouds = CloudStatus.byId(p_lambda$static$75_0_.renderClouds.getId() + p_lambda$static$75_1_);
 
         if (Minecraft.useShaderTransparency())
         {
@@ -417,179 +422,179 @@
                 rendertarget.clear(Minecraft.ON_OSX);
             }
         }
-    }, (p_216602_0_, p_216602_1_) ->
+    }, (p_lambda$static$76_0_, p_lambda$static$76_1_) ->
     {
-        return p_216602_1_.genericValueLabel(new TranslatableComponent(p_216602_0_.renderClouds.getKey()));
+        return p_lambda$static$76_1_.genericValueLabel(new TranslatableComponent(p_lambda$static$76_0_.renderClouds.getKey()));
     });
-    public static final CycleOption TEXT_BACKGROUND = new CycleOption("options.accessibility.text_background", (p_216665_0_, p_216665_1_) ->
+    public static final CycleOption TEXT_BACKGROUND = new CycleOption("options.accessibility.text_background", (p_lambda$static$77_0_, p_lambda$static$77_1_) ->
     {
-        p_216665_0_.backgroundForChatOnly = !p_216665_0_.backgroundForChatOnly;
-    }, (p_216639_0_, p_216639_1_) ->
+        p_lambda$static$77_0_.backgroundForChatOnly = !p_lambda$static$77_0_.backgroundForChatOnly;
+    }, (p_lambda$static$78_0_, p_lambda$static$78_1_) ->
     {
-        return p_216639_1_.genericValueLabel(new TranslatableComponent(p_216639_0_.backgroundForChatOnly ? "options.accessibility.text_background.chat" : "options.accessibility.text_background.everywhere"));
+        return p_lambda$static$78_1_.genericValueLabel(new TranslatableComponent(p_lambda$static$78_0_.backgroundForChatOnly ? "options.accessibility.text_background.chat" : "options.accessibility.text_background.everywhere"));
     });
     private static final Component CHAT_TOOLTIP_HIDE_MATCHED_NAMES = new TranslatableComponent("options.hideMatchedNames.tooltip");
-    public static final BooleanOption AUTO_JUMP = new BooleanOption("options.autoJump", (p_216619_0_) ->
+    public static final BooleanOption AUTO_JUMP = new BooleanOption("options.autoJump", (p_lambda$static$79_0_) ->
     {
-        return p_216619_0_.autoJump;
-    }, (p_216621_0_, p_216621_1_) ->
+        return p_lambda$static$79_0_.autoJump;
+    }, (p_lambda$static$80_0_, p_lambda$static$80_1_) ->
     {
-        p_216621_0_.autoJump = p_216621_1_;
+        p_lambda$static$80_0_.autoJump = p_lambda$static$80_1_;
     });
-    public static final BooleanOption AUTO_SUGGESTIONS = new BooleanOption("options.autoSuggestCommands", (p_216643_0_) ->
+    public static final BooleanOption AUTO_SUGGESTIONS = new BooleanOption("options.autoSuggestCommands", (p_lambda$static$81_0_) ->
     {
-        return p_216643_0_.autoSuggestions;
-    }, (p_216656_0_, p_216656_1_) ->
+        return p_lambda$static$81_0_.autoSuggestions;
+    }, (p_lambda$static$82_0_, p_lambda$static$82_1_) ->
     {
-        p_216656_0_.autoSuggestions = p_216656_1_;
+        p_lambda$static$82_0_.autoSuggestions = p_lambda$static$82_1_;
     });
-    public static final BooleanOption HIDE_MATCHED_NAMES = new BooleanOption("options.hideMatchedNames", CHAT_TOOLTIP_HIDE_MATCHED_NAMES, (p_244790_0_) ->
+    public static final BooleanOption HIDE_MATCHED_NAMES = new BooleanOption("options.hideMatchedNames", CHAT_TOOLTIP_HIDE_MATCHED_NAMES, (p_lambda$static$83_0_) ->
     {
-        return p_244790_0_.hideMatchedNames;
-    }, (p_244791_0_, p_244791_1_) ->
+        return p_lambda$static$83_0_.hideMatchedNames;
+    }, (p_lambda$static$84_0_, p_lambda$static$84_1_) ->
     {
-        p_244791_0_.hideMatchedNames = p_244791_1_;
+        p_lambda$static$84_0_.hideMatchedNames = p_lambda$static$84_1_;
     });
-    public static final BooleanOption CHAT_COLOR = new BooleanOption("options.chat.color", (p_216669_0_) ->
+    public static final BooleanOption CHAT_COLOR = new BooleanOption("options.chat.color", (p_lambda$static$85_0_) ->
     {
-        return p_216669_0_.chatColors;
-    }, (p_216659_0_, p_216659_1_) ->
+        return p_lambda$static$85_0_.chatColors;
+    }, (p_lambda$static$86_0_, p_lambda$static$86_1_) ->
     {
-        p_216659_0_.chatColors = p_216659_1_;
+        p_lambda$static$86_0_.chatColors = p_lambda$static$86_1_;
     });
-    public static final BooleanOption CHAT_LINKS = new BooleanOption("options.chat.links", (p_216583_0_) ->
+    public static final BooleanOption CHAT_LINKS = new BooleanOption("options.chat.links", (p_lambda$static$87_0_) ->
     {
-        return p_216583_0_.chatLinks;
-    }, (p_216670_0_, p_216670_1_) ->
+        return p_lambda$static$87_0_.chatLinks;
+    }, (p_lambda$static$88_0_, p_lambda$static$88_1_) ->
     {
-        p_216670_0_.chatLinks = p_216670_1_;
+        p_lambda$static$88_0_.chatLinks = p_lambda$static$88_1_;
     });
-    public static final BooleanOption CHAT_LINKS_PROMPT = new BooleanOption("options.chat.links.prompt", (p_216610_0_) ->
+    public static final BooleanOption CHAT_LINKS_PROMPT = new BooleanOption("options.chat.links.prompt", (p_lambda$static$89_0_) ->
     {
-        return p_216610_0_.chatLinksPrompt;
-    }, (p_216652_0_, p_216652_1_) ->
+        return p_lambda$static$89_0_.chatLinksPrompt;
+    }, (p_lambda$static$90_0_, p_lambda$static$90_1_) ->
     {
-        p_216652_0_.chatLinksPrompt = p_216652_1_;
+        p_lambda$static$90_0_.chatLinksPrompt = p_lambda$static$90_1_;
     });
-    public static final BooleanOption DISCRETE_MOUSE_SCROLL = new BooleanOption("options.discrete_mouse_scroll", (p_216634_0_) ->
+    public static final BooleanOption DISCRETE_MOUSE_SCROLL = new BooleanOption("options.discrete_mouse_scroll", (p_lambda$static$91_0_) ->
     {
-        return p_216634_0_.discreteMouseScroll;
-    }, (p_216625_0_, p_216625_1_) ->
+        return p_lambda$static$91_0_.discreteMouseScroll;
+    }, (p_lambda$static$92_0_, p_lambda$static$92_1_) ->
     {
-        p_216625_0_.discreteMouseScroll = p_216625_1_;
+        p_lambda$static$92_0_.discreteMouseScroll = p_lambda$static$92_1_;
     });
-    public static final BooleanOption ENABLE_VSYNC = new BooleanOption("options.vsync", (p_216661_0_) ->
+    public static final BooleanOption ENABLE_VSYNC = new BooleanOption("options.vsync", (p_lambda$static$93_0_) ->
     {
-        return p_216661_0_.enableVsync;
-    }, (p_216635_0_, p_216635_1_) ->
+        return p_lambda$static$93_0_.enableVsync;
+    }, (p_lambda$static$94_0_, p_lambda$static$94_1_) ->
     {
-        p_216635_0_.enableVsync = p_216635_1_;
+        p_lambda$static$94_0_.enableVsync = p_lambda$static$94_1_;
 
         if (Minecraft.getInstance().getWindow() != null)
         {
-            Minecraft.getInstance().getWindow().updateVsync(p_216635_0_.enableVsync);
+            Minecraft.getInstance().getWindow().updateVsync(p_lambda$static$94_0_.enableVsync);
         }
     });
-    public static final BooleanOption ENTITY_SHADOWS = new BooleanOption("options.entityShadows", (p_216576_0_) ->
+    public static final BooleanOption ENTITY_SHADOWS = new BooleanOption("options.entityShadows", (p_lambda$static$95_0_) ->
     {
-        return p_216576_0_.entityShadows;
-    }, (p_216588_0_, p_216588_1_) ->
+        return p_lambda$static$95_0_.entityShadows;
+    }, (p_lambda$static$96_0_, p_lambda$static$96_1_) ->
     {
-        p_216588_0_.entityShadows = p_216588_1_;
+        p_lambda$static$96_0_.entityShadows = p_lambda$static$96_1_;
     });
-    public static final BooleanOption FORCE_UNICODE_FONT = new BooleanOption("options.forceUnicodeFont", (p_216657_0_) ->
+    public static final BooleanOption FORCE_UNICODE_FONT = new BooleanOption("options.forceUnicodeFont", (p_lambda$static$97_0_) ->
     {
-        return p_216657_0_.forceUnicodeFont;
-    }, (p_216631_0_, p_216631_1_) ->
+        return p_lambda$static$97_0_.forceUnicodeFont;
+    }, (p_lambda$static$98_0_, p_lambda$static$98_1_) ->
     {
-        p_216631_0_.forceUnicodeFont = p_216631_1_;
+        p_lambda$static$98_0_.forceUnicodeFont = p_lambda$static$98_1_;
         Minecraft minecraft = Minecraft.getInstance();
 
         if (minecraft.getWindow() != null)
         {
-            minecraft.selectMainFont(p_216631_1_);
+            minecraft.selectMainFont(p_lambda$static$98_1_);
         }
     });
-    public static final BooleanOption INVERT_MOUSE = new BooleanOption("options.invertMouse", (p_216627_0_) ->
+    public static final BooleanOption INVERT_MOUSE = new BooleanOption("options.invertMouse", (p_lambda$static$99_0_) ->
     {
-        return p_216627_0_.invertYMouse;
-    }, (p_216603_0_, p_216603_1_) ->
+        return p_lambda$static$99_0_.invertYMouse;
+    }, (p_lambda$static$100_0_, p_lambda$static$100_1_) ->
     {
-        p_216603_0_.invertYMouse = p_216603_1_;
+        p_lambda$static$100_0_.invertYMouse = p_lambda$static$100_1_;
     });
-    public static final BooleanOption REALMS_NOTIFICATIONS = new BooleanOption("options.realmsNotifications", (p_216606_0_) ->
+    public static final BooleanOption REALMS_NOTIFICATIONS = new BooleanOption("options.realmsNotifications", (p_lambda$static$101_0_) ->
     {
-        return p_216606_0_.realmsNotifications;
-    }, (p_216618_0_, p_216618_1_) ->
+        return p_lambda$static$101_0_.realmsNotifications;
+    }, (p_lambda$static$102_0_, p_lambda$static$102_1_) ->
     {
-        p_216618_0_.realmsNotifications = p_216618_1_;
+        p_lambda$static$102_0_.realmsNotifications = p_lambda$static$102_1_;
     });
-    public static final BooleanOption REDUCED_DEBUG_INFO = new BooleanOption("options.reducedDebugInfo", (p_216582_0_) ->
+    public static final BooleanOption REDUCED_DEBUG_INFO = new BooleanOption("options.reducedDebugInfo", (p_lambda$static$103_0_) ->
     {
-        return p_216582_0_.reducedDebugInfo;
-    }, (p_216613_0_, p_216613_1_) ->
+        return p_lambda$static$103_0_.reducedDebugInfo;
+    }, (p_lambda$static$104_0_, p_lambda$static$104_1_) ->
     {
-        p_216613_0_.reducedDebugInfo = p_216613_1_;
+        p_lambda$static$104_0_.reducedDebugInfo = p_lambda$static$104_1_;
     });
-    public static final BooleanOption SHOW_SUBTITLES = new BooleanOption("options.showSubtitles", (p_216663_0_) ->
+    public static final BooleanOption SHOW_SUBTITLES = new BooleanOption("options.showSubtitles", (p_lambda$static$105_0_) ->
     {
-        return p_216663_0_.showSubtitles;
-    }, (p_216662_0_, p_216662_1_) ->
+        return p_lambda$static$105_0_.showSubtitles;
+    }, (p_lambda$static$106_0_, p_lambda$static$106_1_) ->
     {
-        p_216662_0_.showSubtitles = p_216662_1_;
+        p_lambda$static$106_0_.showSubtitles = p_lambda$static$106_1_;
     });
-    public static final BooleanOption SNOOPER_ENABLED = new BooleanOption("options.snooper", (p_216638_0_) ->
+    public static final BooleanOption SNOOPER_ENABLED = new BooleanOption("options.snooper", (p_lambda$static$107_0_) ->
     {
-        if (p_216638_0_.snooperEnabled)
+        if (p_lambda$static$107_0_.snooperEnabled)
         {
         }
 
         return false;
-    }, (p_216676_0_, p_216676_1_) ->
+    }, (p_lambda$static$108_0_, p_lambda$static$108_1_) ->
     {
-        p_216676_0_.snooperEnabled = p_216676_1_;
+        p_lambda$static$108_0_.snooperEnabled = p_lambda$static$108_1_;
     });
-    public static final CycleOption TOGGLE_CROUCH = new CycleOption("key.sneak", (p_228043_0_, p_228043_1_) ->
+    public static final CycleOption TOGGLE_CROUCH = new CycleOption("key.sneak", (p_lambda$static$109_0_, p_lambda$static$109_1_) ->
     {
-        p_228043_0_.toggleCrouch = !p_228043_0_.toggleCrouch;
-    }, (p_228041_0_, p_228041_1_) ->
+        p_lambda$static$109_0_.toggleCrouch = !p_lambda$static$109_0_.toggleCrouch;
+    }, (p_lambda$static$110_0_, p_lambda$static$110_1_) ->
     {
-        return p_228041_1_.genericValueLabel(new TranslatableComponent(p_228041_0_.toggleCrouch ? "options.key.toggle" : "options.key.hold"));
+        return p_lambda$static$110_1_.genericValueLabel(new TranslatableComponent(p_lambda$static$110_0_.toggleCrouch ? "options.key.toggle" : "options.key.hold"));
     });
-    public static final CycleOption TOGGLE_SPRINT = new CycleOption("key.sprint", (p_228039_0_, p_228039_1_) ->
+    public static final CycleOption TOGGLE_SPRINT = new CycleOption("key.sprint", (p_lambda$static$111_0_, p_lambda$static$111_1_) ->
     {
-        p_228039_0_.toggleSprint = !p_228039_0_.toggleSprint;
-    }, (p_228037_0_, p_228037_1_) ->
+        p_lambda$static$111_0_.toggleSprint = !p_lambda$static$111_0_.toggleSprint;
+    }, (p_lambda$static$112_0_, p_lambda$static$112_1_) ->
     {
-        return p_228037_1_.genericValueLabel(new TranslatableComponent(p_228037_0_.toggleSprint ? "options.key.toggle" : "options.key.hold"));
+        return p_lambda$static$112_1_.genericValueLabel(new TranslatableComponent(p_lambda$static$112_0_.toggleSprint ? "options.key.toggle" : "options.key.hold"));
     });
-    public static final BooleanOption TOUCHSCREEN = new BooleanOption("options.touchscreen", (p_216647_0_) ->
+    public static final BooleanOption TOUCHSCREEN = new BooleanOption("options.touchscreen", (p_lambda$static$113_0_) ->
     {
-        return p_216647_0_.touchscreen;
-    }, (p_216580_0_, p_216580_1_) ->
+        return p_lambda$static$113_0_.touchscreen;
+    }, (p_lambda$static$114_0_, p_lambda$static$114_1_) ->
     {
-        p_216580_0_.touchscreen = p_216580_1_;
+        p_lambda$static$114_0_.touchscreen = p_lambda$static$114_1_;
     });
-    public static final BooleanOption USE_FULLSCREEN = new BooleanOption("options.fullscreen", (p_228040_0_) ->
+    public static final BooleanOption USE_FULLSCREEN = new BooleanOption("options.fullscreen", (p_lambda$static$115_0_) ->
     {
-        return p_228040_0_.fullscreen;
-    }, (p_228042_0_, p_228042_1_) ->
+        return p_lambda$static$115_0_.fullscreen;
+    }, (p_lambda$static$116_0_, p_lambda$static$116_1_) ->
     {
-        p_228042_0_.fullscreen = p_228042_1_;
+        p_lambda$static$116_0_.fullscreen = p_lambda$static$116_1_;
         Minecraft minecraft = Minecraft.getInstance();
 
-        if (minecraft.getWindow() != null && minecraft.getWindow().isFullscreen() != p_228042_0_.fullscreen)
+        if (minecraft.getWindow() != null && minecraft.getWindow().isFullscreen() != p_lambda$static$116_0_.fullscreen)
         {
             minecraft.getWindow().toggleFullScreen();
-            p_228042_0_.fullscreen = minecraft.getWindow().isFullscreen();
+            p_lambda$static$116_0_.fullscreen = minecraft.getWindow().isFullscreen();
         }
     });
-    public static final BooleanOption VIEW_BOBBING = new BooleanOption("options.viewBobbing", (p_228036_0_) ->
+    public static final BooleanOption VIEW_BOBBING = new BooleanOption("options.viewBobbing", (p_lambda$static$117_0_) ->
     {
-        return p_228036_0_.bobView;
-    }, (p_228038_0_, p_228038_1_) ->
+        return p_lambda$static$117_0_.bobView;
+    }, (p_lambda$static$118_0_, p_lambda$static$118_1_) ->
     {
-        p_228038_0_.bobView = p_228038_1_;
+        p_lambda$static$118_0_.bobView = p_lambda$static$118_1_;
     });
     private final Component caption;
     private Optional<List<FormattedCharSequence>> toolTip = Optional.empty();
