--- a/net/minecraft/client/renderer/blockentity/BlockEntityRenderDispatcher.java
+++ b/net/minecraft/client/renderer/blockentity/BlockEntityRenderDispatcher.java
@@ -174,7 +174,7 @@
         return this.context;
     }
 
-    public Map<BlockEntityType, BlockEntityRenderer> getBlockEntityRenderMap()
+    public Map getBlockEntityRenderMap()
     {
         if (this.renderers instanceof ImmutableMap)
         {
