--- a/net/minecraft/client/renderer/texture/Stitcher.java
+++ b/net/minecraft/client/renderer/texture/Stitcher.java
@@ -12,7 +12,7 @@
 
 public class Stitcher
 {
-    private static final Comparator<Stitcher.Holder> HOLDER_COMPARATOR = Comparator.comparing((holderIn) ->
+    private static final Comparator<Stitcher.Holder> HOLDER_COMPARATOR = Comparator.comparing(( Stitcher.Holder holderIn) ->
     {
         return -holderIn.height;
     }).thenComparing((holderIn) ->
