--- a/net/minecraft/world/level/block/Block.java
+++ b/net/minecraft/world/level/block/Block.java
@@ -354,9 +354,9 @@
     {
         if (pLevel instanceof ServerLevel)
         {
-            getDrops(pState, (ServerLevel)pLevel, pPos, pBlockEntity, pEntity, pStack).forEach((pStack) ->
+            getDrops(pState, (ServerLevel)pLevel, pPos, pBlockEntity, pEntity, pStack).forEach((pStack1) ->
             {
-                popResource(pLevel, pPos, pStack);
+                popResource(pLevel, pPos, pStack1);
             });
             pState.spawnAfterBreak((ServerLevel)pLevel, pPos, pStack);
         }
