--- a/net/minecraft/data/HashCache.java
+++ b/net/minecraft/data/HashCache.java
@@ -40,10 +40,10 @@
 
         if (Files.isReadable(this.cachePath))
         {
-            IOUtils.readLines(Files.newInputStream(this.cachePath), Charsets.UTF_8).forEach((pFile) ->
+            IOUtils.readLines(Files.newInputStream(this.cachePath), Charsets.UTF_8).forEach((pFile1) ->
             {
-                int i = pFile.indexOf(32);
-                this.oldCache.put(p_123935_.resolve(pFile.substring(i + 1)), pFile.substring(0, i));
+                int i = pFile1.indexOf(32);
+                this.oldCache.put(p_123935_.resolve(pFile1.substring(i + 1)), pFile1.substring(0, i));
             });
         }
     }
