--- a/net/minecraft/world/entity/vehicle/Boat.java
+++ b/net/minecraft/world/entity/vehicle/Boat.java
@@ -600,7 +600,7 @@
 
                             if (!(blockstate.getBlock() instanceof WaterlilyBlock) && Shapes.joinIsNotEmpty(blockstate.getCollisionShape(this.level, blockpos$mutableblockpos).move((double)l1, (double)k2, (double)i2), voxelshape, BooleanOp.AND))
                             {
-                                f += blockstate.getBlock().getFriction();
+                                    f += blockstate.getBlock().getFriction();
                                 ++k1;
                             }
                         }
