--- a/net/minecraft/client/renderer/chunk/ChunkRenderDispatcher.java
+++ b/net/minecraft/client/renderer/chunk/ChunkRenderDispatcher.java
@@ -1199,7 +1199,7 @@
                     Random random = new Random();
                     BlockRenderDispatcher blockrenderdispatcher = Minecraft.getInstance().getBlockRenderer();
 
-                    for (BlockPosM blockposm : BlockPosM.getAllInBoxMutable(blockpos, blockpos1))
+                    for (BlockPosM blockposm : (Iterable<BlockPosM>)BlockPosM.getAllInBoxMutable(blockpos, blockpos1))
                     {
                         BlockState blockstate = chunkcacheof.getBlockState(blockposm);
 
