--- a/net/optifine/shaders/ShadersFramebuffer.java
+++ b/net/optifine/shaders/ShadersFramebuffer.java
@@ -36,7 +36,7 @@
     private int[] colorImageUnits;
     private int glFramebuffer;
     private FlipTextures colorTexturesFlip;
-    private IntBuffer depthTextures;
+    public IntBuffer depthTextures;
     private final DrawBuffers drawBuffers;
     private DrawBuffers activeDrawBuffers;
     private int[] drawColorTextures;
@@ -107,10 +107,25 @@
                 GL30.glTexParameteri(3553, 34892, 34894);
             }
 
-            GL30.glTexImage2D(3553, 0, 6402, this.width, this.height, 0, 6402, 5126, (FloatBuffer)null);
+            if (this.name == "dfb")
+            {
+                GL30.glTexImage2D(3553, 0, 35056, this.width, this.height, 0, 34041, 34042, (FloatBuffer)null);
+            }
+            else
+            {
+                GL30.glTexImage2D(3553, 0, 6402, this.width, this.height, 0, 6402, 5126, (FloatBuffer)null);
+            }
+        }
+
+        if (this.name == "dfb")
+        {
+            this.setFramebufferTexture2D(36160, 33306, 3553, this.depthTextures.get(0), 0);
+        }
+        else
+        {
+            this.setFramebufferTexture2D(36160, 36096, 3553, this.depthTextures.get(0), 0);
         }
 
-        this.setFramebufferTexture2D(36160, 36096, 3553, this.depthTextures.get(0), 0);
         Shaders.checkGLError("FBS " + this.name + " depth");
 
         for (int l = 0; l < this.usedColorBuffers; ++l)
