--- a/net/minecraft/client/resources/model/ModelBakery.java
+++ b/net/minecraft/client/resources/model/ModelBakery.java
@@ -588,7 +588,7 @@
                     });
                     map3.forEach((keyIn, setIn) ->
                     {
-                        Iterator<BlockState> iterator = setIn.iterator();
+                        Iterator<BlockState> iterator = ((Iterable<BlockState>)setIn).iterator();
 
                         while (iterator.hasNext())
                         {
@@ -601,9 +601,9 @@
                             }
                         }
 
-                        if (setIn.size() > 1)
+                        if (((Set) setIn).size() > 1)
                         {
-                            this.registerModelGroup(setIn);
+                            this.registerModelGroup((Iterable<BlockState>) setIn);
                         }
                     });
                 }
