--- a/net/minecraft/client/Minecraft.java
+++ b/net/minecraft/client/Minecraft.java
@@ -54,6 +54,7 @@
 import java.util.Locale;
 import java.util.MissingResourceException;
 import java.util.Queue;
+import java.util.Set;
 import java.util.UUID;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ExecutionException;
@@ -248,11 +249,13 @@
 import org.apache.commons.io.FileUtils;
 import org.lwjgl.util.tinyfd.TinyFileDialogs;
 import org.slf4j.Logger;
+import org.vivecraft.render.PlayerModelController;
 
 public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements WindowEventHandler
 {
     private static Minecraft instance;
     private static final Logger LOGGER = LogUtils.getLogger();
+    public final String minecriftVerString = "Vivecraft 1.18.2  jrbudda-NONVR-2-a4";
     public static final boolean ON_OSX = Util.getPlatform() == Util.OS.OSX;
     private static final int MAX_TICKS_PER_UPDATE = 10;
     public static final ResourceLocation DEFAULT_FONT = new ResourceLocation("default");
@@ -599,7 +602,7 @@
 
     private String createTitle()
     {
-        StringBuilder stringbuilder = new StringBuilder("Minecraft");
+        StringBuilder stringbuilder = new StringBuilder(minecriftVerString);
 
         if (checkModStatus().shouldReportAsModified())
         {
@@ -2008,7 +2011,9 @@
         {
             this.gameRenderer.shutdownEffect();
         }
-
+        //Vivecraft
+        PlayerModelController.getInstance().tick();
+        //
         if (!this.pause)
         {
             this.musicManager.tick();
