--- a/net/optifine/shaders/ShadersRender.java
+++ b/net/optifine/shaders/ShadersRender.java
@@ -369,7 +369,7 @@
             int k = minecraft.level.getMaxBuildHeight();
             LongOpenHashSet longopenhashset = levelrenderer.getRenderChunksEntities();
 
-            for (Entity entity : Shaders.isRenderShadowEntities() ? Shaders.getCurrentWorld().entitiesForRendering() : Collections.EMPTY_LIST)
+            for (Entity entity : Shaders.isRenderShadowEntities() ? Shaders.getCurrentWorld().entitiesForRendering() : (List<Entity>)Collections.EMPTY_LIST)
             {
                 BlockPos blockpos = entity.blockPosition();
 
@@ -396,7 +396,7 @@
             Frustum frustum1 = frustum;
             label103:
 
-            for (LevelRenderer.RenderChunkInfo levelrenderer$renderchunkinfo : Shaders.isRenderShadowBlockEntities() ? levelrenderer1.getRenderInfosTileEntities() : Collections.EMPTY_LIST)
+            for (LevelRenderer.RenderChunkInfo levelrenderer$renderchunkinfo : Shaders.isRenderShadowBlockEntities() ? levelrenderer1.getRenderInfosTileEntities() : (List<LevelRenderer.RenderChunkInfo>)Collections.EMPTY_LIST)
             {
                 List<BlockEntity> list = levelrenderer$renderchunkinfo.chunk.getCompiledChunk().getRenderableBlockEntities();
 
