--- a/net/optifine/shaders/Shaders.java
+++ b/net/optifine/shaders/Shaders.java
@@ -129,6 +129,7 @@
 import net.optifine.util.WorldUtils;
 import org.apache.commons.io.IOUtils;
 import org.lwjgl.BufferUtils;
+import org.lwjgl.opengl.ARBFragmentShader;
 import org.lwjgl.opengl.ARBGeometryShader4;
 import org.lwjgl.opengl.ARBShaderObjects;
 import org.lwjgl.opengl.ARBVertexShader;
@@ -891,6 +892,7 @@
 
     public static void loadShaderPack()
     {
+    	mc = Minecraft.getInstance();
         boolean flag = shaderPackLoaded;
         boolean flag1 = isOldLighting();
 
@@ -1204,7 +1206,7 @@
         Set set = props.keySet();
         List<ICustomTexture> list = new ArrayList<>();
 
-        for (String s1 : set)
+        for (String s1 : (Set<String>)(Set<?>)set)
         {
             if (s1.startsWith(s))
             {
@@ -6433,7 +6435,7 @@
                         Lang.loadLocaleData(inputstream, properties);
                         inputstream.close();
 
-                        for (String s5 : properties.keySet())
+                        for (String s5 : (Set<String>)(Set<?>)properties.keySet())
                         {
                             String s6 = properties.getProperty(s5);
                             shaderPackResources.put(s5, s6);
