--- a/net/optifine/CrashReporter.java
+++ b/net/optifine/CrashReporter.java
@@ -107,10 +107,10 @@
             cat.setDetail("Multitexture", "" + Config.isMultiTexture());
         }
 
-        cat.setDetail("Shaders", Shaders.getShaderPackName().makeConcatWithConstants<invokedynamic>(Shaders.getShaderPackName()));
-        cat.setDetail("OpenGlVersion", Config.openGlVersion.makeConcatWithConstants<invokedynamic>(Config.openGlVersion));
-        cat.setDetail("OpenGlRenderer", Config.openGlRenderer.makeConcatWithConstants<invokedynamic>(Config.openGlRenderer));
-        cat.setDetail("OpenGlVendor", Config.openGlVendor.makeConcatWithConstants<invokedynamic>(Config.openGlVendor));
+        cat.setDetail("Shaders", Shaders.getShaderPackName());
+        cat.setDetail("OpenGlVersion", Config.openGlVersion);
+        cat.setDetail("OpenGlRenderer", Config.openGlRenderer);
+        cat.setDetail("OpenGlVendor", Config.openGlVendor);
         cat.setDetail("CpuCount", "" + Config.getAvailableProcessors());
     }
 }
