--- a/net/minecraft/client/renderer/culling/Frustum.java
+++ b/net/minecraft/client/renderer/culling/Frustum.java
@@ -233,4 +233,16 @@
     {
         return this.camZ;
     }
+
+	@Override
+	public void setCameraPosition(double var1, double var3, double var5) {
+		// TODO Auto-generated method stub
+		
+	}
+
+	@Override
+	public boolean isBoundingBoxInFrustum(AABB var1) {
+		// TODO Auto-generated method stub
+		return false;
+	}
 }
