--- a/net/optifine/override/PlayerControllerOF.java
+++ b/net/optifine/override/PlayerControllerOF.java
@@ -14,6 +14,7 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.EntityHitResult;
+import org.vivecraft.api.NetworkHelper;
 
 public class PlayerControllerOF extends MultiPlayerGameMode
 {
@@ -47,13 +48,21 @@
     public InteractionResult useItem(Player p_105236_, Level pPlayer, InteractionHand pLevel)
     {
         this.acting = true;
+        NetworkHelper.overrideLook(p_105236_, Minecraft.getInstance().vrPlayer.getRightClickLookOverride(p_105236_, pLevel.ordinal()));
         InteractionResult interactionresult = super.useItem(p_105236_, pPlayer, pLevel);
         this.acting = false;
         return interactionresult;
     }
 
+    public void releaseUsingItem(Player p_105278_)
+    {
+        NetworkHelper.overrideLook(p_105278_, Minecraft.getInstance().vrPlayer.getRightClickLookOverride(p_105278_, p_105278_.getUsedItemHand().ordinal()));
+        super.releaseUsingItem(p_105278_);
+    }
+
     public InteractionResult useItemOn(LocalPlayer p_105263_, ClientLevel p_105264_, InteractionHand p_105265_, BlockHitResult p_105266_)
     {
+        NetworkHelper.overrideLook(p_105263_, p_105266_.getLocation().subtract(p_105263_.getEyePosition(1.0F)).normalize());
         this.acting = true;
         this.lastClickBlockPos = p_105266_.getBlockPos();
         InteractionResult interactionresult = super.useItemOn(p_105263_, p_105264_, p_105265_, p_105266_);
