--- a/net/optifine/entity/model/ModelAdapterSheepWool.java
+++ b/net/optifine/entity/model/ModelAdapterSheepWool.java
@@ -28,7 +28,7 @@
     public IEntityRenderer makeEntityRender(Model modelBase, float shadowSize)
     {
         EntityRenderDispatcher entityrenderdispatcher = Minecraft.getInstance().getEntityRenderDispatcher();
-        EntityRenderer entityrenderer = entityrenderdispatcher.getEntityRenderMap().get(EntityType.SHEEP);
+        EntityRenderer entityrenderer = (EntityRenderer) entityrenderdispatcher.getEntityRenderMap().get(EntityType.SHEEP);
 
         if (!(entityrenderer instanceof SheepRenderer))
         {
