--- a/net/minecraft/client/gui/screens/PauseScreen.java
+++ b/net/minecraft/client/gui/screens/PauseScreen.java
@@ -1,13 +1,21 @@
 package net.minecraft.client.gui.screens;
 
+import org.vivecraft.gui.settings.GuiQuickCommandsInGame;
+import org.vivecraft.settings.AutoCalibration;
+import org.vivecraft.settings.VRHotkeys;
+import org.vivecraft.utils.external.jkatvr;
+
 import com.mojang.blaze3d.vertex.PoseStack;
 import com.mojang.realmsclient.RealmsMainScreen;
 import net.minecraft.SharedConstants;
 import net.minecraft.Util;
+import net.minecraft.client.Minecraft;
 import net.minecraft.client.gui.components.Button;
 import net.minecraft.client.gui.screens.achievement.StatsScreen;
 import net.minecraft.client.gui.screens.advancements.AdvancementsScreen;
 import net.minecraft.client.gui.screens.multiplayer.JoinMultiplayerScreen;
+import net.minecraft.client.gui.screens.social.SocialInteractionsScreen;
+import net.minecraft.client.main.Main;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.TranslatableComponent;
 
@@ -30,6 +38,82 @@
         {
             this.createPauseMenu();
         }
+
+        if (!Minecraft.getInstance().isMultiplayerServer())
+        {
+            this.addRenderableWidget(new Button(this.width / 2 - 102, this.height / 4 + 72 + -16, 98, 20, new TranslatableComponent("vivecraft.gui.chat"), (p) ->
+            {
+                this.minecraft.setScreen(new ChatScreen(""));
+            }));
+        }
+        else
+        {
+            this.addRenderableWidget(new Button(this.width / 2 - 102, this.height / 4 + 72 + -16, 46, 20, new TranslatableComponent("vivecraft.gui.chat"), (p) ->
+            {
+                this.minecraft.setScreen(new ChatScreen(""));
+            }));
+            this.addRenderableWidget(new Button(this.width / 2 - 102 + 48, this.height / 4 + 72 + -16, 46, 20, new TranslatableComponent("vivecraft.gui.social"), (p) ->
+            {
+                this.minecraft.setScreen(new SocialInteractionsScreen());
+            }));
+        }
+
+        this.addRenderableWidget(new Button(this.width / 2 + 4, this.height / 4 + 72 + -16, 98, 20, new TranslatableComponent("vivecraft.gui.commands"), (p) ->
+        {
+            this.minecraft.setScreen(new GuiQuickCommandsInGame(this));
+            this.init();
+        }));
+        this.addRenderableWidget(new Button(this.width / 2 - 102, this.height / 4 + 120 + -16, 49, 20, new TranslatableComponent("vivecraft.gui.overlay"), (p) ->
+        {
+            this.minecraft.options.renderDebug = !this.minecraft.options.renderDebug;
+            this.minecraft.setScreen((Screen)null);
+        }));
+        this.addRenderableWidget(new Button(this.width / 2 - 52, this.height / 4 + 120 + -16, 49, 20, new TranslatableComponent("vivecraft.gui.profiler"), (p) ->
+        {
+            this.minecraft.options.renderDebugCharts = !this.minecraft.options.renderDebugCharts;
+            this.minecraft.options.ofProfiler = this.minecraft.options.renderDebugCharts;
+            this.minecraft.setScreen((Screen)null);
+        }));
+        this.addRenderableWidget(new Button(this.width / 2 + 4, this.height / 4 + 120 + -16, 98, 20, new TranslatableComponent("vivecraft.gui.screenshot"), (p) ->
+        {
+            this.minecraft.setScreen((Screen)null);
+            this.minecraft.grabScreenShot = true;
+        }));
+
+        if (!this.minecraft.vrSettings.seated)
+        {
+            this.addRenderableWidget(new Button(this.width / 2 - 102, this.height / 4 + 144 + -16, 98, 20, new TranslatableComponent("vivecraft.gui.calibrateheight"), (p) ->
+            {
+                AutoCalibration.calibrateManual();
+                this.minecraft.vrSettings.saveOptions();
+                this.minecraft.setScreen((Screen)null);
+            }));
+        }
+
+        if (Main.katvr)
+        {
+            this.addRenderableWidget(new Button(this.width / 2 + 106, this.height / 4 + 144 + -16, 98, 20, new TranslatableComponent("vivecraft.gui.alignkatwalk"), (p) ->
+            {
+                jkatvr.resetYaw(this.minecraft.vrPlayer.vrdata_room_pre.hmd.getYaw());
+                this.minecraft.setScreen((Screen)null);
+            }));
+        }
+
+        if (!this.minecraft.vrSettings.seated || this.minecraft.vrSettings.displayMirrorMode == 14 || this.minecraft.vrSettings.displayMirrorMode == 15)
+        {
+            this.addRenderableWidget(new Button(this.width / 2 + 4, this.height / 4 + 144 + -16, 98, 20, new TranslatableComponent("vivecraft.gui.movethirdpersoncam"), (p) ->
+            {
+                if (!VRHotkeys.isMovingThirdPersonCam())
+                {
+                    VRHotkeys.startMovingThirdPersonCam(1, VRHotkeys.Triggerer.MENUBUTTON);
+                }
+                else if (VRHotkeys.getMovingThirdPersonCamTriggerer() == VRHotkeys.Triggerer.MENUBUTTON)
+                {
+                    VRHotkeys.stopMovingThirdPersonCam();
+                    this.minecraft.vrSettings.saveOptions();
+                }
+            }));
+        }
     }
 
     private void createPauseMenu()
