--- a/net/minecraft/client/gui/screens/Screen.java
+++ b/net/minecraft/client/gui/screens/Screen.java
@@ -16,6 +16,7 @@
 import java.net.URI;
 import java.net.URISyntaxException;
 import java.nio.file.Path;
+import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
 import java.util.Locale;
@@ -31,6 +32,7 @@
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.gui.Font;
 import net.minecraft.client.gui.chat.NarratorChatListener;
+import net.minecraft.client.gui.components.AbstractWidget;
 import net.minecraft.client.gui.components.Widget;
 import net.minecraft.client.gui.components.events.AbstractContainerEventHandler;
 import net.minecraft.client.gui.components.events.GuiEventListener;
@@ -51,6 +53,9 @@
 import net.minecraft.world.inventory.tooltip.TooltipComponent;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.TooltipFlag;
+import net.optifine.reflect.Reflector;
+import net.optifine.util.GuiUtils;
+
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
@@ -68,7 +73,8 @@
     protected ItemRenderer itemRenderer;
     public int width;
     public int height;
-    private final List<Widget> renderables = Lists.newArrayList();
+    //vivecraft idk why this was private...
+    protected final List<Widget> renderables = Lists.newArrayList();
     public boolean passEvents;
     protected Font font;
     private URI clickedLink;
@@ -183,7 +189,20 @@
 
     protected void renderTooltip(PoseStack p_96566_, ItemStack p_96567_, int p_96568_, int p_96569_)
     {
-        this.renderTooltip(p_96566_, this.getTooltipFromItem(p_96567_), p_96567_.getTooltipImage(), p_96568_, p_96569_);
+    	//TODO: fix
+    	if (Reflector.GuiUtils.exists())
+    	{
+//    		Reflector.call(Reflector.GuiUtils_preItemToolTip, p_96567_);
+//    		Font font = (Font)Reflector.call(p_96567_.getItem(), Reflector.IForgeItem_getFontRenderer, p_96567_);
+//    		this.renderTooltip(p_96566_, Lists.transform(this.getTooltipFromItem(p_96567_), Component::getVisualOrderText), p_96568_, p_96569_, font == null ? this.font : font);
+//    		Reflector.call(Reflector.GuiUtils_postItemToolTip);
+    	}
+    	else
+    	{
+    	//	this.renderComponentTooltip(p_96566_, this.getTooltipFromItem(p_96567_), p_96568_, p_96569_);
+    	}
+    	
+    	this.renderTooltip(p_96566_, this.getTooltipFromItem(p_96567_), p_96567_.getTooltipImage(), p_96568_, p_96569_);
     }
 
     public void renderTooltip(PoseStack p_169389_, List<Component> p_169390_, Optional<TooltipComponent> p_169391_, int p_169392_, int p_169393_)
@@ -208,7 +227,16 @@
 
     public void renderComponentTooltip(PoseStack p_96598_, List<Component> p_96599_, int p_96600_, int p_96601_)
     {
-        this.renderTooltip(p_96598_, Lists.transform(p_96599_, Component::getVisualOrderText), p_96600_, p_96601_);
+    	if (Reflector.GuiUtils.exists())
+    	{
+    		//this.renderWrappedToolTip(p_96598_, p_96599_, p_96600_, p_96601_, this.font);
+    	}
+    	else
+    	{
+    		this.renderTooltip(p_96598_, Lists.transform(p_96599_, Component::getVisualOrderText), p_96600_, p_96601_);
+    	}
+    	
+        //this.renderTooltip(p_96598_, Lists.transform(p_96599_, Component::getVisualOrderText), p_96600_, p_96601_);
     }
 
     public void renderTooltip(PoseStack p_96618_, List <? extends FormattedCharSequence > p_96619_, int p_96620_, int p_96621_)
@@ -434,12 +462,20 @@
 
     public void sendMessage(String p_96613_, boolean p_96614_)
     {
-        if (p_96614_)
+        if (Reflector.ForgeEventFactory_onClientSendMessage.exists())
         {
-            this.minecraft.gui.getChat().addRecentChat(p_96613_);
+            p_96613_ = Reflector.callString(Reflector.ForgeEventFactory_onClientSendMessage, p_96613_);
         }
 
-        this.minecraft.player.chat(p_96613_);
+        if (!p_96613_.isEmpty())
+        {
+            if (p_96614_)
+            {
+                this.minecraft.gui.getChat().addRecentChat(p_96613_);
+            }
+
+            this.minecraft.player.chat(p_96613_);
+        }
     }
 
     public final void init(Minecraft p_96607_, int p_96608_, int p_96609_)
@@ -449,9 +485,14 @@
         this.font = p_96607_.font;
         this.width = p_96608_;
         this.height = p_96609_;
-        this.clearWidgets();
-        this.setFocused((GuiEventListener)null);
-        this.init();
+        //TODO: fix
+        //if (!Reflector.postForgeBusEvent(Reflector.InitGuiEvent_Pre_Constructor, this, this.buttons, consumer1, consumer))
+        //{
+            this.clearWidgets();
+            this.setFocused((GuiEventListener)null);
+            this.init();
+        //}
+        //Reflector.postForgeBusEvent(Reflector.InitGuiEvent_Post_Constructor, this, this.buttons, consumer1, consumer);
         this.triggerImmediateNarration(false);
         this.suppressNarration(NARRATE_SUPPRESS_AFTER_INIT_TIME);
     }
@@ -482,7 +523,16 @@
     {
         if (this.minecraft.level != null)
         {
-            this.fillGradient(p_96559_, 0, 0, this.width, this.height, -1072689136, -804253680);
+            if (this.minecraft.vrSettings != null && !this.minecraft.vrSettings.menuBackground)
+            {
+                this.fillGradient(p_96559_, 0, 0, this.width, this.height, 0, 0);
+            }
+            else
+            {
+                this.fillGradient(p_96559_, 0, 0, this.width, this.height, -1072689136, -804253680);
+            }
+
+            Reflector.postForgeBusEvent(Reflector.BackgroundDrawnEvent_Constructor, this, p_96559_);
         }
         else
         {
@@ -504,6 +554,7 @@
         bufferbuilder.vertex((double)this.width, 0.0D, 0.0D).uv((float)this.width / 32.0F, (float)p_96627_).color(64, 64, 64, 255).endVertex();
         bufferbuilder.vertex(0.0D, 0.0D, 0.0D).uv(0.0F, (float)p_96627_).color(64, 64, 64, 255).endVertex();
         tesselator.end();
+        Reflector.postForgeBusEvent(Reflector.BackgroundDrawnEvent_Constructor, this, new PoseStack());
     }
 
     public boolean isPauseScreen()
@@ -769,4 +820,60 @@
             this.priority = p_169426_;
         }
     }
+
+    public Minecraft getMinecraft()
+    {
+        return this.minecraft;
+    }
+    
+    //Vivecraft / OF
+    protected List<AbstractWidget> getButtonList()
+    {
+        List<AbstractWidget> list = new ArrayList<>();
+
+        for (GuiEventListener guieventlistener : this.children())
+        {
+            if (guieventlistener instanceof AbstractWidget)
+            {
+                list.add((AbstractWidget)guieventlistener);
+            }
+        }
+
+        return list;
+    }
+    
+    protected AbstractWidget getSelectedButton(int x, int y)
+    {
+        for (AbstractWidget butt : getButtonList())
+        {
+            if (butt.visible && butt.isHovered())
+            {
+                return butt;
+            }
+        }
+
+        return null;
+    }
+    
+    public static AbstractWidget getSelectedButton(int x, int y, List<AbstractWidget> listButtons)
+    {
+        for (int i = 0; i < listButtons.size(); ++i)
+        {
+            AbstractWidget abstractwidget = listButtons.get(i);
+
+            if (abstractwidget.visible)
+            {
+                int j = GuiUtils.getWidth(abstractwidget);
+                int k = GuiUtils.getHeight(abstractwidget);
+
+                if (x >= abstractwidget.x && y >= abstractwidget.y && x < abstractwidget.x + j && y < abstractwidget.y + k)
+                {
+                    return abstractwidget;
+                }
+            }
+        }
+
+        return null;
+    }
+    //
 }
