--- a/net/minecraft/client/gui/screens/SoundOptionsScreen.java
+++ b/net/minecraft/client/gui/screens/SoundOptionsScreen.java
@@ -7,8 +7,10 @@
 import net.minecraft.client.gui.components.OptionButton;
 import net.minecraft.client.gui.components.VolumeSlider;
 import net.minecraft.network.chat.CommonComponents;
+import net.minecraft.network.chat.TextComponent;
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.sounds.SoundSource;
+import org.vivecraft.settings.VRSettings;
 
 public class SoundOptionsScreen extends OptionsSubScreen
 {
@@ -32,16 +34,24 @@
             }
         }
 
+        this.addButton(new Button(this.width / 2 - 155 + i % 2 * 160, this.height / 6 - 12 + 24 * (i >> 1), 150, 20, new TextComponent(this.minecraft.vrSettings.getButtonDisplayString(VRSettings.VrOptions.HRTF_SELECTION)), (p_213104_1_) ->
+        {
+            this.buttons.clear();
+            this.children.clear();
+            this.minecraft.vrSettings.setOptionValue(VRSettings.VrOptions.HRTF_SELECTION);
+            this.minecraft.vrSettings.saveOptions();
+            this.init();
+        }));
         int j = this.width / 2 - 75;
         int k = this.height / 6 - 12;
         ++i;
-        this.addButton(new OptionButton(j, k + 24 * (i >> 1), 150, 20, Option.SHOW_SUBTITLES, Option.SHOW_SUBTITLES.getMessage(this.options), (p_96713_) ->
+        this.addButton(new OptionButton(j, k + 24 * (i >> 1), 150, 20, Option.SHOW_SUBTITLES, Option.SHOW_SUBTITLES.getMessage(this.options), (p_213105_1_) ->
         {
             Option.SHOW_SUBTITLES.toggle(this.minecraft.options);
-            p_96713_.setMessage(Option.SHOW_SUBTITLES.getMessage(this.minecraft.options));
+            p_213105_1_.setMessage(Option.SHOW_SUBTITLES.getMessage(this.minecraft.options));
             this.minecraft.options.save();
         }));
-        this.addButton(new Button(this.width / 2 - 100, this.height / 6 + 168, 200, 20, CommonComponents.GUI_DONE, (p_96710_) ->
+        this.addButton(new Button(this.width / 2 - 100, this.height / 6 + 168, 200, 20, CommonComponents.GUI_DONE, (p_213104_1_) ->
         {
             this.minecraft.setScreen(this.lastScreen);
         }));
