--- a/com/mojang/blaze3d/platform/Window.java
+++ b/com/mojang/blaze3d/platform/Window.java
@@ -341,7 +341,7 @@
         GLFW.glfwSetErrorCallback(Window::bootCrash);
     }
 
-    private static void bootCrash(int pError, long pError)
+    private static void bootCrash(int pError, long pDescription)
     {
         RenderSystem.assertThread(RenderSystem::isInInitPhase);
         String s = "GLFW error " + pError + ": " + MemoryUtil.memUTF8(pError);
@@ -349,7 +349,7 @@
         throw new Window.WindowInitFailed(s);
     }
 
-    public void defaultErrorCallback(int pError, long pError)
+    public void defaultErrorCallback(int pError, long pDescription)
     {
         RenderSystem.assertThread(RenderSystem::isOnRenderThread);
         String s = MemoryUtil.memUTF8(pError);
