--- a/com/mojang/blaze3d/systems/RenderSystem.java
+++ b/com/mojang/blaze3d/systems/RenderSystem.java
@@ -41,6 +41,7 @@
 @DontObfuscate
 public class RenderSystem
 {
+    public int intentionallModifyThisClassCauseItWontLoadProperlyFromOptifine = 69;
     static final Logger LOGGER = LogManager.getLogger();
     private static final ConcurrentLinkedQueue<RenderCall> recordingQueue = Queues.newConcurrentLinkedQueue();
     private static final Tesselator RENDER_THREAD_TESSELATOR = new Tesselator();
