--- a/net/minecraft/client/renderer/block/LiquidBlockRenderer.java
+++ b/net/minecraft/client/renderer/block/LiquidBlockRenderer.java
@@ -83,7 +83,7 @@
     public boolean tesselate(BlockAndTintGetter p_203174_, BlockPos p_203175_, VertexConsumer p_203176_, BlockState p_203177_, FluidState p_203178_)
     {
         BlockState blockstate = p_203178_.createLegacyBlock();
-        boolean f;
+        boolean flg;
 
         try
         {
@@ -481,7 +481,7 @@
                 return flag9;
             }
 
-            f = false;
+            flg = flag7;
         }
         finally
         {
@@ -491,7 +491,7 @@
             }
         }
 
-        return f;
+        return flg;
     }
 
     private float calculateAverageHeight(BlockAndTintGetter p_203150_, Fluid p_203151_, float p_203152_, float p_203153_, float p_203154_, BlockPos p_203155_)
