--- a/net/minecraft/client/renderer/block/LiquidBlockRenderer.java
+++ b/net/minecraft/client/renderer/block/LiquidBlockRenderer.java
@@ -87,7 +87,7 @@
     public boolean tesselate(BlockAndTintGetter pLightReader, BlockPos pPos, VertexConsumer pVertexBuilder, FluidState pFluidState)
     {
         BlockState blockstate = pFluidState.createLegacyBlock();
-        boolean f;
+        boolean flg, flag7 = false;
 
         try
         {
@@ -142,7 +142,6 @@
                 float f28 = (float)(i >> 16 & 255) / 255.0F;
                 float f1 = (float)(i >> 8 & 255) / 255.0F;
                 float f2 = (float)(i & 255) / 255.0F;
-                boolean flag7 = false;
                 float f3 = pLightReader.getShade(Direction.DOWN, true);
                 float f4 = pLightReader.getShade(Direction.UP, true);
                 float f5 = pLightReader.getShade(Direction.NORTH, true);
@@ -430,7 +429,7 @@
                 return flag7;
             }
 
-            f = false;
+            flg = flag7;
         }
         finally
         {
@@ -440,7 +439,7 @@
             }
         }
 
-        return f;
+        return flg;
     }
 
     private void vertex(VertexConsumer pVertexBuilder, double pX, double p_110987_, double pY, float p_110989_, float pZ, float p_110991_, float pRed, float pGreen, int pBlue)
