--- a/net/minecraft/client/renderer/block/LiquidBlockRenderer.java
+++ b/net/minecraft/client/renderer/block/LiquidBlockRenderer.java
@@ -87,7 +87,7 @@
     public boolean tesselate(BlockAndTintGetter pLightReader, BlockPos pPos, VertexConsumer pVertexBuilder, FluidState pFluidState)
     {
         BlockState blockstate = pFluidState.createLegacyBlock();
-        boolean f;
+        boolean flg, flag7 = false;
 
         try
         {
@@ -456,7 +456,7 @@
                 return flag8;
             }
 
-            f = false;
+            flg = flag7;
         }
         finally
         {
@@ -466,7 +466,7 @@
             }
         }
 
-        return f;
+        return flg;
     }
 
     private void vertex(VertexConsumer pVertexBuilder, double pX, double p_110987_, double pY, float p_110989_, float pZ, float p_110991_, float pRed, float pGreen, int pBlue)
