--- a/net/minecraft/client/gui/screens/inventory/CreativeModeInventoryScreen.java
+++ b/net/minecraft/client/gui/screens/inventory/CreativeModeInventoryScreen.java
@@ -14,6 +14,9 @@
 import java.util.Set;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
+
+import org.vivecraft.utils.ASMInjections;
+
 import net.minecraft.ChatFormatting;
 import net.minecraft.client.HotbarManager;
 import net.minecraft.client.Minecraft;
@@ -452,6 +455,7 @@
             (this.menu).items.addAll(searchtree.search(s.toLowerCase(Locale.ROOT)));
         }
 
+        ASMInjections.addCreativeSearch(this.searchBox.getValue(), (this.menu).items);
         this.scrollOffs = 0.0F;
         this.menu.scrollTo(0.0F);
     }
@@ -589,6 +593,7 @@
         else if (pTab != CreativeModeTab.TAB_SEARCH)
         {
             pTab.fillItemList((this.menu).items);
+            ASMInjections.addCreativeItems(pTab, (this.menu).items);
         }
 
         if (pTab == CreativeModeTab.TAB_INVENTORY)
