--- a/net/minecraft/client/renderer/ItemInHandRenderer.java
+++ b/net/minecraft/client/renderer/ItemInHandRenderer.java
@@ -1,11 +1,19 @@
 package net.minecraft.client.renderer;
 
 import com.google.common.base.MoreObjects;
+import com.mojang.blaze3d.systems.RenderSystem;
 import com.mojang.blaze3d.vertex.PoseStack;
 import com.mojang.blaze3d.vertex.VertexConsumer;
 import com.mojang.math.Matrix4f;
+import com.mojang.math.Quaternion;
 import com.mojang.math.Vector3f;
+import java.util.HashMap;
+import java.util.Map;
 import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
+import net.minecraft.Util;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.player.AbstractClientPlayer;
 import net.minecraft.client.player.LocalPlayer;
@@ -14,25 +22,54 @@
 import net.minecraft.client.renderer.entity.ItemRenderer;
 import net.minecraft.client.renderer.entity.player.PlayerRenderer;
 import net.minecraft.client.renderer.texture.OverlayTexture;
+import net.minecraft.client.resources.model.BakedModel;
+import net.minecraft.core.BlockPos;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.util.Mth;
 import net.minecraft.world.InteractionHand;
 import net.minecraft.world.entity.HumanoidArm;
 import net.minecraft.world.entity.LivingEntity;
+import net.minecraft.world.item.ArrowItem;
+import net.minecraft.world.item.BlockItem;
+import net.minecraft.world.item.CompassItem;
 import net.minecraft.world.item.CrossbowItem;
+import net.minecraft.world.item.FishingRodItem;
+import net.minecraft.world.item.FoodOnAStickItem;
+import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.Items;
 import net.minecraft.world.item.MapItem;
 import net.minecraft.world.item.ProjectileWeaponItem;
+import net.minecraft.world.item.ShieldItem;
+import net.minecraft.world.item.SwordItem;
+import net.minecraft.world.item.TridentItem;
+import net.minecraft.world.item.UseAnim;
+import net.minecraft.world.item.enchantment.EnchantmentHelper;
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.TorchBlock;
+import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.saveddata.maps.MapItemSavedData;
+import net.minecraft.world.phys.AABB;
+import net.minecraft.world.phys.Vec3;
 import net.optifine.Config;
 import net.optifine.CustomItems;
 import net.optifine.reflect.Reflector;
 import net.optifine.shaders.Shaders;
+import org.apache.commons.lang3.tuple.Triple;
+import org.vivecraft.gameplay.trackers.BowTracker;
+import org.vivecraft.gameplay.trackers.SwingTracker;
+import org.vivecraft.gameplay.trackers.TelescopeTracker;
+import org.vivecraft.provider.ControllerType;
+import org.vivecraft.render.RenderPass;
+import org.vivecraft.render.VRArmRenderer;
+import org.vivecraft.render.VRFirstPersonArmSwing;
+import org.vivecraft.render.VRRenderType;
+import org.vivecraft.render.VivecraftItemRendering;
 
 public class ItemInHandRenderer
 {
-    private static final RenderType MAP_BACKGROUND = RenderType.text(new ResourceLocation("textures/map/map_background.png"));
-    private static final RenderType MAP_BACKGROUND_CHECKERBOARD = RenderType.text(new ResourceLocation("textures/map/map_background_checkerboard.png"));
+    private static final RenderType MAP_BACKGROUND = VRRenderType.getTextNoCull(new ResourceLocation("textures/map/map_background.png"));
+    private static final RenderType MAP_BACKGROUND_CHECKERBOARD = VRRenderType.getTextNoCull(new ResourceLocation("textures/map/map_background_checkerboard.png"));
     private final Minecraft minecraft;
     private ItemStack mainHandItem = ItemStack.EMPTY;
     private ItemStack offHandItem = ItemStack.EMPTY;
@@ -42,12 +79,19 @@
     private float oOffHandHeight;
     private final EntityRenderDispatcher entityRenderDispatcher;
     private final ItemRenderer itemRenderer;
+    private final VRArmRenderer myRenderVRPlayer;
+    private final Map<String, VRArmRenderer> renderVRPlayerSkinMap = new HashMap<>();
+    public VRFirstPersonArmSwing swingType = VRFirstPersonArmSwing.Attack;
+    public float xdist = 0.0F;
 
     public ItemInHandRenderer(Minecraft p_109310_)
     {
         this.minecraft = p_109310_;
         this.entityRenderDispatcher = p_109310_.getEntityRenderDispatcher();
         this.itemRenderer = p_109310_.getItemRenderer();
+        this.myRenderVRPlayer = new VRArmRenderer(this.entityRenderDispatcher);
+        this.renderVRPlayerSkinMap.put("default", this.myRenderVRPlayer);
+        this.renderVRPlayerSkinMap.put("slim", new VRArmRenderer(this.entityRenderDispatcher, true));
     }
 
     public void renderItem(LivingEntity p_109323_, ItemStack p_109324_, ItemTransforms.TransformType p_109325_, boolean p_109326_, PoseStack p_109327_, MultiBufferSource p_109328_, int p_109329_)
@@ -169,33 +213,37 @@
     {
         boolean flag = p_109352_ != HumanoidArm.LEFT;
         float f = flag ? 1.0F : -1.0F;
-        float f1 = Mth.sqrt(p_109351_);
-        float f2 = -0.3F * Mth.sin(f1 * (float)Math.PI);
-        float f3 = 0.4F * Mth.sin(f1 * ((float)Math.PI * 2F));
-        float f4 = -0.4F * Mth.sin(p_109351_ * (float)Math.PI);
-        p_109347_.translate((double)(f * (f2 + 0.64000005F)), (double)(f3 + -0.6F + p_109350_ * -0.6F), (double)(f4 + -0.71999997F));
-        p_109347_.mulPose(Vector3f.YP.rotationDegrees(f * 45.0F));
-        float f5 = Mth.sin(p_109351_ * p_109351_ * (float)Math.PI);
-        float f6 = Mth.sin(f1 * (float)Math.PI);
-        p_109347_.mulPose(Vector3f.YP.rotationDegrees(f * f6 * 70.0F));
-        p_109347_.mulPose(Vector3f.ZP.rotationDegrees(f * f5 * -20.0F));
         AbstractClientPlayer abstractclientplayer = this.minecraft.player;
         this.minecraft.getTextureManager().bind(abstractclientplayer.getSkinTextureLocation());
-        p_109347_.translate((double)(f * -1.0F), (double)3.6F, 3.5D);
-        p_109347_.mulPose(Vector3f.ZP.rotationDegrees(f * 120.0F));
-        p_109347_.mulPose(Vector3f.XP.rotationDegrees(200.0F));
-        p_109347_.mulPose(Vector3f.YP.rotationDegrees(f * -135.0F));
-        p_109347_.translate((double)(f * 5.6F), 0.0D, 0.0D);
-        PlayerRenderer playerrenderer = (PlayerRenderer)this.entityRenderDispatcher.<AbstractClientPlayer>getRenderer(abstractclientplayer);
+        VRArmRenderer vrarmrenderer = this.renderVRPlayerSkinMap.get(abstractclientplayer.getModelName());
+        p_109347_.pushPose();
+
+        if (abstractclientplayer.swingingArm == InteractionHand.MAIN_HAND && flag)
+        {
+            this.transformFirstPersonVR(p_109347_, p_109352_, p_109351_);
+        }
+
+        if (abstractclientplayer.swingingArm == InteractionHand.OFF_HAND && !flag)
+        {
+            this.transformFirstPersonVR(p_109347_, p_109352_, p_109351_);
+        }
+
+        p_109347_.scale(0.4F, 0.4F, 0.4F);
+        boolean flag1 = abstractclientplayer.getModelName().equals("slim");
+        p_109347_.translate(flag1 ? (double)(0.345F * -f) : (double)(0.375F * -f), 0.0D, flag1 ? (double)0.785F : 0.75D);
+        p_109347_.mulPose(Vector3f.XP.rotationDegrees(-90.0F));
+        p_109347_.mulPose(Vector3f.YP.rotationDegrees(180.0F));
 
         if (flag)
         {
-            playerrenderer.renderRightHand(p_109347_, p_109348_, p_109349_, abstractclientplayer);
+            vrarmrenderer.renderRightHand(p_109347_, p_109348_, p_109349_, abstractclientplayer);
         }
         else
         {
-            playerrenderer.renderLeftHand(p_109347_, p_109348_, p_109349_, abstractclientplayer);
+            vrarmrenderer.renderLeftHand(p_109347_, p_109348_, p_109349_, abstractclientplayer);
         }
+
+        p_109347_.popPose();
     }
 
     private void applyEatTransform(PoseStack p_109331_, float p_109332_, HumanoidArm p_109333_, ItemStack p_109334_)
@@ -228,6 +276,49 @@
         p_109336_.mulPose(Vector3f.YP.rotationDegrees((float)i * -45.0F));
     }
 
+    private void transformFirstPersonVR(PoseStack matrixStackIn, HumanoidArm hand, float swingProgress)
+    {
+        if (swingProgress != 0.0F)
+        {
+            switch (this.swingType)
+            {
+                case Attack:
+                    float f2 = Mth.sin((float)((double)(swingProgress * 3.0F) * Math.PI));
+
+                    if ((double)swingProgress > 0.5D)
+                    {
+                        f2 = Mth.sin((float)((double)swingProgress * Math.PI + Math.PI));
+                    }
+
+                    matrixStackIn.translate(0.0D, 0.0D, (double)0.2F);
+                    matrixStackIn.mulPose(Vector3f.XP.rotationDegrees(f2 * 30.0F));
+                    matrixStackIn.translate(0.0D, 0.0D, (double) - 0.2F);
+                    break;
+
+                case Interact:
+                    float f1 = Mth.sin((float)((double)(swingProgress * 3.0F) * Math.PI));
+
+                    if ((double)swingProgress > 0.5D)
+                    {
+                        f1 = Mth.sin((float)((double)swingProgress * Math.PI + Math.PI));
+                    }
+
+                    matrixStackIn.mulPose(Vector3f.ZP.rotationDegrees((float)(hand == HumanoidArm.RIGHT ? -1 : 1) * f1 * 45.0F));
+                    break;
+
+                case Use:
+                    float f = Mth.sin((float)((double)(swingProgress * 2.0F) * Math.PI));
+
+                    if ((double)swingProgress > 0.25D)
+                    {
+                        f = Mth.sin((float)((double)(swingProgress / 2.0F) * Math.PI + Math.PI));
+                    }
+
+                    matrixStackIn.translate(0.0D, 0.0D, (double)(-(1.0F + f) * 0.1F));
+            }
+        }
+    }
+
     private void applyItemArmTransform(PoseStack p_109383_, HumanoidArm p_109384_, float p_109385_)
     {
         int i = p_109384_ == HumanoidArm.RIGHT ? 1 : -1;
@@ -311,182 +402,478 @@
         p_109317_.endBatch();
     }
 
-    private void renderArmWithItem(AbstractClientPlayer p_109372_, float p_109373_, float p_109374_, InteractionHand p_109375_, float p_109376_, ItemStack p_109377_, float p_109378_, PoseStack p_109379_, MultiBufferSource p_109380_, int p_109381_)
+    public void renderArmWithItem(AbstractClientPlayer p_109372_, float p_109373_, float p_109374_, InteractionHand p_109375_, float p_109376_, ItemStack p_109377_, float p_109378_, PoseStack p_109379_, MultiBufferSource p_109380_, int p_109381_)
     {
         if (!Config.isShaders() || !Shaders.isSkipRenderHand(p_109375_))
         {
             boolean flag = p_109375_ == InteractionHand.MAIN_HAND;
             HumanoidArm humanoidarm = flag ? p_109372_.getMainArm() : p_109372_.getMainArm().getOpposite();
+            p_109378_ = this.getEquipProgress(p_109375_, p_109373_);
             p_109379_.pushPose();
+            boolean flag1 = true;
 
-            if (p_109377_.isEmpty())
+            if (this.minecraft.currentPass == RenderPass.THIRD && !this.minecraft.vrSettings.mixedRealityRenderHands)
             {
-                if (flag && !p_109372_.isInvisible())
-                {
-                    this.renderPlayerArm(p_109379_, p_109380_, p_109381_, p_109378_, p_109376_, humanoidarm);
-                }
+                flag1 = false;
             }
-            else if (p_109377_.getItem() instanceof MapItem)
+
+            if (this.minecraft.currentPass == RenderPass.CAMERA)
             {
-                if (flag && this.offHandItem.isEmpty())
-                {
-                    this.renderTwoHandedMap(p_109379_, p_109380_, p_109381_, p_109374_, p_109378_, p_109376_);
-                }
-                else
-                {
-                    this.renderOneHandedMap(p_109379_, p_109380_, p_109381_, p_109378_, humanoidarm, p_109376_, p_109377_);
-                }
+                flag1 = false;
             }
-            else if (p_109377_.getItem() instanceof CrossbowItem)
+
+            if (BowTracker.isBow(p_109377_) && this.minecraft.bowTracker.isActive((LocalPlayer)p_109372_))
+            {
+                flag1 = false;
+            }
+
+            if (TelescopeTracker.isTelescope(p_109377_) && (p_109375_ == InteractionHand.OFF_HAND && this.minecraft.currentPass == RenderPass.SCOPEL || p_109375_ == InteractionHand.MAIN_HAND && this.minecraft.currentPass == RenderPass.SCOPER))
+            {
+                flag1 = false;
+            }
+
+            if (flag1 && !p_109372_.isInvisible())
+            {
+                this.renderPlayerArm(p_109379_, p_109380_, p_109381_, p_109378_, p_109376_, humanoidarm);
+            }
+
+            if (!p_109377_.isEmpty())
             {
-                boolean flag1 = CrossbowItem.isCharged(p_109377_);
-                boolean flag2 = humanoidarm == HumanoidArm.RIGHT;
-                int i = flag2 ? 1 : -1;
+                Item item = p_109377_.getItem();
+                boolean flag2 = false;
+                VivecraftItemRendering vivecraftitemrendering = VivecraftItemRendering.Item;
 
-                if (p_109372_.isUsingItem() && p_109372_.getUseItemRemainingTicks() > 0 && p_109372_.getUsedItemHand() == p_109375_)
+                if (p_109377_.getUseAnimation() != UseAnim.EAT && p_109377_.getUseAnimation() != UseAnim.DRINK)
                 {
-                    this.applyItemArmTransform(p_109379_, humanoidarm, p_109378_);
-                    p_109379_.translate((double)((float)i * -0.4785682F), (double) - 0.094387F, (double)0.05731531F);
-                    p_109379_.mulPose(Vector3f.XP.rotationDegrees(-11.935F));
-                    p_109379_.mulPose(Vector3f.YP.rotationDegrees((float)i * 65.3F));
-                    p_109379_.mulPose(Vector3f.ZP.rotationDegrees((float)i * -9.785F));
-                    float f9 = (float)p_109377_.getUseDuration() - ((float)this.minecraft.player.getUseItemRemainingTicks() - p_109373_ + 1.0F);
-                    float f12 = f9 / (float)CrossbowItem.getChargeDuration(p_109377_);
+                    if (item instanceof BlockItem)
+                    {
+                        Block block = ((BlockItem)item).getBlock();
+
+                        if (block instanceof TorchBlock)
+                        {
+                            vivecraftitemrendering = VivecraftItemRendering.Block_Stick;
+                        }
+                        else
+                        {
+                            BakedModel bakedmodel = this.itemRenderer.getModel(p_109377_, this.minecraft.level, this.minecraft.player);
 
-                    if (f12 > 1.0F)
+                            if (bakedmodel.isGui3d())
+                            {
+                                vivecraftitemrendering = VivecraftItemRendering.Block_3D;
+                            }
+                            else
+                            {
+                                vivecraftitemrendering = VivecraftItemRendering.Block_Item;
+                            }
+                        }
+                    }
+                    else if (item instanceof MapItem)
                     {
-                        f12 = 1.0F;
+                        vivecraftitemrendering = VivecraftItemRendering.Map;
                     }
+                    else if (p_109377_.getUseAnimation() == UseAnim.BOW)
+                    {
+                        vivecraftitemrendering = VivecraftItemRendering.Bow_Seated;
 
-                    if (f12 > 0.1F)
+                        if (this.minecraft.bowTracker.isActive((LocalPlayer)p_109372_))
+                        {
+                            if (this.minecraft.bowTracker.isDrawing)
+                            {
+                                vivecraftitemrendering = VivecraftItemRendering.Bow_Roomscale_Drawing;
+                            }
+                            else
+                            {
+                                vivecraftitemrendering = VivecraftItemRendering.Bow_Roomscale;
+                            }
+                        }
+                    }
+                    else if (item instanceof SwordItem)
+                    {
+                        vivecraftitemrendering = VivecraftItemRendering.Sword;
+                    }
+                    else if (item instanceof ShieldItem)
+                    {
+                        vivecraftitemrendering = VivecraftItemRendering.Shield;
+                    }
+                    else if (item instanceof TridentItem)
                     {
-                        float f15 = Mth.sin((f9 - 0.1F) * 1.3F);
-                        float f3 = f12 - 0.1F;
-                        float f4 = f15 * f3;
-                        p_109379_.translate((double)(f4 * 0.0F), (double)(f4 * 0.004F), (double)(f4 * 0.0F));
+                        vivecraftitemrendering = VivecraftItemRendering.Spear;
                     }
+                    else if (item instanceof CrossbowItem)
+                    {
+                        vivecraftitemrendering = VivecraftItemRendering.Crossbow;
+                    }
+                    else if (!(item instanceof CompassItem) && item != Items.CLOCK)
+                    {
+                        if (SwingTracker.isTool(item))
+                        {
+                            vivecraftitemrendering = VivecraftItemRendering.Tool;
 
-                    p_109379_.translate((double)(f12 * 0.0F), (double)(f12 * 0.0F), (double)(f12 * 0.04F));
-                    p_109379_.scale(1.0F, 1.0F, 1.0F + f12 * 0.2F);
-                    p_109379_.mulPose(Vector3f.YN.rotationDegrees((float)i * 45.0F));
+                            if (item instanceof FoodOnAStickItem || item instanceof FishingRodItem)
+                            {
+                                vivecraftitemrendering = VivecraftItemRendering.Tool_Rod;
+                            }
+                        }
+                        else if (TelescopeTracker.isTelescope(p_109377_))
+                        {
+                            vivecraftitemrendering = VivecraftItemRendering.Telescope;
+                        }
+                    }
+                    else
+                    {
+                        vivecraftitemrendering = VivecraftItemRendering.Compass;
+                    }
                 }
                 else
                 {
-                    float f = -0.4F * Mth.sin(Mth.sqrt(p_109376_) * (float)Math.PI);
-                    float f1 = 0.2F * Mth.sin(Mth.sqrt(p_109376_) * ((float)Math.PI * 2F));
-                    float f2 = -0.2F * Mth.sin(p_109376_ * (float)Math.PI);
-                    p_109379_.translate((double)((float)i * f), (double)f1, (double)f2);
-                    this.applyItemArmTransform(p_109379_, humanoidarm, p_109378_);
-                    this.applyItemArmAttackTransform(p_109379_, humanoidarm, p_109376_);
+                    vivecraftitemrendering = VivecraftItemRendering.Noms;
+                }
 
-                    if (flag1 && p_109376_ < 0.001F)
+                int k = flag ? 1 : -1;
+                double d8 = 0.7D;
+                double d0 = -0.05D;
+                double d1 = 0.005D;
+                double d2 = 0.0D;
+                double d3 = this.minecraft.vr.getGunAngle();
+                Quaternion quaternion = Vector3f.YP.rotationDegrees(0.0F);
+                Quaternion quaternion1 = Vector3f.YP.rotationDegrees(0.0F);
+                quaternion.mul(Vector3f.XP.rotationDegrees((float)(-110.0D + d3)));
+                p_109379_.pushPose();
+                boolean flag3 = false;
+
+                if (vivecraftitemrendering == VivecraftItemRendering.Bow_Seated)
+                {
+                    d1 += -0.1D;
+                    d2 += 0.1D;
+                    quaternion.mul(Vector3f.XP.rotationDegrees((float)(90.0D - d3)));
+                    d8 = (double)0.7F;
+                }
+                else if (vivecraftitemrendering == VivecraftItemRendering.Bow_Roomscale)
+                {
+                    quaternion = Vector3f.XP.rotationDegrees(0.0F);
+                    p_109379_.mulPose(Vector3f.XP.rotationDegrees((float)(-110.0D + d3)));
+                    d1 -= 0.25D;
+                    d2 += (double)0.025F + 0.03D * d3 / 40.0D;
+                    d0 += -0.0225D;
+                    d8 = 1.0D;
+                }
+                else if (vivecraftitemrendering == VivecraftItemRendering.Bow_Roomscale_Drawing)
+                {
+                    quaternion = Vector3f.YP.rotationDegrees(0.0F);
+                    d8 = 1.0D;
+                    int i = 0;
+
+                    if (this.minecraft.vrSettings.vrReverseShootingEye)
                     {
-                        p_109379_.translate((double)((float)i * -0.641864F), 0.0D, 0.0D);
-                        p_109379_.mulPose(Vector3f.YP.rotationDegrees((float)i * 10.0F));
+                        i = 1;
                     }
-                }
 
-                this.renderItem(p_109372_, p_109377_, flag2 ? ItemTransforms.TransformType.FIRST_PERSON_RIGHT_HAND : ItemTransforms.TransformType.FIRST_PERSON_LEFT_HAND, !flag2, p_109379_, p_109380_, p_109381_);
-            }
-            else
-            {
-                boolean flag3 = humanoidarm == HumanoidArm.RIGHT;
+                    Vec3 vec3 = this.minecraft.bowTracker.getAimVector();
+                    Vec3 vec31 = new Vec3(vec3.x, vec3.y, vec3.z);
+                    Vec3 vec32 = this.minecraft.vrPlayer.vrdata_world_render.getHand(1).getCustomVector(new Vec3(0.0D, -1.0D, 0.0D));
+                    Vec3 vec33 = this.minecraft.vrPlayer.vrdata_world_render.getHand(1).getCustomVector(new Vec3(0.0D, 0.0D, -1.0D));
+                    vec31.cross(vec32);
+                    double d4 = (180D / Math.PI) * Math.acos(vec31.dot(vec32));
+                    float f = (float)Math.toDegrees(Math.asin(vec31.y / vec31.length()));
+                    float f1 = (float)Math.toDegrees(Math.atan2(vec31.x, vec31.z));
+                    Vec3 vec34 = new Vec3(0.0D, 1.0D, 0.0D);
+                    Vec3 vec35 = new Vec3(vec31.x, 0.0D, vec31.z);
+                    Vec3 vec36 = Vec3.ZERO;
+                    double d5 = vec33.dot(vec35);
 
-                if (p_109372_.isUsingItem() && p_109372_.getUseItemRemainingTicks() > 0 && p_109372_.getUsedItemHand() == p_109375_)
-                {
-                    int k = flag3 ? 1 : -1;
+                    if (d5 != 0.0D)
+                    {
+                        vec36 = vec35.scale(d5);
+                    }
 
-                    switch (p_109377_.getUseAnimation())
+                    double d6 = 0.0D;
+                    Vec3 vec37 = vec33.subtract(vec36).normalize();
+                    d6 = vec37.dot(vec34);
+                    double d7 = vec35.dot(vec37.cross(vec34));
+                    float f2;
+
+                    if (d7 < 0.0D)
+                    {
+                        f2 = -((float)Math.acos(d6));
+                    }
+                    else
                     {
-                        case NONE:
-                            this.applyItemArmTransform(p_109379_, humanoidarm, p_109378_);
-                            break;
+                        f2 = (float)Math.acos(d6);
+                    }
 
-                        case EAT:
-                        case DRINK:
-                            this.applyEatTransform(p_109379_, p_109373_, humanoidarm, p_109377_);
-                            this.applyItemArmTransform(p_109379_, humanoidarm, p_109378_);
-                            break;
+                    float f3 = (float)((180D / Math.PI) * (double)f2);
 
-                        case BLOCK:
-                            this.applyItemArmTransform(p_109379_, humanoidarm, p_109378_);
-                            break;
+                    if (this.minecraft.bowTracker.isCharged())
+                    {
+                        long j = Util.getMillis() - this.minecraft.bowTracker.startDrawTime;
+                        d0 += 0.003D * Math.sin((double)j);
+                    }
 
-                        case BOW:
-                            this.applyItemArmTransform(p_109379_, humanoidarm, p_109378_);
-                            p_109379_.translate((double)((float)k * -0.2785682F), (double)0.18344387F, (double)0.15731531F);
-                            p_109379_.mulPose(Vector3f.XP.rotationDegrees(-13.935F));
-                            p_109379_.mulPose(Vector3f.YP.rotationDegrees((float)k * 35.3F));
-                            p_109379_.mulPose(Vector3f.ZP.rotationDegrees((float)k * -9.785F));
-                            float f8 = (float)p_109377_.getUseDuration() - ((float)this.minecraft.player.getUseItemRemainingTicks() - p_109373_ + 1.0F);
-                            float f11 = f8 / 20.0F;
-                            f11 = (f11 * f11 + f11 * 2.0F) / 3.0F;
+                    p_109379_.translate(0.0D, 0.0D, 0.1D);
+                    p_109379_.last().pose().multiply(this.minecraft.vrPlayer.vrdata_world_render.getController(1).getMatrix().transposed().toMCMatrix());
+                    quaternion.mul(Vector3f.YP.rotationDegrees(f1));
+                    quaternion.mul(Vector3f.XP.rotationDegrees(-f));
+                    quaternion.mul(Vector3f.ZP.rotationDegrees(-f3));
+                    quaternion.mul(Vector3f.ZP.rotationDegrees(180.0F));
+                    p_109379_.last().pose().multiply(quaternion);
+                    quaternion = Vector3f.YP.rotationDegrees(0.0F);
+                    quaternion.mul(Vector3f.YP.rotationDegrees(180.0F));
+                    quaternion.mul(Vector3f.XP.rotationDegrees(160.0F));
+                    d1 += 0.1225D;
+                    d0 += 0.125D;
+                    d2 += 0.16D;
+                }
+                else if (vivecraftitemrendering == VivecraftItemRendering.Crossbow)
+                {
+                    d0 += (double)0.01F;
+                    d2 += (double) - 0.02F;
+                    d1 += (double) - 0.02F;
+                    d8 = 0.5D;
+                    quaternion = Vector3f.XP.rotationDegrees(0.0F);
+                    quaternion.mul(Vector3f.YP.rotationDegrees(10.0F));
+                }
+                else if (vivecraftitemrendering == VivecraftItemRendering.Map)
+                {
+                    flag2 = true;
+                    quaternion = Vector3f.XP.rotationDegrees(-45.0F);
+                    d0 = 0.0D;
+                    d1 = 0.16D;
+                    d2 = -0.075D;
+                    d8 = 0.75D;
+                }
+                else if (vivecraftitemrendering == VivecraftItemRendering.Noms)
+                {
+                    long l = (long)this.minecraft.player.getUseItemRemainingTicks();
+                    quaternion = Vector3f.ZP.rotationDegrees(180.0F);
+                    quaternion.mul(Vector3f.XP.rotationDegrees(-135.0F));
+                    d2 = d2 + 0.006D * Math.sin((double)l);
+                    d2 = d2 + (double)0.02F;
+                    d0 += (double)0.08F;
+                    d8 = (double)0.4F;
+                }
+                else if (vivecraftitemrendering != VivecraftItemRendering.Item && vivecraftitemrendering != VivecraftItemRendering.Block_Item)
+                {
+                    if (vivecraftitemrendering == VivecraftItemRendering.Compass)
+                    {
+                        quaternion = Vector3f.YP.rotationDegrees(90.0F);
+                        quaternion.mul(Vector3f.XP.rotationDegrees(25.0F));
+                        d8 = (double)0.4F;
+                    }
+                    else if (vivecraftitemrendering == VivecraftItemRendering.Block_3D)
+                    {
+                        d8 = (double)0.3F;
+                        d2 += (double) - 0.1F;
+                        d0 += (double)0.05F;
+                    }
+                    else if (vivecraftitemrendering == VivecraftItemRendering.Block_Stick)
+                    {
+                        quaternion = Vector3f.XP.rotationDegrees(0.0F);
+                        d1 += -0.105D + 0.06D * d3 / 40.0D;
+                        d2 += (double) - 0.1F;
+                        quaternion.mul(Vector3f.XP.rotationDegrees(-45.0F));
+                        quaternion.mul(Vector3f.XP.rotationDegrees((float)d3));
+                    }
+                    else if (vivecraftitemrendering == VivecraftItemRendering.Shield)
+                    {
+                        flag3 = !flag;
+                        d8 = (double)0.4F;
+                        d1 += (double)0.21F;
+
+                        if (flag)
+                        {
+                            d0 += (double)0.11F;
+                        }
+                        else
+                        {
+                            d0 += -0.015D;
+                        }
+
+                        d2 += (double) - 0.01F;
+                        quaternion.mul(Vector3f.XP.rotationDegrees((float)(105.0D - d3)));
+
+                        if (p_109372_.isUsingItem() && p_109372_.getUseItemRemainingTicks() > 0 && p_109372_.getUsedItemHand() == p_109375_)
+                        {
+                            quaternion.mul(Vector3f.ZP.rotationDegrees((float)(k * -5)));
+                            d1 += (double) - 0.13F;
+                            d0 += (double)((float)k * 0.05F);
+                            d2 += (double) - 0.1F;
 
-                            if (f11 > 1.0F)
+                            if (p_109372_.isBlocking())
                             {
-                                f11 = 1.0F;
+                                quaternion.mul(Vector3f.YP.rotationDegrees((float)k * 90.0F));
                             }
-
-                            if (f11 > 0.1F)
+                            else
                             {
-                                float f14 = Mth.sin((f8 - 0.1F) * 1.3F);
-                                float f17 = f11 - 0.1F;
-                                float f19 = f14 * f17;
-                                p_109379_.translate((double)(f19 * 0.0F), (double)(f19 * 0.004F), (double)(f19 * 0.0F));
+                                quaternion.mul(Vector3f.YP.rotationDegrees((1.0F - p_109378_) * (float)k * 90.0F));
                             }
+                        }
 
-                            p_109379_.translate((double)(f11 * 0.0F), (double)(f11 * 0.0F), (double)(f11 * 0.04F));
-                            p_109379_.scale(1.0F, 1.0F, 1.0F + f11 * 0.2F);
-                            p_109379_.mulPose(Vector3f.YN.rotationDegrees((float)k * 45.0F));
-                            break;
-
-                        case SPEAR:
-                            this.applyItemArmTransform(p_109379_, humanoidarm, p_109378_);
-                            p_109379_.translate((double)((float)k * -0.5F), (double)0.7F, (double)0.1F);
-                            p_109379_.mulPose(Vector3f.XP.rotationDegrees(-55.0F));
-                            p_109379_.mulPose(Vector3f.YP.rotationDegrees((float)k * 35.3F));
-                            p_109379_.mulPose(Vector3f.ZP.rotationDegrees((float)k * -9.785F));
-                            float f13 = (float)p_109377_.getUseDuration() - ((float)this.minecraft.player.getUseItemRemainingTicks() - p_109373_ + 1.0F);
-                            float f16 = f13 / 10.0F;
+                        quaternion.mul(Vector3f.YP.rotationDegrees((float)k * -90.0F));
+                    }
+                    else if (vivecraftitemrendering == VivecraftItemRendering.Spear)
+                    {
+                        quaternion = Vector3f.XP.rotationDegrees(0.0F);
+                        d0 += (double) - 0.135F;
+                        d2 = d2 + (double)0.575F;
+                        d8 = (double)0.6F;
+                        float f4 = 0.0F;
+                        boolean flag5 = false;
+                        int i1 = 0;
+
+                        if (p_109372_.isUsingItem() && p_109372_.getUseItemRemainingTicks() > 0 && p_109372_.getUsedItemHand() == p_109375_)
+                        {
+                            flag5 = true;
+                            i1 = EnchantmentHelper.getRiptide(p_109377_);
 
-                            if (f16 > 1.0F)
+                            if (i1 <= 0 || i1 > 0 && p_109372_.isInWaterOrRain())
                             {
-                                f16 = 1.0F;
+                                f4 = (float)p_109377_.getUseDuration() - ((float)this.minecraft.player.getUseItemRemainingTicks() - p_109373_ + 1.0F);
+
+                                if (f4 > 10.0F)
+                                {
+                                    f4 = 10.0F;
+
+                                    if (i1 > 0 && p_109372_.isInWaterOrRain())
+                                    {
+                                        p_109379_.mulPose(Vector3f.ZP.rotationDegrees((float)(-this.minecraft.tickCounter * 10 * i1 % 360) - p_109373_ * 10.0F * (float)i1));
+                                    }
+
+                                    if (this.minecraft.frameIndex % 4L == 0L)
+                                    {
+                                        this.minecraft.vr.triggerHapticPulse(flag ? 0 : 1, 200);
+                                    }
+
+                                    long j1 = Util.getMillis() - this.minecraft.bowTracker.startDrawTime;
+                                    d0 += 0.003D * Math.sin((double)j1);
+                                }
                             }
+                        }
+
+                        if (p_109372_.isAutoSpinAttack())
+                        {
+                            i1 = 5;
+                            d2 += (double) - 0.15F;
+                            p_109379_.mulPose(Vector3f.ZP.rotationDegrees((float)(-this.minecraft.tickCounter * 10 * i1 % 360) - p_109373_ * 10.0F * (float)i1));
+                            flag5 = true;
+                        }
+
+                        if (!flag5)
+                        {
+                            d1 += 0.0D + 0.2D * d3 / 40.0D;
+                            quaternion.mul(Vector3f.XP.rotationDegrees((float)d3));
+                        }
 
-                            if (f16 > 0.1F)
+                        quaternion.mul(Vector3f.XP.rotationDegrees(-65.0F));
+                        d2 = d2 + (double)(-0.75F + f4 / 10.0F * 0.25F);
+                    }
+                    else if (vivecraftitemrendering != VivecraftItemRendering.Sword)
+                    {
+                        if (vivecraftitemrendering == VivecraftItemRendering.Tool_Rod)
+                        {
+                            d2 += (double) - 0.15F;
+                            d1 += -0.02D + d3 / 40.0D * 0.1D;
+                            d0 += (double)0.05F;
+                            quaternion.mul(Vector3f.XP.rotationDegrees(40.0F));
+                            d8 = (double)0.8F;
+                        }
+                        else if (vivecraftitemrendering == VivecraftItemRendering.Tool)
+                        {
+                            boolean flag4 = this.minecraft.climbTracker.isClaws(p_109377_) && this.minecraft.climbTracker.isClimbeyClimb();
+
+                            if (flag4)
                             {
-                                float f18 = Mth.sin((f13 - 0.1F) * 1.3F);
-                                float f20 = f16 - 0.1F;
-                                float f5 = f18 * f20;
-                                p_109379_.translate((double)(f5 * 0.0F), (double)(f5 * 0.004F), (double)(f5 * 0.0F));
+                                quaternion.mul(Vector3f.XP.rotationDegrees((float)(-d3)));
+                                d8 = (double)0.3F;
+                                d2 += (double)0.075F;
+                                d1 += (double)0.02F;
+                                d0 += (double)0.03F;
+
+                                if (this.minecraft.vr.keyClimbeyGrab.isKeyDown(ControllerType.RIGHT) && flag || this.minecraft.vr.keyClimbeyGrab.isKeyDown(ControllerType.LEFT) && !flag)
+                                {
+                                    d2 += (double) - 0.2F;
+                                }
                             }
 
-                            p_109379_.translate(0.0D, 0.0D, (double)(f16 * 0.2F));
-                            p_109379_.scale(1.0F, 1.0F, 1.0F + f16 * 0.2F);
-                            p_109379_.mulPose(Vector3f.YN.rotationDegrees((float)k * 45.0F));
+                            if (item instanceof ArrowItem)
+                            {
+                                quaternion1 = Vector3f.ZP.rotationDegrees(-180.0F);
+                                quaternion.mul(Vector3f.XP.rotationDegrees((float)(-d3)));
+                            }
+                        }
+                        else if (vivecraftitemrendering == VivecraftItemRendering.Telescope)
+                        {
+                            quaternion1 = Vector3f.XP.rotationDegrees(0.0F);
+                            quaternion = Vector3f.XP.rotationDegrees(0.0F);
+                            d2 = 0.0D;
+                            d1 = 0.0D;
+                            d0 = 0.0D;
+                        }
                     }
                 }
-                else if (p_109372_.isAutoSpinAttack())
-                {
-                    this.applyItemArmTransform(p_109379_, humanoidarm, p_109378_);
-                    int j = flag3 ? 1 : -1;
-                    p_109379_.translate((double)((float)j * -0.4F), (double)0.8F, (double)0.3F);
-                    p_109379_.mulPose(Vector3f.YP.rotationDegrees((float)j * 65.0F));
-                    p_109379_.mulPose(Vector3f.ZP.rotationDegrees((float)j * -85.0F));
+                else
+                {
+                    quaternion = Vector3f.ZP.rotationDegrees(180.0F);
+                    quaternion.mul(Vector3f.XP.rotationDegrees(-135.0F));
+                    d8 = (double)0.4F;
+                    d0 += (double)0.08F;
+                    d2 += (double) - 0.08F;
+                }
+
+                if (p_109372_.swingingArm == p_109375_)
+                {
+                    this.transformFirstPersonVR(p_109379_, humanoidarm, p_109376_);
+                }
+
+                ItemRenderer itemrenderer = this.itemRenderer;
+                ItemRenderer.isfphand = true;
+                ItemTransforms.TransformType itemtransforms$transformtype = flag ? ItemTransforms.TransformType.FIRST_PERSON_RIGHT_HAND : (flag3 ? ItemTransforms.TransformType.FIRST_PERSON_LEFT_HAND : ItemTransforms.TransformType.FIRST_PERSON_RIGHT_HAND);
+
+                if (p_109377_.hasTag() && p_109377_.getTag().getInt("CustomModelData") > 0)
+                {
+                    itemtransforms$transformtype = ItemTransforms.TransformType.THIRD_PERSON_RIGHT_HAND;
+                    p_109379_.scale(0.7F, 0.7F, 0.7F);
+                }
+                else
+                {
+                    p_109379_.mulPose(quaternion1);
+                    p_109379_.translate(d0, d1, d2);
+                    p_109379_.mulPose(quaternion);
+                    p_109379_.scale((float)d8, (float)d8, (float)d8);
+                }
+
+                if (flag2)
+                {
+                    RenderSystem.disableCull();
+                    this.renderMap(p_109379_, p_109380_, p_109381_, p_109377_);
+                }
+                else if (vivecraftitemrendering == VivecraftItemRendering.Telescope)
+                {
+                    if (this.minecraft.currentPass != RenderPass.SCOPEL && this.minecraft.currentPass != RenderPass.SCOPER)
+                    {
+                        p_109379_.pushPose();
+                        p_109379_.translate((double) - 0.1325F, 0.25D, (double)0.08F);
+                        p_109379_.mulPose(Vector3f.XP.rotationDegrees(180.0F));
+                        p_109379_.scale(0.25F, 1.5F, 0.25F);
+                        this.minecraft.getBlockRenderer().getModelRenderer().renderModel(p_109379_.last(), p_109380_.getBuffer(Sheets.solidBlockSheet()), (BlockState)null, this.minecraft.getModelManager().getModel(TelescopeTracker.scopeModel), 0.5F, 0.5F, 1.0F, p_109381_, OverlayTexture.NO_OVERLAY);
+                        p_109379_.popPose();
+                    }
+
+                    p_109379_.pushPose();
+                    p_109379_.translate(0.0D, (double)0.2501F, (double) - 0.053F);
+                    p_109379_.mulPose(Vector3f.XP.rotationDegrees(90.0F));
+                    p_109379_.mulPose(Vector3f.YP.rotationDegrees(180.0F));
+                    p_109379_.mulPose(Vector3f.ZP.rotationDegrees(180.0F));
+                    this.minecraft.gameRenderer.DrawScopeFB(p_109379_, p_109375_ == InteractionHand.MAIN_HAND ? 0 : 1);
+                    p_109379_.popPose();
                 }
                 else
                 {
-                    float f6 = -0.4F * Mth.sin(Mth.sqrt(p_109376_) * (float)Math.PI);
-                    float f7 = 0.2F * Mth.sin(Mth.sqrt(p_109376_) * ((float)Math.PI * 2F));
-                    float f10 = -0.2F * Mth.sin(p_109376_ * (float)Math.PI);
-                    int l = flag3 ? 1 : -1;
-                    p_109379_.translate((double)((float)l * f6), (double)f7, (double)f10);
-                    this.applyItemArmTransform(p_109379_, humanoidarm, p_109378_);
-                    this.applyItemArmAttackTransform(p_109379_, humanoidarm, p_109376_);
+                    this.renderItem(p_109372_, p_109377_, itemtransforms$transformtype, flag3, p_109379_, p_109380_, p_109381_);
                 }
 
-                this.renderItem(p_109372_, p_109377_, flag3 ? ItemTransforms.TransformType.FIRST_PERSON_RIGHT_HAND : ItemTransforms.TransformType.FIRST_PERSON_LEFT_HAND, !flag3, p_109379_, p_109380_, p_109381_);
+                itemrenderer = this.itemRenderer;
+                ItemRenderer.isfphand = false;
+                p_109379_.popPose();
             }
 
             p_109379_.popPose();
@@ -572,4 +959,41 @@
             this.offHandHeight = 0.0F;
         }
     }
+
+    public float getEquipProgress(InteractionHand hand, float partialTicks)
+    {
+        return hand == InteractionHand.MAIN_HAND ? 1.0F - (this.oMainHandHeight + (this.mainHandHeight - this.oMainHandHeight) * partialTicks) : 1.0F - (this.oOffHandHeight + (this.offHandHeight - this.oOffHandHeight) * partialTicks);
+    }
+
+    public boolean isInsideOpaqueBlock(Vec3 in)
+    {
+        if (this.minecraft.level == null)
+        {
+            return false;
+        }
+        else
+        {
+            BlockPos blockpos = new BlockPos(in);
+            return this.minecraft.level.getBlockState(blockpos).isSolidRender(this.minecraft.level, blockpos);
+        }
+    }
+
+    @Nullable
+    public Triple<Float, BlockState, BlockPos> getNearOpaqueBlock(Vec3 in, double dist)
+    {
+        if (this.minecraft.level == null)
+        {
+            return null;
+        }
+        else
+        {
+            AABB aabb = new AABB(in.subtract(dist, dist, dist), in.add(dist, dist, dist));
+            Stream<BlockPos> stream = BlockPos.betweenClosedStream(aabb).filter((bp) ->
+            {
+                return this.minecraft.level.getBlockState(bp).isViewBlocking(this.minecraft.level, bp);
+            });
+            Optional<BlockPos> optional = stream.findFirst();
+            return optional.isPresent() ? Triple.of(1.0F, this.minecraft.level.getBlockState(optional.get()), optional.get()) : null;
+        }
+    }
 }
