--- a/net/minecraft/client/renderer/LevelRenderer.java
+++ b/net/minecraft/client/renderer/LevelRenderer.java
@@ -163,7 +163,8 @@
 import net.optifine.util.RenderChunkUtils;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
-import org.lwjgl.opengl.GL11;
+import org.vivecraft.gameplay.screenhandlers.KeyboardHandler;
+import org.vivecraft.render.RenderPass;
 
 public class LevelRenderer implements ResourceManagerReloadListener, AutoCloseable
 {
@@ -287,6 +288,15 @@
     private static boolean ambientOcclusion = false;
     private Map<String, List<Entity>> mapEntityLists = new HashMap<>();
     private Map<RenderType, Map> mapRegionLayers = new LinkedHashMap<>();
+    @Nullable
+    public RenderTarget alphaSortVROccludedFramebuffer;
+    @Nullable
+    public RenderTarget alphaSortVRUnoccludedFramebuffer;
+    @Nullable
+    public RenderTarget alphaSortVRHandsFramebuffer;
+    public float selR;
+    public float selG;
+    public float selB;
 
     public LevelRenderer(Minecraft p_109480_, RenderBuffers p_109481_)
     {
@@ -337,9 +347,16 @@
 
             pLightmap.turnOnLightLayer();
             Level level = this.minecraft.level;
-            int i = Mth.floor(pX);
-            int j = Mth.floor(p_109707_);
-            int k = Mth.floor(pY);
+            Vec3 vec3 = this.minecraft.vrPlayer.vrdata_world_render.getEye(RenderPass.CENTER).getPosition();
+
+            if (this.minecraft.currentPass == RenderPass.THIRD || this.minecraft.currentPass == RenderPass.CAMERA)
+            {
+                vec3 = this.minecraft.vrPlayer.vrdata_world_render.getEye(this.minecraft.currentPass).getPosition();
+            }
+
+            int i = Mth.floor(vec3.x);
+            int j = Mth.floor(vec3.y);
+            int k = Mth.floor(vec3.z);
             Tesselator tesselator = Tesselator.getInstance();
             BufferBuilder bufferbuilder = tesselator.getBuilder();
             RenderSystem.disableCull();
@@ -563,7 +580,9 @@
     public void onResourceManagerReload(ResourceManager pResourceManager)
     {
         this.initOutline();
-
+        //Vivecraft
+        this.minecraft.vrRenderer.reinitFrameBuffers("Resource Reload");
+        //
         if (Minecraft.useShaderTransparency())
         {
             this.initTransparency();
@@ -599,10 +618,10 @@
         }
     }
 
-    private void initTransparency()
+    public void initTransparency()
     {
         this.deinitTransparency();
-        ResourceLocation resourcelocation = new ResourceLocation("shaders/post/transparency.json");
+        ResourceLocation resourcelocation = new ResourceLocation("shaders/post/vrtransparency.json");
 
         try
         {
@@ -620,11 +639,11 @@
             this.weatherTarget = rendertarget4;
             this.cloudsTarget = rendertarget;
         }
-        catch (Exception exception1)
+        catch (Exception exception)
         {
-            String s = exception1 instanceof JsonSyntaxException ? "parse" : "load";
+            String s = exception instanceof JsonSyntaxException ? "parse" : "load";
             String s1 = "Failed to " + s + " shader: " + resourcelocation;
-            LevelRenderer.TransparencyShaderException levelrenderer$transparencyshaderexception = new LevelRenderer.TransparencyShaderException(s1, exception1);
+            LevelRenderer.TransparencyShaderException levelrenderer$transparencyshaderexception = new LevelRenderer.TransparencyShaderException(s1, exception);
 
             if (this.minecraft.getResourcePackRepository().getSelectedIds().size() > 1)
             {
@@ -634,7 +653,7 @@
                 {
                     component = new TextComponent(this.minecraft.getResourceManager().getResource(resourcelocation).getSourceName());
                 }
-                catch (IOException ioexception1)
+                catch (IOException ioexception)
                 {
                     component = null;
                 }
@@ -654,7 +673,7 @@
         }
     }
 
-    private void deinitTransparency()
+    public void deinitTransparency()
     {
         if (this.transparencyChain != null)
         {
@@ -1030,7 +1049,7 @@
             this.viewArea.repositionCamera(d0, d2);
         }
 
-        if (Config.isDynamicLights())
+        if (Config.isDynamicLights() && this.minecraft.currentPass == RenderPass.LEFT)
         {
             DynamicLights.update(this);
         }
@@ -1044,7 +1063,7 @@
         BlockPos blockpos1 = new BlockPos(Mth.floor(vec3.x / 16.0D) * 16, Mth.floor(vec3.y / 16.0D) * 16, Mth.floor(vec3.z / 16.0D) * 16);
         float f = pActiveRenderInfo.getXRot();
         float f1 = pActiveRenderInfo.getYRot();
-        this.needsUpdate = this.needsUpdate || !this.chunksToCompile.isEmpty() || vec3.x != this.prevCamX || vec3.y != this.prevCamY || vec3.z != this.prevCamZ || (double)f != this.prevCamRotX || (double)f1 != this.prevCamRotY;
+        this.needsUpdate = true;
         this.prevCamX = vec3.x;
         this.prevCamY = vec3.y;
         this.prevCamZ = vec3.z;
@@ -1368,25 +1387,25 @@
     }
 
     @Nullable
-    private ChunkRenderDispatcher.RenderChunk getRenderChunkOffset(BlockPos playerPos, ChunkRenderDispatcher.RenderChunk renderChunkBase, Direction facing, boolean fog, int yMax)
+    private ChunkRenderDispatcher.RenderChunk getRenderChunkOffset(BlockPos p_getRenderChunkOffset_1_, ChunkRenderDispatcher.RenderChunk p_getRenderChunkOffset_2_, Direction p_getRenderChunkOffset_3_, boolean p_getRenderChunkOffset_4_, int p_getRenderChunkOffset_5_)
     {
-        ChunkRenderDispatcher.RenderChunk chunkrenderdispatcher$renderchunk = renderChunkBase.getRenderChunkNeighbour(facing);
+        ChunkRenderDispatcher.RenderChunk chunkrenderdispatcher$renderchunk = p_getRenderChunkOffset_2_.getRenderChunkNeighbour(p_getRenderChunkOffset_3_);
 
         if (chunkrenderdispatcher$renderchunk == null)
         {
             return null;
         }
-        else if (chunkrenderdispatcher$renderchunk.getOrigin().getY() > yMax)
+        else if (chunkrenderdispatcher$renderchunk.getOrigin().getY() > p_getRenderChunkOffset_5_)
         {
             return null;
         }
         else
         {
-            if (fog)
+            if (p_getRenderChunkOffset_4_)
             {
                 BlockPos blockpos = chunkrenderdispatcher$renderchunk.getOrigin();
-                int i = playerPos.getX() - blockpos.getX();
-                int j = playerPos.getZ() - blockpos.getZ();
+                int i = p_getRenderChunkOffset_1_.getX() - blockpos.getX();
+                int j = p_getRenderChunkOffset_1_.getZ() - blockpos.getZ();
                 int k = i * i + j * j;
 
                 if (k > this.renderDistanceSq)
@@ -1440,8 +1459,14 @@
         this.blockEntityRenderDispatcher.prepare(this.level, pDrawBlockOutline, this.minecraft.hitResult);
         this.entityRenderDispatcher.prepare(this.level, pDrawBlockOutline, this.minecraft.crosshairPickEntity);
         ProfilerFiller profilerfiller = this.level.getProfiler();
-        profilerfiller.popPush("light_updates");
-        this.minecraft.level.getChunkSource().getLightEngine().runUpdates(Integer.MAX_VALUE, true, true);
+
+        if (this.minecraft.currentPass == RenderPass.LEFT)
+        {
+            profilerfiller.popPush("light_updates");
+            this.minecraft.level.getChunkSource().getLightEngine().runUpdates(Integer.MAX_VALUE, true, true);
+        }
+
+        this.setShaderGroup();
         Vec3 vec3 = pDrawBlockOutline.getPosition();
         double d0 = vec3.x();
         double d1 = vec3.y();
@@ -1551,13 +1576,6 @@
         profilerfiller.popPush("terrain");
         Lagometer.timerTerrain.start();
 
-        if (this.minecraft.options.ofSmoothFps)
-        {
-            this.minecraft.getProfiler().popPush("finish");
-            GL11.glFinish();
-            this.minecraft.getProfiler().popPush("terrain");
-        }
-
         if (Config.isFogOff() && FogRenderer.fogStandard)
         {
             RenderSystem.setFogAllowed(false);
@@ -1573,16 +1591,21 @@
         }
 
         Lagometer.timerTerrain.end();
+        Matrix4f matrix4f1 = new Matrix4f();
+        matrix4f1.setIdentity();
 
         if (this.level.effects().constantAmbientLight())
         {
-            Lighting.setupNetherLevel(pMatrixStack.last().pose());
+            Lighting.setupNetherLevel(matrix4f1);
         }
         else
         {
-            Lighting.setupLevel(pMatrixStack.last().pose());
+            Lighting.setupLevel(matrix4f1);
         }
 
+        this.minecraft.getProfiler().popPush("stencil");
+        pActiveRenderInfo.drawEyeStencil(flag1);
+
         if (flag1)
         {
             Shaders.beginEntities();
@@ -1624,15 +1647,19 @@
         int i2 = this.level.getMinBuildHeight();
         int j2 = this.level.getMaxBuildHeight();
 
+        boolean renderSelf = this.minecraft.currentPass == RenderPass.THIRD && this.minecraft.vrSettings.displayMirrorMode == 14 || this.minecraft.currentPass == RenderPass.CAMERA;
+        renderSelf = renderSelf | (this.minecraft.vrSettings.shouldRenderSelf || this.minecraft.vrSettings.tmpRenderSelf);
+
+        
         for (Entity entity : this.level.entitiesForRendering())
         {
             BlockPos blockpos = entity.blockPosition();
 
             if (this.renderInfosEntities.contains(SectionPos.asLong(blockpos)) || blockpos.getY() <= i2 || blockpos.getY() >= j2)
             {
-                boolean flag4 = entity == this.minecraft.player && !this.minecraft.player.isSpectator();
+                boolean flag5 = entity == this.minecraft.player && !this.minecraft.player.isSpectator();
 
-                if ((this.entityRenderDispatcher.shouldRender(entity, frustum, d0, d1, d2) || entity.hasIndirectPassenger(this.minecraft.player)) && (entity != pDrawBlockOutline.getEntity() || pDrawBlockOutline.isDetached() || pDrawBlockOutline.getEntity() instanceof LivingEntity && ((LivingEntity)pDrawBlockOutline.getEntity()).isSleeping()) && (!(entity instanceof LocalPlayer) || pDrawBlockOutline.getEntity() == entity || flag4))
+                if ((this.entityRenderDispatcher.shouldRender(entity, frustum, d0, d1, d2) || entity.hasIndirectPassenger(this.minecraft.player)) && (renderSelf || entity != pDrawBlockOutline.getEntity() || pDrawBlockOutline.isDetached() || pDrawBlockOutline.getEntity() instanceof LivingEntity && ((LivingEntity)pDrawBlockOutline.getEntity()).isSleeping()) && (!(entity instanceof LocalPlayer) || pDrawBlockOutline.getEntity() == entity || flag5))
                 {
                     String s = entity.getClass().getName();
                     List<Entity> list = this.mapEntityLists.get(s);
@@ -1654,6 +1681,11 @@
             {
                 ++this.renderedEntities;
 
+                if (entity1 == pDrawBlockOutline.getEntity())
+                {
+                    pActiveRenderInfo.restoreRVEPos((LivingEntity)entity1);
+                }
+
                 if (entity1.tickCount == 0)
                 {
                     entity1.xOld = entity1.getX();
@@ -1689,6 +1721,12 @@
 
                 this.renderEntity(entity1, d0, d1, d2, pPartialTicks, pMatrixStack, multibuffersource);
                 this.renderedEntity = null;
+
+                if (entity1 == pDrawBlockOutline.getEntity())
+                {
+                    pActiveRenderInfo.cacheRVEPos((LivingEntity)entity1);
+                    pActiveRenderInfo.setupRVE();
+                }
             }
 
             list1.clear();
@@ -1951,11 +1989,14 @@
             multibuffersource$buffersource.endBatch();
             ShadersRender.endDebug();
             Shaders.preRenderHand();
-            Matrix4f matrix4f1 = RenderSystem.getProjectionMatrix().copy();
-            ShadersRender.renderHand0(pActiveRenderInfo, pMatrixStack, pDrawBlockOutline, pPartialTicks);
-            RenderSystem.setProjectionMatrix(matrix4f1);
+//            Matrix4f matrix4f1 = RenderSystem.getProjectionMatrix().copy();
+//            ShadersRender.renderHand0(pActiveRenderInfo, pMatrixStack, pDrawBlockOutline, pPartialTicks);
+//            RenderSystem.setProjectionMatrix(matrix4f1);
             Shaders.preWater();
         }
+        
+        boolean menuHandleft = pActiveRenderInfo.isInMenuRoom() || this.minecraft.screen != null || KeyboardHandler.Showing;
+        boolean menuhandright = menuHandleft || this.minecraft.interactTracker.hotbar >= 0 && this.minecraft.vrSettings.vrTouchHotbar;
 
         if (this.transparencyChain != null)
         {
@@ -1982,9 +2023,11 @@
             }
 
             RenderStateShard.PARTICLES_TARGET.clearRenderState();
+            pActiveRenderInfo.renderVRFabulous(pPartialTicks, this, menuhandright, menuHandleft);
         }
         else
         {
+            pActiveRenderInfo.renderVrFast(pPartialTicks, false, menuhandright, menuHandleft);
             profilerfiller.popPush("translucent");
 
             if (flag1)
@@ -2266,7 +2309,7 @@
                     if (chunkrenderdispatcher$renderchunk.regionX != l || chunkrenderdispatcher$renderchunk.regionZ != i1)
                     {
                         PairInt pairint = PairInt.of(chunkrenderdispatcher$renderchunk.regionX, chunkrenderdispatcher$renderchunk.regionZ);
-                        map1 = map.computeIfAbsent(pairint, (kx) ->
+                        map1 = map.computeIfAbsent(pairint, (p_lambda$renderBlockLayer$4_0_) ->
                         {
                             return new LinkedHashMap(8);
                         });
@@ -3269,19 +3312,19 @@
             while (iterator1.hasNext())
             {
                 ChunkRenderDispatcher.RenderChunk chunkrenderdispatcher$renderchunk1 = iterator1.next();
-                boolean flag1 = chunkrenderdispatcher$renderchunk1.isChunkRegionEmpty();
-                boolean flag;
+                boolean flag = chunkrenderdispatcher$renderchunk1.isChunkRegionEmpty();
+                boolean flag1;
 
-                if (!chunkrenderdispatcher$renderchunk1.isDirtyFromPlayer() && !flag1)
+                if (!chunkrenderdispatcher$renderchunk1.isDirtyFromPlayer() && !flag)
                 {
-                    flag = this.chunkRenderDispatcher.updateChunkLater(chunkrenderdispatcher$renderchunk1);
+                    flag1 = this.chunkRenderDispatcher.updateChunkLater(chunkrenderdispatcher$renderchunk1);
                 }
                 else
                 {
-                    flag = this.chunkRenderDispatcher.updateChunkNow(chunkrenderdispatcher$renderchunk1);
+                    flag1 = this.chunkRenderDispatcher.updateChunkNow(chunkrenderdispatcher$renderchunk1);
                 }
 
-                if (!flag)
+                if (!flag1)
                 {
                     break;
                 }
@@ -3289,7 +3332,7 @@
                 chunkrenderdispatcher$renderchunk1.setNotDirty();
                 iterator1.remove();
 
-                if (!flag1)
+                if (!flag)
                 {
                     double d0 = 2.0D * RenderChunkUtils.getRelativeBufferSize(chunkrenderdispatcher$renderchunk1);
                     d1 += d0;
@@ -3439,7 +3482,7 @@
     {
         if (!Config.isCustomEntityModels() || !CustomEntityModels.isCustomModel(pZ))
         {
-            renderShape(pMatrixStack, pBuffer, pZ.getShape(this.level, p_109644_, CollisionContext.of(pEntity)), (double)p_109644_.getX() - pX, (double)p_109644_.getY() - p_109642_, (double)p_109644_.getZ() - pY, 0.0F, 0.0F, 0.0F, 0.4F);
+            renderShape(pMatrixStack, pBuffer, pZ.getShape(this.level, p_109644_, CollisionContext.of(pEntity)), (double)p_109644_.getX() - pX, (double)p_109644_.getY() - p_109642_, (double)p_109644_.getZ() - pY, this.selR, this.selG, this.selB, 0.4F);
         }
     }
 
@@ -3632,23 +3675,23 @@
         this.playRecord(pSound, pPos, pSound == null ? null : RecordItem.getBySound(pSound));
     }
 
-    public void playRecord(@Nullable SoundEvent soundIn, BlockPos pos, @Nullable RecordItem musicDiscItem)
+    public void playRecord(@Nullable SoundEvent p_playRecord_1_, BlockPos p_playRecord_2_, @Nullable RecordItem p_playRecord_3_)
     {
-        SoundInstance soundinstance = this.playingRecords.get(pos);
+        SoundInstance soundinstance = this.playingRecords.get(p_playRecord_2_);
 
         if (soundinstance != null)
         {
             this.minecraft.getSoundManager().stop(soundinstance);
-            this.playingRecords.remove(pos);
+            this.playingRecords.remove(p_playRecord_2_);
         }
 
-        if (soundIn != null)
+        if (p_playRecord_1_ != null)
         {
-            RecordItem recorditem = RecordItem.getBySound(soundIn);
+            RecordItem recorditem = RecordItem.getBySound(p_playRecord_1_);
 
             if (Reflector.MinecraftForgeClient.exists())
             {
-                recorditem = musicDiscItem;
+                recorditem = p_playRecord_3_;
             }
 
             if (recorditem != null)
@@ -3656,12 +3699,12 @@
                 this.minecraft.gui.setNowPlaying(recorditem.getDisplayName());
             }
 
-            SoundInstance soundinstance1 = SimpleSoundInstance.forRecord(soundIn, (double)pos.getX(), (double)pos.getY(), (double)pos.getZ());
-            this.playingRecords.put(pos, soundinstance1);
+            SoundInstance soundinstance1 = SimpleSoundInstance.forRecord(p_playRecord_1_, (double)p_playRecord_2_.getX(), (double)p_playRecord_2_.getY(), (double)p_playRecord_2_.getZ());
+            this.playingRecords.put(p_playRecord_2_, soundinstance1);
             this.minecraft.getSoundManager().play(soundinstance1);
         }
 
-        this.notifyNearbyEntities(this.level, pos, soundIn != null);
+        this.notifyNearbyEntities(this.level, p_playRecord_2_, p_playRecord_1_ != null);
     }
 
     private void notifyNearbyEntities(Level pLevel, BlockPos pPos, boolean pIsPartying)
@@ -3920,6 +3963,7 @@
     public void levelEvent(Player pPlayer, int pType, BlockPos pBlockPos, int pData)
     {
         Random random = this.level.random;
+        boolean flag = this.minecraft.player != null && this.minecraft.player.isAlive() && this.minecraft.player.blockPosition().distSqr(pBlockPos) < 25.0D;
 
         switch (pType)
         {
@@ -3991,18 +4035,38 @@
                 break;
 
             case 1011:
+                if (flag)
+                {
+                    this.minecraft.vr.triggerHapticPulse(0, 250);
+                }
+
                 this.level.playLocalSound(pBlockPos, SoundEvents.IRON_DOOR_CLOSE, SoundSource.BLOCKS, 1.0F, random.nextFloat() * 0.1F + 0.9F, false);
                 break;
 
             case 1012:
+                if (flag)
+                {
+                    this.minecraft.vr.triggerHapticPulse(0, 250);
+                }
+
                 this.level.playLocalSound(pBlockPos, SoundEvents.WOODEN_DOOR_CLOSE, SoundSource.BLOCKS, 1.0F, random.nextFloat() * 0.1F + 0.9F, false);
                 break;
 
             case 1013:
+                if (flag)
+                {
+                    this.minecraft.vr.triggerHapticPulse(0, 250);
+                }
+
                 this.level.playLocalSound(pBlockPos, SoundEvents.WOODEN_TRAPDOOR_CLOSE, SoundSource.BLOCKS, 1.0F, random.nextFloat() * 0.1F + 0.9F, false);
                 break;
 
             case 1014:
+                if (flag)
+                {
+                    this.minecraft.vr.triggerHapticPulse(0, 250);
+                }
+
                 this.level.playLocalSound(pBlockPos, SoundEvents.FENCE_GATE_CLOSE, SoundSource.BLOCKS, 1.0F, random.nextFloat() * 0.1F + 0.9F, false);
                 break;
 
@@ -4023,14 +4087,32 @@
                 break;
 
             case 1019:
+                if (flag)
+                {
+                    this.minecraft.vr.triggerHapticPulse(0, 750);
+                    this.minecraft.vr.triggerHapticPulse(1, 750);
+                }
+
                 this.level.playLocalSound(pBlockPos, SoundEvents.ZOMBIE_ATTACK_WOODEN_DOOR, SoundSource.HOSTILE, 2.0F, (random.nextFloat() - random.nextFloat()) * 0.2F + 1.0F, false);
                 break;
 
             case 1020:
+                if (flag)
+                {
+                    this.minecraft.vr.triggerHapticPulse(0, 750);
+                    this.minecraft.vr.triggerHapticPulse(1, 750);
+                }
+
                 this.level.playLocalSound(pBlockPos, SoundEvents.ZOMBIE_ATTACK_IRON_DOOR, SoundSource.HOSTILE, 2.0F, (random.nextFloat() - random.nextFloat()) * 0.2F + 1.0F, false);
                 break;
 
             case 1021:
+                if (flag)
+                {
+                    this.minecraft.vr.triggerHapticPulse(0, 750);
+                    this.minecraft.vr.triggerHapticPulse(1, 750);
+                }
+
                 this.level.playLocalSound(pBlockPos, SoundEvents.ZOMBIE_BREAK_WOODEN_DOOR, SoundSource.HOSTILE, 2.0F, (random.nextFloat() - random.nextFloat()) * 0.2F + 1.0F, false);
                 break;
 
@@ -4059,10 +4141,21 @@
                 break;
 
             case 1030:
+                if (flag)
+                {
+                    this.minecraft.vr.triggerHapticPulse(0, 500);
+                }
+
                 this.level.playLocalSound(pBlockPos, SoundEvents.ANVIL_USE, SoundSource.BLOCKS, 1.0F, random.nextFloat() * 0.1F + 0.9F, false);
                 break;
 
             case 1031:
+                if (flag)
+                {
+                    this.minecraft.vr.triggerHapticPulse(0, 1250);
+                    this.minecraft.vr.triggerHapticPulse(1, 1250);
+                }
+
                 this.level.playLocalSound(pBlockPos, SoundEvents.ANVIL_LAND, SoundSource.BLOCKS, 0.3F, this.level.random.nextFloat() * 0.1F + 0.9F, false);
                 break;
 
@@ -4083,6 +4176,11 @@
                 break;
 
             case 1036:
+                if (flag)
+                {
+                    this.minecraft.vr.triggerHapticPulse(0, 250);
+                }
+
                 this.level.playLocalSound(pBlockPos, SoundEvents.IRON_TRAPDOOR_CLOSE, SoundSource.BLOCKS, 1.0F, random.nextFloat() * 0.1F + 0.9F, false);
                 break;
 
@@ -4442,9 +4540,9 @@
         return this.chunksToCompile.size();
     }
 
-    public ChunkRenderDispatcher.RenderChunk getRenderChunk(BlockPos pos)
+    public ChunkRenderDispatcher.RenderChunk getRenderChunk(BlockPos p_getRenderChunk_1_)
     {
-        return this.viewArea.getRenderChunkAt(pos);
+        return this.viewArea.getRenderChunkAt(p_getRenderChunk_1_);
     }
 
     public ClientLevel getWorld()
@@ -4529,11 +4627,11 @@
         return this.renderInfosTileEntities;
     }
 
-    private void checkLoadVisibleChunks(Camera activeRenderInfo, Frustum icamera, boolean spectator)
+    private void checkLoadVisibleChunks(Camera p_checkLoadVisibleChunks_1_, Frustum p_checkLoadVisibleChunks_2_, boolean p_checkLoadVisibleChunks_3_)
     {
         if (this.loadVisibleChunksCounter == 0)
         {
-            this.loadAllVisibleChunks(activeRenderInfo, icamera, spectator);
+            this.loadAllVisibleChunks(p_checkLoadVisibleChunks_1_, p_checkLoadVisibleChunks_2_, p_checkLoadVisibleChunks_3_);
             this.minecraft.gui.getChat().removeById(201435902);
         }
 
@@ -4543,7 +4641,7 @@
         }
     }
 
-    private void loadAllVisibleChunks(Camera activeRenderInfo, Frustum icamera, boolean spectator)
+    private void loadAllVisibleChunks(Camera p_loadAllVisibleChunks_1_, Frustum p_loadAllVisibleChunks_2_, boolean p_loadAllVisibleChunks_3_)
     {
         int i = this.minecraft.options.ofChunkUpdates;
         boolean flag = this.minecraft.options.ofLazyChunkLoading;
@@ -4567,7 +4665,7 @@
                 for (int j1 = 0; j1 < 100; ++j1)
                 {
                     levelrenderer.needsUpdate();
-                    levelrenderer.setupRender(activeRenderInfo, icamera, false, this.frameId++, spectator);
+                    levelrenderer.setupRender(p_loadAllVisibleChunks_1_, p_loadAllVisibleChunks_2_, false, this.frameId++, p_loadAllVisibleChunks_3_);
 
                     if (!levelrenderer.hasRenderedAllChunks())
                     {
@@ -4697,6 +4795,43 @@
         return this.cloudsTarget;
     }
 
+    public void setShaderGroup()
+    {
+        this.transparencyChain = null;
+        this.translucentTarget = null;
+        this.itemEntityTarget = null;
+        this.particlesTarget = null;
+        this.weatherTarget = null;
+        this.cloudsTarget = null;
+        this.alphaSortVRHandsFramebuffer = null;
+        this.alphaSortVROccludedFramebuffer = null;
+        this.alphaSortVRUnoccludedFramebuffer = null;
+        PostChain postchain = this.minecraft.vrRenderer.alphaShaders.get(this.minecraft.getMainRenderTarget().name);
+
+        if (postchain != null)
+        {
+            this.transparencyChain = postchain;
+            this.translucentTarget = postchain.getTempTarget("translucent");
+            this.itemEntityTarget = postchain.getTempTarget("itemEntity");
+            this.particlesTarget = postchain.getTempTarget("particles");
+            this.weatherTarget = postchain.getTempTarget("weather");
+            this.cloudsTarget = postchain.getTempTarget("clouds");
+            this.alphaSortVRHandsFramebuffer = postchain.getTempTarget("vrhands");
+            this.alphaSortVROccludedFramebuffer = postchain.getTempTarget("vroccluded");
+            this.alphaSortVRUnoccludedFramebuffer = postchain.getTempTarget("vrunoccluded");
+        }
+
+        this.entityEffect = null;
+        this.entityTarget = null;
+        postchain = this.minecraft.vrRenderer.entityShaders.get(this.minecraft.getMainRenderTarget().name);
+
+        if (postchain != null)
+        {
+            this.entityEffect = postchain;
+            this.entityTarget = this.entityEffect.getTempTarget("final");
+        }
+    }
+
     public static class RenderChunkInfo
     {
         public final ChunkRenderDispatcher.RenderChunk chunk;
