--- a/net/minecraft/client/renderer/RenderStateShard.java
+++ b/net/minecraft/client/renderer/RenderStateShard.java
@@ -679,7 +679,7 @@
                 RenderSystem.translatef(17.0F / (float)p_110311_, (2.0F + (float)p_110311_ / 1.5F) * ((float)(Util.getMillis() % 800000L) / 800000.0F), 0.0F);
                 RenderSystem.rotatef(((float)(p_110311_ * p_110311_) * 4321.0F + (float)p_110311_ * 9.0F) * 2.0F, 0.0F, 0.0F, 1.0F);
                 RenderSystem.scalef(4.5F - (float)p_110311_ / 4.0F, 4.5F - (float)p_110311_ / 4.0F, 1.0F);
-                RenderSystem.mulTextureByProjModelView();
+                GlStateManager.mulTextureByModelView();
                 RenderSystem.matrixMode(5888);
                 RenderSystem.setupEndPortalTexGen();
             }, () ->
