--- a/net/minecraft/client/gui/components/DebugScreenOverlay.java
+++ b/net/minecraft/client/gui/components/DebugScreenOverlay.java
@@ -583,7 +583,7 @@
         {
             list.add("");
 
-            for (String s1 : (Collection)Reflector.call(Reflector.BrandingControl_getBrandings, true, false))
+            for (String s1 : (Collection<String>)Reflector.call(Reflector.BrandingControl_getBrandings, true, false))
             {
                 if (!s1.startsWith("Minecraft "))
                 {
