--- a/net/minecraft/client/tutorial/Tutorial.java
+++ b/net/minecraft/client/tutorial/Tutorial.java
@@ -95,8 +95,8 @@
         {
             this.stop();
         }
-
-        this.instance = this.minecraft.options.tutorialStep.create(this);
+        //some steps don't work in VR.
+        this.instance = TutorialSteps.NONE.create(this);
     }
 
     public void addTimedToast(TutorialToast pToast, int pDurationTicks)
