--- a/net/minecraft/world/scores/Scoreboard.java
+++ b/net/minecraft/world/scores/Scoreboard.java
@@ -107,7 +107,7 @@
             {
                 return Maps.newHashMap();
             });
-            return map.computeIfAbsent(pUsername, (pUsername) ->
+            return map.computeIfAbsent(pUsername, (pUsername1) ->
             {
                 Score score = new Score(this, pUsername, p_83472_);
                 score.setScore(0);
