--- a/net/minecraft/client/gui/screens/inventory/CreativeModeInventoryScreen.java
+++ b/net/minecraft/client/gui/screens/inventory/CreativeModeInventoryScreen.java
@@ -46,6 +46,7 @@
 import net.minecraft.world.item.TooltipFlag;
 import net.minecraft.world.item.enchantment.Enchantment;
 import net.minecraft.world.item.enchantment.EnchantmentHelper;
+import org.vivecraft.asm.ASMDelegator;
 
 public class CreativeModeInventoryScreen extends EffectRenderingInventoryScreen<CreativeModeInventoryScreen.ItemPickerMenu>
 {
@@ -455,6 +456,7 @@
             (this.menu).items.addAll(searchtree.search(s.toLowerCase(Locale.ROOT)));
         }
 
+        ASMDelegator.addCreativeSearch(this.searchBox.getValue(), (this.menu).items);
         this.scrollOffs = 0.0F;
         this.menu.scrollTo(0.0F);
     }
@@ -466,25 +468,25 @@
 
         if (i == -1)
         {
-            predicate = (p_98609_) ->
+            predicate = (p_214084_1_) ->
             {
-                return p_98609_.getPath().contains(p_98620_);
+                return p_214084_1_.getPath().contains(p_98620_);
             };
         }
         else
         {
             String s = p_98620_.substring(0, i).trim();
             String s1 = p_98620_.substring(i + 1).trim();
-            predicate = (p_98606_) ->
+            predicate = (p_214081_2_) ->
             {
-                return p_98606_.getNamespace().contains(s) && p_98606_.getPath().contains(s1);
+                return p_214081_2_.getNamespace().contains(s) && p_214081_2_.getPath().contains(s1);
             };
         }
 
         TagCollection<Item> tagcollection = ItemTags.getAllTags();
-        tagcollection.getAvailableTags().stream().filter(predicate).forEach((p_98552_) ->
+        tagcollection.getAvailableTags().stream().filter(predicate).forEach((p_214082_2_) ->
         {
-            Tag tag = this.visibleTags.put(p_98552_, tagcollection.getTag(p_98552_));
+            Tag tag = this.visibleTags.put(p_214082_2_, tagcollection.getTag(p_214082_2_));
         });
     }
 
@@ -593,6 +595,7 @@
         else if (p_98561_ != CreativeModeTab.TAB_SEARCH)
         {
             p_98561_.fillItemList((this.menu).items);
+            ASMDelegator.addCreativeItems(p_98561_, (this.menu).items);
         }
 
         if (p_98561_ == CreativeModeTab.TAB_INVENTORY)
@@ -789,11 +792,11 @@
                 }
             }
 
-            this.visibleTags.forEach((p_98569_, p_98570_) ->
+            this.visibleTags.forEach((p_214083_2_, p_214083_3_) ->
             {
-                if (p_98570_.contains(item))
+                if (p_214083_3_.contains(item))
                 {
-                    list1.add(1, (new TextComponent("#" + p_98569_)).withStyle(ChatFormatting.DARK_PURPLE));
+                    list1.add(1, (new TextComponent("#" + p_214083_2_)).withStyle(ChatFormatting.DARK_PURPLE));
                 }
             });
 
@@ -1017,7 +1020,7 @@
 
         public ItemPickerMenu(Player p_98641_)
         {
-            super((MenuType<?>)null, 0);
+            super((MenuType)null, 0);
             Inventory inventory = p_98641_.inventory;
 
             for (int i = 0; i < 5; ++i)
