--- a/net/minecraft/client/gui/screens/inventory/SignEditScreen.java
+++ b/net/minecraft/client/gui/screens/inventory/SignEditScreen.java
@@ -27,6 +27,7 @@
 import net.minecraft.world.level.block.StandingSignBlock;
 import net.minecraft.world.level.block.entity.SignBlockEntity;
 import net.minecraft.world.level.block.state.BlockState;
+import org.vivecraft.gameplay.screenhandlers.KeyboardHandler;
 
 public class SignEditScreen extends Screen
 {
@@ -40,9 +41,9 @@
     public SignEditScreen(SignBlockEntity p_99260_)
     {
         super(new TranslatableComponent("sign.edit"));
-        this.messages = IntStream.range(0, 4).mapToObj(p_99260_::getMessage).map(Component::getString).toArray((p_99265_) ->
+        this.messages = IntStream.range(0, 4).mapToObj(p_99260_::getMessage).map(Component::getString).toArray((p_243354_0_) ->
         {
-            return new String[p_99265_];
+            return new String[p_243354_0_];
         });
         this.sign = p_99260_;
     }
@@ -50,21 +51,22 @@
     protected void init()
     {
         this.minecraft.keyboardHandler.setSendRepeatsToGui(true);
-        this.addButton(new Button(this.width / 2 - 100, this.height / 4 + 120, 200, 20, CommonComponents.GUI_DONE, (p_99276_) ->
+        this.addButton(new Button(this.width / 2 - 100, this.height / 4 + 120, 200, 20, CommonComponents.GUI_DONE, (p_238847_1_) ->
         {
             this.onDone();
         }));
         this.sign.setEditable(false);
+        KeyboardHandler.setOverlayShowing(true);
         this.signField = new TextFieldHelper(() ->
         {
             return this.messages[this.line];
-        }, (p_99282_) ->
+        }, (p_238850_1_) ->
         {
-            this.messages[this.line] = p_99282_;
-            this.sign.setMessage(this.line, new TextComponent(p_99282_));
-        }, TextFieldHelper.createClipboardGetter(this.minecraft), TextFieldHelper.createClipboardSetter(this.minecraft), (p_99280_) ->
+            this.messages[this.line] = p_238850_1_;
+            this.sign.setMessage(this.line, new TextComponent(p_238850_1_));
+        }, TextFieldHelper.createClipboardGetter(this.minecraft), TextFieldHelper.createClipboardSetter(this.minecraft), (p_238848_1_) ->
         {
-            return this.minecraft.font.width(p_99280_) <= 90;
+            return this.minecraft.font.width(p_238848_1_) <= 90;
         });
     }
 
@@ -79,6 +81,7 @@
         }
 
         this.sign.setEditable(true);
+        KeyboardHandler.setOverlayShowing(false);
     }
 
     public void tick()
