--- a/net/minecraft/world/item/Item.java
+++ b/net/minecraft/world/item/Item.java
@@ -141,14 +141,14 @@
         {
             ItemStack itemstack = p_41433_.getItemInHand(p_41434_);
 
-            if (p_41433_.canEat(this.getFoodProperties().canAlwaysEat()))
+            if (!p_41433_.canEat(this.getFoodProperties().canAlwaysEat()) && !itemstack.getHoverName().getString().equals("EAT ME"))
             {
-                p_41433_.startUsingItem(p_41434_);
-                return InteractionResultHolder.consume(itemstack);
+                return InteractionResultHolder.fail(itemstack);
             }
             else
             {
-                return InteractionResultHolder.fail(itemstack);
+                p_41433_.startUsingItem(p_41434_);
+                return InteractionResultHolder.consume(itemstack);
             }
         }
         else
