--- a/net/minecraft/world/item/Item.java
+++ b/net/minecraft/world/item/Item.java
@@ -141,14 +141,14 @@
         {
             ItemStack itemstack = pLevel.getItemInHand(pPlayer);
 
-            if (pLevel.canEat(this.getFoodProperties().canAlwaysEat()))
+            if (!pLevel.canEat(this.getFoodProperties().canAlwaysEat()) && !itemstack.getHoverName().getString().equals("EAT ME"))
             {
-                pLevel.startUsingItem(pPlayer);
-                return InteractionResultHolder.consume(itemstack);
+                return InteractionResultHolder.fail(itemstack);
             }
             else
             {
-                return InteractionResultHolder.fail(itemstack);
+                pLevel.startUsingItem(pPlayer);
+                return InteractionResultHolder.consume(itemstack);
             }
         }
         else
