--- a/net/optifine/shaders/config/ShaderPackParser.java
+++ b/net/optifine/shaders/config/ShaderPackParser.java
@@ -256,7 +256,7 @@
         String s = "profile.";
         List<ShaderProfile> list = new ArrayList<>();
 
-        for (String s1 : props.keySet())
+        for (String s1 : (Set<String>)(Set<?>)props.keySet())
         {
             if (s1.startsWith(s))
             {
@@ -289,7 +289,7 @@
         Pattern pattern = Pattern.compile("program\\.([^.]+)\\.enabled");
         Map<String, IExpressionBool> map = new HashMap<>();
 
-        for (String s1 : props.keySet())
+        for (String s1 : (Set<String>)(Set<?>)props.keySet())
         {
             Matcher matcher = pattern.matcher(s1);
 
@@ -818,7 +818,7 @@
         Map<String, IExpression> map = new HashMap<>();
         List<CustomUniform> list = new ArrayList<>();
 
-        for (String s4 : props.keySet())
+        for (String s4 : (Set<String>)(Set<?>)props.keySet())
         {
             String[] astring = Config.tokenize(s4, ".");
 
@@ -913,7 +913,7 @@
 
     public static void parseAlphaStates(Properties props)
     {
-        for (String s : props.keySet())
+        for (String s : (Set<String>)(Set<?>)props.keySet())
         {
             String[] astring = Config.tokenize(s, ".");
 
@@ -984,7 +984,7 @@
 
     public static void parseBlendStates(Properties props)
     {
-        for (String s : props.keySet())
+        for (String s : (Set<String>)(Set<?>)props.keySet())
         {
             String[] astring = Config.tokenize(s, ".");
 
@@ -1085,7 +1085,7 @@
 
     public static void parseRenderScales(Properties props)
     {
-        for (String s : props.keySet())
+        for (String s : (Set<String>)(Set<?>)props.keySet())
         {
             String[] astring = Config.tokenize(s, ".");
 
@@ -1156,7 +1156,7 @@
 
     public static void parseBuffersFlip(Properties props)
     {
-        for (String s : props.keySet())
+        for (String s : (Set<String>)(Set<?>)props.keySet())
         {
             String[] astring = Config.tokenize(s, ".");
 
@@ -1242,7 +1242,7 @@
     {
         DynamicDimension[] adynamicdimension = new DynamicDimension[countBuffers];
 
-        for (String s : props.keySet())
+        for (String s : (Set<String>)(Set<?>)props.keySet())
         {
             if (s.startsWith("size.buffer."))
             {
