--- a/net/minecraft/client/renderer/entity/ItemRenderer.java
+++ b/net/minecraft/client/renderer/entity/ItemRenderer.java
@@ -17,6 +17,11 @@
 import java.util.Random;
 import java.util.Set;
 import javax.annotation.Nullable;
+
+import org.vivecraft.gameplay.trackers.SwingTracker;
+import org.vivecraft.gameplay.trackers.TelescopeTracker;
+import org.vivecraft.render.SpecialItemRenderer;
+
 import net.minecraft.CrashReport;
 import net.minecraft.CrashReportCategory;
 import net.minecraft.ReportedException;
@@ -65,6 +70,7 @@
 import net.optifine.CustomItems;
 import net.optifine.EmissiveTextures;
 import net.optifine.reflect.Reflector;
+import net.optifine.render.GlAlphaState;
 import net.optifine.render.VertexBuilderWrapper;
 import net.optifine.shaders.Shaders;
 
@@ -84,6 +90,11 @@
     private final BlockEntityWithoutLevelRenderer blockEntityRenderer;
     public ModelManager modelManager = null;
     private static boolean renderItemGui = false;
+    public static boolean ismainhand = false;
+    public static boolean isfphand = false;
+    float fade = 1.0F;
+    float manualFade = 1.0F;
+    private GlAlphaState alphaState = new GlAlphaState();
 
     public ItemRenderer(TextureManager p_174225_, ModelManager p_174226_, ItemColors p_174227_, BlockEntityWithoutLevelRenderer p_174228_)
     {
@@ -166,6 +177,16 @@
             }
 
             p_115147_.translate(-0.5D, -0.5D, -0.5D);
+            LocalPlayer localplayer = Minecraft.getInstance().player;
+
+            if (localplayer != null && isfphand)
+            {
+                this.fade = SwingTracker.getItemFade(localplayer, p_115144_);
+            }
+            else
+            {
+                this.fade = this.manualFade;
+            }
 
             if (!p_115151_.isCustomRenderer() && (!p_115144_.is(Items.TRIDENT) || flag))
             {
@@ -188,6 +209,12 @@
                 else
                 {
                     RenderType rendertype = ItemBlockRenderTypes.getRenderType(p_115144_, flag1);
+
+                    if (isfphand && this.fade < 1.0F)
+                    {
+                        rendertype = Sheets.translucentCullBlockSheet();
+                    }
+
                     VertexConsumer vertexconsumer;
 
                     if (p_115144_.is(Items.COMPASS) && p_115144_.hasFoil())
@@ -360,6 +387,19 @@
                 }
             }
 
+            if (Minecraft.getInstance().jumpTracker.isBoots(p_115166_))
+            {
+                k = this.makeColor(1, 0, 255, 0);
+            }
+            else if (Minecraft.getInstance().climbTracker.isClaws(p_115166_))
+            {
+                k = this.makeColor(1, 130, 0, 75);
+            }
+            else if (TelescopeTracker.isTelescope(p_115166_))
+            {
+                k = this.makeColor(1, 190, 110, 135);
+            }
+
             float f = (float)(k >> 16 & 255) / 255.0F;
             float f1 = (float)(k >> 8 & 255) / 255.0F;
             float f2 = (float)(k & 255) / 255.0F;
@@ -370,7 +410,7 @@
             }
             else
             {
-                p_115164_.putBulkData(posestack$pose, bakedquad, f, f1, f2, p_115167_, p_115168_);
+                SpecialItemRenderer.addQuad(p_115164_, posestack$pose, bakedquad, f, f1, f2, this.fade, p_115167_, p_115168_);
             }
         }
     }
@@ -431,6 +471,10 @@
         RenderSystem.enableBlend();
         RenderSystem.blendFunc(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA);
         RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
+        //Vivecraft.. wut
+        GlStateManager.getAlphaState(this.alphaState);
+        GlStateManager.enableAlphaTest();
+        //
         PoseStack posestack = RenderSystem.getModelViewStack();
         posestack.pushPose();
         posestack.translate((double)p_115129_, (double)p_115130_, (double)(100.0F + this.blitOffset));
@@ -635,4 +679,24 @@
     {
         return VanillaResourceType.MODELS;
     }
+
+    private int makeColor(int a, int r, int g, int b)
+    {
+        return a << 24 | r << 16 | g << 8 | b;
+    }
+
+    public float getCurrentFade()
+    {
+        return this.fade;
+    }
+
+    public float getFade()
+    {
+        return this.manualFade;
+    }
+
+    public void setFade(float fade)
+    {
+        this.manualFade = fade;
+    }
 }
