--- a/net/optifine/model/ListQuadsOverlay.java
+++ b/net/optifine/model/ListQuadsOverlay.java
@@ -11,7 +11,7 @@
 {
     private List<BakedQuad> listQuads = new ArrayList<>();
     private List<BlockState> listBlockStates = new ArrayList<>();
-    private List<BakedQuad> listQuadsSingle = Arrays.asList();
+    private List<BakedQuad> listQuadsSingle = Arrays.asList(new BakedQuad[1]);
 
     public void addQuad(BakedQuad quad, BlockState blockState)
     {
