--- a/net/minecraft/world/level/entity/EntitySectionStorage.java
+++ b/net/minecraft/world/level/entity/EntitySectionStorage.java
@@ -140,7 +140,7 @@
     public LongSet getAllChunksWithExistingSections()
     {
         LongSet longset = new LongOpenHashSet();
-        this.sections.keySet().forEach((p_156884_1_) ->
+        this.sections.keySet().forEach((long p_156884_1_) ->
         {
             longset.add(getChunkKeyFromSectionKey(p_156884_1_));
         });
