--- a/net/optifine/Config.java
+++ b/net/optifine/Config.java
@@ -144,13 +144,10 @@
 
     public static void initGameSettings(Options settings)
     {
-        if (gameSettings == null)
-        {
-            gameSettings = settings;
-            updateAvailableProcessors();
-            ReflectorForge.putLaunchBlackboard("optifine.ForgeSplashCompatible", Boolean.TRUE);
-            antialiasingLevel = gameSettings.ofAaLevel;
-        }
+        gameSettings = settings;
+        updateAvailableProcessors();
+        ReflectorForge.putLaunchBlackboard("optifine.ForgeSplashCompatible", Boolean.TRUE);
+        antialiasingLevel = gameSettings.ofAaLevel;
     }
 
     public static void initDisplay()
@@ -163,14 +160,11 @@
 
     public static void checkInitialized()
     {
-        if (!initialized)
+        if (!initialized && Minecraft.getInstance().getWindow() != null)
         {
-            if (Minecraft.getInstance().getWindow() != null)
-            {
-                initialized = true;
-                checkOpenGlCaps();
-                startVersionCheckThread();
-            }
+            initialized = true;
+            checkOpenGlCaps();
+            startVersionCheckThread();
         }
     }
 
@@ -228,9 +222,9 @@
 
                 build = readLines(inputstream)[0];
             }
-            catch (Exception exception)
+            catch (Exception exception1)
             {
-                warn("" + exception.getClass().getName() + ": " + exception.getMessage());
+                warn("" + exception1.getClass().getName() + ": " + exception1.getMessage());
                 build = "";
             }
         }
@@ -422,9 +416,9 @@
                 }
             }
         }
-        catch (Exception exception)
+        catch (Exception exception1)
         {
-            error("", exception);
+            error("", exception1);
             return def;
         }
     }
@@ -472,9 +466,9 @@
             String s = GL11.glGetString(GL11.GL_EXTENSIONS);
             return s.split(" ");
         }
-        catch (Exception exception)
+        catch (Exception exception1)
         {
-            error("", exception);
+            error("", exception1);
             return new String[0];
         }
     }
@@ -529,9 +523,9 @@
                 }
             }
         }
-        catch (Throwable throwable)
+        catch (Throwable throwable1)
         {
-            warn(throwable.getClass().getName() + ": " + throwable.getMessage());
+            warn(throwable1.getClass().getName() + ": " + throwable1.getMessage());
         }
     }
 
@@ -1161,7 +1155,7 @@
             Resource resource = resourceManager.getResource(location);
             return resource != null;
         }
-        catch (IOException ioexception)
+        catch (IOException ioexception1)
         {
             return false;
         }
@@ -1215,7 +1209,7 @@
                     astring[i] = apackresources[i].getName();
                 }
 
-                return arrayToString((Object[])astring);
+                return arrayToString(astring);
             }
         }
     }
@@ -2064,9 +2058,9 @@
 
         if (showFrameTime)
         {
-            String s1 = String.format("%.1f", 1000.0D / (double)limit(i, 1, Integer.MAX_VALUE));
-            String s = String.format("%.1f", 1000.0D / (double)limit(j, 1, Integer.MAX_VALUE));
-            return "" + s1 + "/" + s + " ms";
+            String s = String.format("%.1f", 1000.0D / (double)limit(i, 1, Integer.MAX_VALUE));
+            String s1 = String.format("%.1f", 1000.0D / (double)limit(j, 1, Integer.MAX_VALUE));
+            return "" + s + "/" + s1 + " ms";
         }
         else
         {
@@ -2350,14 +2344,7 @@
 
     public static boolean isIntegratedServerRunning()
     {
-        if (minecraft.getSingleplayerServer() == null)
-        {
-            return false;
-        }
-        else
-        {
-            return minecraft.isLocalServer();
-        }
+        return minecraft.getSingleplayerServer() == null ? false : minecraft.isLocalServer();
     }
 
     public static IntBuffer createDirectIntBuffer(int capacity)
