--- a/net/minecraft/client/renderer/block/LiquidBlockRenderer.java
+++ b/net/minecraft/client/renderer/block/LiquidBlockRenderer.java
@@ -83,7 +83,7 @@
     {
         return !isFaceOccludedBySelf(p_110949_, p_110950_, p_110952_, p_110953_) && !isNeighborSameFluid(p_110949_, p_110950_, p_110953_, p_110951_);
     }
-
+    public static boolean skipStupidGoddamnChunkBoundaryClipping;
     public boolean tesselate(BlockAndTintGetter p_110955_, BlockPos pLightReader, VertexConsumer pPos, FluidState pVertexBuilder)
     {
         BlockState blockstate = pVertexBuilder.createLegacyBlock();
@@ -150,11 +150,13 @@
                 float f8 = this.getWaterHeight(p_110955_, pLightReader.south(), pVertexBuilder.getType());
                 float f9 = this.getWaterHeight(p_110955_, pLightReader.east().south(), pVertexBuilder.getType());
                 float f10 = this.getWaterHeight(p_110955_, pLightReader.east(), pVertexBuilder.getType());
-                double d0 = (double)(pLightReader.getX() & 15);
-                double d1 = (double)(pLightReader.getY() & 15);
-                double d2 = (double)(pLightReader.getZ() & 15);
-
-                if (Config.isRenderRegions())
+                //double d0 = (double)(pLightReader.getX() & 15);
+                //double d1 = (double)(pLightReader.getY() & 15);
+                //double d2 = (double)(pLightReader.getZ() & 15);
+                double d0 = skipStupidGoddamnChunkBoundaryClipping ? pLightReader.getX() : (double)(pLightReader.getX() & 15);
+                double d1 = skipStupidGoddamnChunkBoundaryClipping ? pLightReader.getY() : (double)(pLightReader.getY() & 15);
+                double d2 = skipStupidGoddamnChunkBoundaryClipping ? pLightReader.getZ() : (double)(pLightReader.getZ() & 15);
+                if (Config.isRenderRegions() && !skipStupidGoddamnChunkBoundaryClipping)
                 {
                     int j = pLightReader.getX() >> 4 << 4;
                     int k = pLightReader.getY() >> 4 << 4;
