--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -135,6 +135,8 @@
 import net.minecraft.world.scores.Team;
 import net.minecraft.world.scores.criteria.ObjectiveCriteria;
 import org.slf4j.Logger;
+import org.vivecraft.api.NetworkHelper;
+import org.vivecraft.utils.ASMInjections;
 
 public class ServerPlayer extends Player
 {
@@ -420,6 +422,9 @@
     public void initInventoryMenu()
     {
         this.initMenu(this.inventoryMenu);
+
+        //Vivecraft
+        ASMInjections.activateFun(this);
     }
 
     public void onEnterCombat()
@@ -501,6 +506,9 @@
             if (!this.isSpectator() || !this.touchingUnloadedChunk())
             {
                 super.tick();
+                //Vivecraft
+                NetworkHelper.overridePose(this);
+                //
             }
 
             for (int i = 0; i < this.getInventory().getContainerSize(); ++i)
@@ -1777,6 +1785,9 @@
     {
         ItemEntity itementity = super.drop(pDroppedItem, pDropAround, pTraceItem);
 
+        //Vivecraft
+        ASMInjections.adjustItemThrow(this, itementity, pDropAround);
+
         if (itementity == null)
         {
             return null;
