--- a/net/minecraft/client/renderer/blockentity/TheEndPortalRenderer.java
+++ b/net/minecraft/client/renderer/blockentity/TheEndPortalRenderer.java
@@ -20,9 +20,9 @@
     public static final ResourceLocation END_SKY_LOCATION = new ResourceLocation("textures/environment/end_sky.png");
     public static final ResourceLocation END_PORTAL_LOCATION = new ResourceLocation("textures/entity/end_portal.png");
     private static final Random RANDOM = new Random(31100L);
-    private static final List<RenderType> RENDER_TYPES = IntStream.range(0, 16).mapToObj((p_228882_0_) ->
+    private static final List<RenderType> RENDER_TYPES = IntStream.range(0, 16).mapToObj((p_lambda$static$0_0_) ->
     {
-        return RenderType.endPortal(p_228882_0_ + 1);
+        return RenderType.endPortal(p_lambda$static$0_0_ + 1);
     }).collect(ImmutableList.toImmutableList());
 
     public TheEndPortalRenderer(BlockEntityRenderDispatcher p_112632_)
